{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "armed-modeling",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import altair as alt\n",
    "import os, random\n",
    "import gensim\n",
    "from Bio import SeqIO\n",
    "from gensim.models import Word2Vec\n",
    "from sklearn.cluster import KMeans, OPTICS\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn import preprocessing, model_selection, metrics\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "isolated-leadership",
   "metadata": {},
   "source": [
    "## Defining paths for each and every omic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "optical-category",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_root_data = os.path.join ('..', 'Data', 'Extracted', 'Databases')\n",
    "path_all_fasta = os.path.join (path_root_data, 'fasta_files', 'AllBins')\n",
    "path_normalised_metabolomics = os.path.join (path_root_data, 'Metabolomics', 'Normalised_Tables')\n",
    "path_model_save_root = 'Saved_models'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "departmental-missile",
   "metadata": {},
   "outputs": [],
   "source": [
    "num_of_mags = len([i for i in os.listdir(path_all_fasta) if (i[-2:] == 'fa')])\n",
    "SEED = 42\n",
    "END = 25\n",
    "num_of_workers = 3\n",
    "random.seed(SEED)\n",
    "np.random.seed(SEED)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "opposite-entrance",
   "metadata": {},
   "source": [
    "---\n",
    "# GENOMIC ANALYSIS\n",
    "---\n",
    "## MAG-related functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "fallen-anchor",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function that splits each genome into k-mers thus creating even longer sentence (MAG)\n",
    "# It returns tokenized genome i.e. [kmer, kmer,...]\n",
    "def split_genome (genome, k = 5):\n",
    "    new_genome = []\n",
    "    n = len(genome)\n",
    "    \n",
    "    if n-k <=0:\n",
    "        return genome\n",
    "    else:\n",
    "        for i in range(n-k):\n",
    "            new_genome.append(genome[i:i+k])\n",
    "        \n",
    "        return new_genome\n",
    "\n",
    "\n",
    "def vectorize_one_mag (one_mag, w2v_model):\n",
    "    \n",
    "    # We have to generate vectors for each word in one MAG and then create vector representation of that MAG\n",
    "    # by averaging vectors of its words\n",
    "    zero_vector = np.zeros(w2v_model.vector_size)\n",
    "    word_vectors = []\n",
    "    one_mag_vector = []\n",
    "    \n",
    "    for sentence in one_mag:\n",
    "        for word in sentence:\n",
    "            if word in w2v_model.wv:\n",
    "                try:\n",
    "                    word_vectors.append(w2v_model.wv[word])\n",
    "                except KeyError:\n",
    "                    print ('Key Error')\n",
    "                    continue\n",
    "    \n",
    "    if word_vectors:\n",
    "        word_vectors = np.asarray(word_vectors)\n",
    "        one_mag_vector = word_vectors.mean (axis=0)\n",
    "    \n",
    "    else:\n",
    "        one_mag_vector = zero_vector\n",
    "    \n",
    "    return one_mag_vector\n",
    "\n",
    "\n",
    "# Function that vectorizes a MAG (document) with a pretrained word2vec model. It returns vector representation of a given MAG\n",
    "# Vectorization is done by averaging word (k-mer) vectors for the whole document (MAG)\n",
    "def vectorize_mags (w2v_model, end = 25):\n",
    "    \n",
    "    print ('Vectorizing MAGs')\n",
    "    \n",
    "    fasta_files = [i for i in os.listdir(path_all_fasta) if (i[-2:] == 'fa')]\n",
    "    list_of_mag_vectors = []\n",
    "    \n",
    "    # This was done so that I could work with first 'end' FASTA files only. Otherwise, I should just remove: i, and enumerate\n",
    "    for i, fasta_file_name in enumerate(fasta_files):\n",
    "        \n",
    "        if i == end:\n",
    "            break\n",
    "        \n",
    "        else:\n",
    "            with open(os.path.join(path_all_fasta, fasta_file_name), 'r') as input_file:\n",
    "                \n",
    "                one_mag = []\n",
    "                for fasta_string in SeqIO.parse(input_file, \"fasta\"):\n",
    "                    \n",
    "                    # Get kmers of a genome and create a sentence (list of words)\n",
    "                    temp_kmers = split_genome (str(fasta_string.seq))\n",
    "                    \n",
    "                    # Create a document (list of sentences)\n",
    "                    one_mag.append(temp_kmers)\n",
    "                    \n",
    "                # Vectorizing MAGs one by one\n",
    "                list_of_mag_vectors.append (vectorize_one_mag (one_mag, w2v_model))\n",
    "    \n",
    "    print ('Finished vectorizing')\n",
    "    \n",
    "    return list_of_mag_vectors\n",
    "    \n",
    "    \n",
    "\n",
    "# If one wants to import MAGs in order to vectorize them, one should use start argument in order to skip first 'start' MAGs\n",
    "# If one wants to import MAGs to train word2vec model, one should use only end argument, so that first 'end' MAGs are used for training\n",
    "# Todo: Implement randomisation in picking MAGs for training, and don't use first 'start' MAGs for training\n",
    "# (create list of indexes from 0 to 1364, use sklearn split train test, then traverse directory and use only MAGs with indexes in train for training w2v)\n",
    "def import_mags_and_build_model (end = 25, path_all_fasta = path_all_fasta):\n",
    "    \n",
    "    print ('Importing MAGs')\n",
    "    \n",
    "    # There are 1364 MAGs enclosed in FASTA files\n",
    "    # I have to traverse every FASTA file, and in each file every sequence\n",
    "    \n",
    "    fasta_files = [i for i in os.listdir(path_all_fasta) if (i[-2:] == 'fa')]\n",
    "    \n",
    "    # This was done so that I could work with first 100 FASTA files only. Otherwise, I should just remove: i, and enumerate\n",
    "    for i, fasta_file_name in enumerate(fasta_files):\n",
    "        \n",
    "        if i == end:\n",
    "            break\n",
    "        \n",
    "        else:\n",
    "            with open(os.path.join(path_all_fasta, fasta_file_name), 'r') as input_file:\n",
    "                \n",
    "                one_mag = []\n",
    "                for fasta_string in SeqIO.parse(input_file, \"fasta\"):\n",
    "                    \n",
    "                    # Get kmers of a genome and create a sentence (list of words)\n",
    "                    temp_kmers = split_genome (str(fasta_string.seq))\n",
    "                    \n",
    "                    # Create a document (list of sentences)\n",
    "                    one_mag.append(temp_kmers)\n",
    "                    \n",
    "                # If we do not have a model, we build one\n",
    "                if i == 0:\n",
    "                    print ('Building w2v model')\n",
    "                    # We build our model on the first MAG\n",
    "                    w2v_model = Word2Vec (sentences = one_mag, size = 100, workers = num_of_workers, seed=SEED)\n",
    "                \n",
    "                # Else we just expand its vocabulary\n",
    "                else:\n",
    "                    # Now we expand our vocabulary\n",
    "                    w2v_model.build_vocab (one_mag, update = True)\n",
    "                    \n",
    "    print ('Finished building')\n",
    "    \n",
    "    return w2v_model\n",
    "\n",
    "\n",
    "def train_model (w2v_model, epochs, end = 25):\n",
    "    \n",
    "    print ('Starting model training')\n",
    "    \n",
    "    # There are 1364 MAGs enclosed in FASTA files\n",
    "    # I have to traverse every FASTA file, and in each file every sequence\n",
    "    \n",
    "    fasta_files = [i for i in os.listdir(path_all_fasta) if (i[-2:] == 'fa')]\n",
    "    \n",
    "    # This was done so that I could work with first 100 FASTA files only. Otherwise, I should just remove: i, and enumerate\n",
    "    for i, fasta_file_name in enumerate(fasta_files):\n",
    "        \n",
    "        if i == end:\n",
    "            break\n",
    "        \n",
    "        else:\n",
    "            with open(os.path.join(path_all_fasta, fasta_file_name), 'r') as input_file:\n",
    "                \n",
    "                one_mag = []\n",
    "                for fasta_string in SeqIO.parse(input_file, \"fasta\"):\n",
    "                    \n",
    "                    # Get kmers of a genome and create a sentence (list of words)\n",
    "                    temp_kmers = split_genome (str(fasta_string.seq))\n",
    "                    \n",
    "                    # Create a document (list of sentences)\n",
    "                    one_mag.append(temp_kmers)\n",
    "                    \n",
    "                \n",
    "                w2v_model.train (one_mag, total_examples = w2v_model.corpus_count, epochs = epochs)\n",
    "                    \n",
    "    print ('Model training finished')\n",
    "    \n",
    "    return w2v_model\n",
    "\n",
    "\n",
    "def visualize_with_pca (data, labels, centers):\n",
    "    \n",
    "    pca_model = PCA (n_components = 2, random_state = SEED)\n",
    "    data_transformed = pca_model.fit_transform (data)\n",
    "    \n",
    "    data_transformed = pd.DataFrame (data_transformed)\n",
    "    data_transformed.columns = ['PC_1', 'PC_2']\n",
    "    data_transformed['Labels'] = labels\n",
    "    \n",
    "    chart_data = alt.Chart(data_transformed).mark_circle(opacity = 1).encode(\n",
    "        alt.X ('PC_1:Q'),\n",
    "        alt.Y ('PC_2:Q'),\n",
    "        alt.Color ('Labels:N', legend = alt.Legend())\n",
    "    )\n",
    "    \n",
    "    # This means we are visualising centroids from k_means (there are less centroids that data points)\n",
    "    if labels.shape[0] != centers.shape[0]:\n",
    "        \n",
    "        centers_transformed = pca_model.fit_transform (centers)\n",
    "        centers_transformed = pd.DataFrame (centers_transformed)\n",
    "        centers_transformed.columns = ['PC_1', 'PC_2']\n",
    "        \n",
    "        chart_centers = alt.Chart(centers_transformed).mark_point(shape = 'diamond', color = 'black', size = 50, opacity = 0.7).encode(\n",
    "            alt.X ('PC_1:Q'),\n",
    "            alt.Y ('PC_2:Q'),\n",
    "        )\n",
    "        \n",
    "        return chart_data + chart_centers\n",
    "    \n",
    "    # For DBSCAN there are no centroids\n",
    "    else:\n",
    "        return chart_data\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "knowing-excuse",
   "metadata": {},
   "source": [
    "## MAG examination"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "exterior-invasion",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing MAGs\n",
      "Building w2v model\n",
      "Finished building\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# FOR CLUSTERING I SHOULD CREATE A DATAFRAME WITH MAGs INDEXES AND THEIR VECTOR REPRESENTATIONS\n",
    "\n",
    "final_model = import_mags_and_build_model (end = END, path_all_fasta = path_all_fasta)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "excess-tuesday",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('CCGGC', 0.5897153615951538),\n",
       " ('TCGGC', 0.560882568359375),\n",
       " ('ACGGA', 0.5486097931861877),\n",
       " ('GCGGC', 0.5190006494522095),\n",
       " ('ACGGG', 0.5120842456817627),\n",
       " ('ACGGT', 0.48747333884239197),\n",
       " ('ACGAA', 0.31071770191192627),\n",
       " ('ACGAC', 0.3037322759628296),\n",
       " ('ACGTC', 0.29194021224975586),\n",
       " ('ACGTT', 0.28757983446121216)]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_model.wv.most_similar('ACGGC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "molecular-strand",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting model training\n",
      "Model training finished\n"
     ]
    }
   ],
   "source": [
    "# Train model. It tooks ~10 minutes for END = 25 amount of MAGs\n",
    "EPOCHS = 10\n",
    "\n",
    "final_model = train_model (final_model, epochs = EPOCHS, end = END)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "serial-external",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_model.wv.save_word2vec_format(os.path.join (path_model_save_root, 'model_25.bin'), binary=True) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "prompt-discipline",
   "metadata": {},
   "source": [
    "Now I should vectorize documents with this model. For further use, I could save this model's weights, and use it to vectorize all mags. That would take a lot, but every MAG will have its vector representation\n",
    "> This could be done by importing one MAG at a time, then tokenizing it (like before), then getting vector representations of that MAG's sentences (genomes) and then finding the vector representation of the whole MAG (document). If I do that for one MAG at a time, There is no need to worry about memory\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "complimentary-prescription",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Vectorizing MAGs\n",
      "Finished vectorizing\n"
     ]
    }
   ],
   "source": [
    "list_of_mag_vectors = vectorize_mags (final_model, end = END)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "typical-calibration",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[array([-0.3412855 , -1.3074803 ,  1.6673468 , -0.18193825,  0.6783584 ,\n",
       "         0.5612512 ,  0.24554321,  1.012197  ,  0.79908276, -0.33281672,\n",
       "        -0.15648904,  1.0270334 , -1.7591202 ,  1.6900983 , -1.4474658 ,\n",
       "         0.14639385,  0.2635378 , -1.107245  , -0.60734785, -0.828628  ,\n",
       "        -1.6702052 , -2.5707538 , -0.855652  , -0.9473689 ,  1.5789722 ,\n",
       "        -0.12598857, -0.56086534,  0.6502867 , -0.45995945, -0.21470664,\n",
       "        -0.59516215,  0.10712713, -0.92095405, -0.44383943,  0.4565169 ,\n",
       "        -1.2524884 ,  0.0095067 ,  0.3513306 ,  1.5343064 , -0.72287744,\n",
       "        -0.5809037 ,  1.7036309 ,  2.6705778 , -0.5011542 ,  0.997374  ,\n",
       "         0.0991397 , -1.9106972 , -0.14362767, -0.38712674,  1.3254927 ,\n",
       "         1.7431389 ,  2.058191  , -0.16283734, -0.07293132, -0.7716484 ,\n",
       "        -1.209459  , -1.3840557 , -1.1185156 ,  0.85279655, -0.21090718,\n",
       "        -0.94948244,  1.3082253 , -0.264265  , -0.6213532 ,  1.2397407 ,\n",
       "         0.424508  ,  0.5835601 ,  0.49468315,  1.5864079 , -0.87738883,\n",
       "         0.6018783 , -1.4273747 , -0.54887307, -2.1267896 ,  1.8002617 ,\n",
       "         0.04051406,  0.1719724 ,  2.1054747 ,  0.6386669 , -0.20759436,\n",
       "         0.4925736 ,  1.9699081 ,  2.108959  ,  1.5508952 , -1.839611  ,\n",
       "         0.20987386, -3.0212228 , -0.26749852,  0.76927215,  1.379773  ,\n",
       "         0.4117722 ,  0.3424971 ,  0.6453142 , -0.0095837 ,  1.2614189 ,\n",
       "         0.9477767 ,  1.0051314 , -1.6120713 ,  0.6649687 ,  0.6371815 ],\n",
       "       dtype=float32),\n",
       " array([-0.35324103,  1.3564781 , -1.5199772 ,  0.48981887,  0.01135122,\n",
       "        -1.7531407 ,  0.638735  ,  0.2601508 , -0.8481854 ,  0.12541911,\n",
       "         0.04091527, -0.44542992,  0.9070896 , -0.84629357,  0.77752644,\n",
       "        -1.533778  ,  0.09223127, -1.0737209 ,  2.0153806 ,  0.98181313,\n",
       "         1.2876089 ,  2.6906297 ,  0.34304506, -0.17475137, -0.578983  ,\n",
       "         0.27712563,  0.9200995 ,  0.10998598,  1.195513  ,  0.57291394,\n",
       "        -0.9929817 , -0.46158782,  0.72046643, -0.14549033, -0.8490781 ,\n",
       "         1.5734891 ,  1.8009242 ,  1.1861869 , -0.40417758,  0.21729736,\n",
       "         0.76139474, -2.6367433 , -0.8887636 , -0.48310426, -1.1935097 ,\n",
       "         0.33771887,  2.130353  , -0.54363614, -1.0204004 , -0.477217  ,\n",
       "        -1.3859844 , -1.536534  ,  0.77992266, -0.30815968,  0.78744435,\n",
       "         1.2376032 ,  1.1307034 ,  1.0900105 , -0.07736622,  0.29449335,\n",
       "         1.1882366 , -0.47047514, -0.32448965,  2.0800388 , -1.1721865 ,\n",
       "         1.282906  , -1.2199047 , -0.68259203, -1.4754184 ,  1.7439377 ,\n",
       "         0.5962887 ,  1.4789145 ,  1.5864042 ,  0.64522135, -1.5487069 ,\n",
       "         0.598926  , -1.468332  , -0.4863637 ,  0.22161585,  0.08987768,\n",
       "         0.65630114, -1.44923   , -1.3009859 ,  0.41596392, -0.14485364,\n",
       "         0.9553539 ,  1.7430012 ,  0.407132  ,  1.9415417 , -1.9079943 ,\n",
       "         0.08612078,  0.10741968,  1.613271  ,  1.6481098 ,  0.01365732,\n",
       "        -1.203047  , -0.6103819 ,  0.529846  , -0.8373521 , -2.1927984 ],\n",
       "       dtype=float32)]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_of_mag_vectors[0:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "enhanced-thickness",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>90</th>\n",
       "      <th>91</th>\n",
       "      <th>92</th>\n",
       "      <th>93</th>\n",
       "      <th>94</th>\n",
       "      <th>95</th>\n",
       "      <th>96</th>\n",
       "      <th>97</th>\n",
       "      <th>98</th>\n",
       "      <th>99</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-0.341285</td>\n",
       "      <td>-1.307480</td>\n",
       "      <td>1.667347</td>\n",
       "      <td>-0.181938</td>\n",
       "      <td>0.678358</td>\n",
       "      <td>0.561251</td>\n",
       "      <td>0.245543</td>\n",
       "      <td>1.012197</td>\n",
       "      <td>0.799083</td>\n",
       "      <td>-0.332817</td>\n",
       "      <td>...</td>\n",
       "      <td>0.411772</td>\n",
       "      <td>0.342497</td>\n",
       "      <td>0.645314</td>\n",
       "      <td>-0.009584</td>\n",
       "      <td>1.261419</td>\n",
       "      <td>0.947777</td>\n",
       "      <td>1.005131</td>\n",
       "      <td>-1.612071</td>\n",
       "      <td>0.664969</td>\n",
       "      <td>0.637182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-0.353241</td>\n",
       "      <td>1.356478</td>\n",
       "      <td>-1.519977</td>\n",
       "      <td>0.489819</td>\n",
       "      <td>0.011351</td>\n",
       "      <td>-1.753141</td>\n",
       "      <td>0.638735</td>\n",
       "      <td>0.260151</td>\n",
       "      <td>-0.848185</td>\n",
       "      <td>0.125419</td>\n",
       "      <td>...</td>\n",
       "      <td>0.086121</td>\n",
       "      <td>0.107420</td>\n",
       "      <td>1.613271</td>\n",
       "      <td>1.648110</td>\n",
       "      <td>0.013657</td>\n",
       "      <td>-1.203047</td>\n",
       "      <td>-0.610382</td>\n",
       "      <td>0.529846</td>\n",
       "      <td>-0.837352</td>\n",
       "      <td>-2.192798</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-1.027619</td>\n",
       "      <td>-1.867537</td>\n",
       "      <td>1.986852</td>\n",
       "      <td>-0.700268</td>\n",
       "      <td>0.068323</td>\n",
       "      <td>1.246489</td>\n",
       "      <td>0.360177</td>\n",
       "      <td>1.379367</td>\n",
       "      <td>0.769427</td>\n",
       "      <td>-0.346021</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.694683</td>\n",
       "      <td>0.318921</td>\n",
       "      <td>0.516075</td>\n",
       "      <td>-0.900386</td>\n",
       "      <td>1.879338</td>\n",
       "      <td>1.581902</td>\n",
       "      <td>0.321573</td>\n",
       "      <td>-1.432031</td>\n",
       "      <td>0.769141</td>\n",
       "      <td>0.652904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.075040</td>\n",
       "      <td>1.481463</td>\n",
       "      <td>-2.178516</td>\n",
       "      <td>0.583901</td>\n",
       "      <td>-0.200483</td>\n",
       "      <td>-1.770640</td>\n",
       "      <td>0.561849</td>\n",
       "      <td>-0.078156</td>\n",
       "      <td>-0.928566</td>\n",
       "      <td>0.127184</td>\n",
       "      <td>...</td>\n",
       "      <td>0.284476</td>\n",
       "      <td>0.141210</td>\n",
       "      <td>1.831504</td>\n",
       "      <td>1.398362</td>\n",
       "      <td>-0.607934</td>\n",
       "      <td>-0.966220</td>\n",
       "      <td>-0.872207</td>\n",
       "      <td>0.923392</td>\n",
       "      <td>-0.848844</td>\n",
       "      <td>-2.399399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.075354</td>\n",
       "      <td>1.427001</td>\n",
       "      <td>-2.124519</td>\n",
       "      <td>0.521798</td>\n",
       "      <td>-0.294279</td>\n",
       "      <td>-1.672292</td>\n",
       "      <td>0.562149</td>\n",
       "      <td>0.005005</td>\n",
       "      <td>-0.853643</td>\n",
       "      <td>0.065581</td>\n",
       "      <td>...</td>\n",
       "      <td>0.274580</td>\n",
       "      <td>0.118297</td>\n",
       "      <td>1.762440</td>\n",
       "      <td>1.379492</td>\n",
       "      <td>-0.546828</td>\n",
       "      <td>-0.996493</td>\n",
       "      <td>-0.796771</td>\n",
       "      <td>0.906181</td>\n",
       "      <td>-0.835940</td>\n",
       "      <td>-2.342209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.129941</td>\n",
       "      <td>1.277619</td>\n",
       "      <td>-1.844996</td>\n",
       "      <td>0.219384</td>\n",
       "      <td>-0.353005</td>\n",
       "      <td>-1.293824</td>\n",
       "      <td>0.321701</td>\n",
       "      <td>0.232485</td>\n",
       "      <td>-0.444988</td>\n",
       "      <td>-0.437443</td>\n",
       "      <td>...</td>\n",
       "      <td>0.159732</td>\n",
       "      <td>-0.018819</td>\n",
       "      <td>1.726185</td>\n",
       "      <td>0.603282</td>\n",
       "      <td>0.019392</td>\n",
       "      <td>-0.371690</td>\n",
       "      <td>-0.961833</td>\n",
       "      <td>0.956784</td>\n",
       "      <td>-0.973640</td>\n",
       "      <td>-2.271333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>-0.989811</td>\n",
       "      <td>-2.985559</td>\n",
       "      <td>2.709233</td>\n",
       "      <td>-0.640486</td>\n",
       "      <td>0.232940</td>\n",
       "      <td>1.623222</td>\n",
       "      <td>0.477722</td>\n",
       "      <td>1.724624</td>\n",
       "      <td>0.602666</td>\n",
       "      <td>-0.209730</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.317823</td>\n",
       "      <td>0.606889</td>\n",
       "      <td>0.257215</td>\n",
       "      <td>-0.741542</td>\n",
       "      <td>2.227737</td>\n",
       "      <td>1.319617</td>\n",
       "      <td>1.197607</td>\n",
       "      <td>-2.712788</td>\n",
       "      <td>1.819405</td>\n",
       "      <td>2.038449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>-0.951616</td>\n",
       "      <td>-2.665345</td>\n",
       "      <td>2.896326</td>\n",
       "      <td>-0.725836</td>\n",
       "      <td>0.373385</td>\n",
       "      <td>1.783689</td>\n",
       "      <td>0.521206</td>\n",
       "      <td>1.706788</td>\n",
       "      <td>0.528568</td>\n",
       "      <td>-0.127032</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.642154</td>\n",
       "      <td>0.893542</td>\n",
       "      <td>0.116415</td>\n",
       "      <td>-0.858817</td>\n",
       "      <td>1.663370</td>\n",
       "      <td>1.752637</td>\n",
       "      <td>1.220060</td>\n",
       "      <td>-2.411797</td>\n",
       "      <td>1.615608</td>\n",
       "      <td>1.523981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.063258</td>\n",
       "      <td>-0.265686</td>\n",
       "      <td>0.631984</td>\n",
       "      <td>-0.196667</td>\n",
       "      <td>0.216028</td>\n",
       "      <td>0.246256</td>\n",
       "      <td>0.385285</td>\n",
       "      <td>0.877592</td>\n",
       "      <td>0.378182</td>\n",
       "      <td>-0.462380</td>\n",
       "      <td>...</td>\n",
       "      <td>0.465767</td>\n",
       "      <td>0.268095</td>\n",
       "      <td>0.891215</td>\n",
       "      <td>-0.241772</td>\n",
       "      <td>1.045041</td>\n",
       "      <td>0.846471</td>\n",
       "      <td>0.344685</td>\n",
       "      <td>-0.573762</td>\n",
       "      <td>-0.044410</td>\n",
       "      <td>-0.331948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>-0.025258</td>\n",
       "      <td>-0.052589</td>\n",
       "      <td>-0.332469</td>\n",
       "      <td>-0.216357</td>\n",
       "      <td>-0.029103</td>\n",
       "      <td>-0.398063</td>\n",
       "      <td>0.019306</td>\n",
       "      <td>0.645865</td>\n",
       "      <td>0.542800</td>\n",
       "      <td>-0.684887</td>\n",
       "      <td>...</td>\n",
       "      <td>0.247554</td>\n",
       "      <td>0.156420</td>\n",
       "      <td>1.397167</td>\n",
       "      <td>-0.127547</td>\n",
       "      <td>1.007807</td>\n",
       "      <td>0.455250</td>\n",
       "      <td>-0.290012</td>\n",
       "      <td>-0.409483</td>\n",
       "      <td>-0.254392</td>\n",
       "      <td>-0.798456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>-0.774891</td>\n",
       "      <td>-0.759372</td>\n",
       "      <td>0.990898</td>\n",
       "      <td>-0.550337</td>\n",
       "      <td>0.525517</td>\n",
       "      <td>0.238583</td>\n",
       "      <td>0.201613</td>\n",
       "      <td>0.957158</td>\n",
       "      <td>1.068931</td>\n",
       "      <td>-0.877320</td>\n",
       "      <td>...</td>\n",
       "      <td>0.296142</td>\n",
       "      <td>-0.024803</td>\n",
       "      <td>0.626915</td>\n",
       "      <td>-0.168473</td>\n",
       "      <td>1.561563</td>\n",
       "      <td>0.982090</td>\n",
       "      <td>0.094608</td>\n",
       "      <td>-0.521890</td>\n",
       "      <td>-0.185521</td>\n",
       "      <td>0.074665</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>-0.162748</td>\n",
       "      <td>-0.904852</td>\n",
       "      <td>1.138327</td>\n",
       "      <td>-0.428986</td>\n",
       "      <td>-0.075081</td>\n",
       "      <td>0.706694</td>\n",
       "      <td>0.313898</td>\n",
       "      <td>1.044261</td>\n",
       "      <td>0.480721</td>\n",
       "      <td>-0.311993</td>\n",
       "      <td>...</td>\n",
       "      <td>0.155758</td>\n",
       "      <td>0.605894</td>\n",
       "      <td>0.799425</td>\n",
       "      <td>-0.595461</td>\n",
       "      <td>1.138482</td>\n",
       "      <td>0.979226</td>\n",
       "      <td>0.447821</td>\n",
       "      <td>-1.221070</td>\n",
       "      <td>0.699337</td>\n",
       "      <td>0.300184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>-0.056239</td>\n",
       "      <td>0.051084</td>\n",
       "      <td>-0.491837</td>\n",
       "      <td>-0.239224</td>\n",
       "      <td>-0.017112</td>\n",
       "      <td>-0.360456</td>\n",
       "      <td>0.228465</td>\n",
       "      <td>0.655526</td>\n",
       "      <td>0.524382</td>\n",
       "      <td>-0.745152</td>\n",
       "      <td>...</td>\n",
       "      <td>0.324269</td>\n",
       "      <td>0.240862</td>\n",
       "      <td>1.238024</td>\n",
       "      <td>-0.215436</td>\n",
       "      <td>0.819516</td>\n",
       "      <td>0.475223</td>\n",
       "      <td>-0.292477</td>\n",
       "      <td>-0.086527</td>\n",
       "      <td>-0.501598</td>\n",
       "      <td>-0.760479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>0.253001</td>\n",
       "      <td>1.930428</td>\n",
       "      <td>-2.134980</td>\n",
       "      <td>0.379541</td>\n",
       "      <td>-0.377623</td>\n",
       "      <td>-1.514137</td>\n",
       "      <td>0.723177</td>\n",
       "      <td>0.247370</td>\n",
       "      <td>-0.625146</td>\n",
       "      <td>-0.328814</td>\n",
       "      <td>...</td>\n",
       "      <td>0.268433</td>\n",
       "      <td>0.023448</td>\n",
       "      <td>1.530563</td>\n",
       "      <td>1.068806</td>\n",
       "      <td>-0.002943</td>\n",
       "      <td>-1.203657</td>\n",
       "      <td>-0.693088</td>\n",
       "      <td>1.118636</td>\n",
       "      <td>-1.292646</td>\n",
       "      <td>-1.889813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>0.306026</td>\n",
       "      <td>1.964842</td>\n",
       "      <td>-2.190711</td>\n",
       "      <td>0.419888</td>\n",
       "      <td>-0.347752</td>\n",
       "      <td>-1.582111</td>\n",
       "      <td>0.711217</td>\n",
       "      <td>0.191082</td>\n",
       "      <td>-0.666909</td>\n",
       "      <td>-0.279569</td>\n",
       "      <td>...</td>\n",
       "      <td>0.265846</td>\n",
       "      <td>0.062000</td>\n",
       "      <td>1.592789</td>\n",
       "      <td>1.070795</td>\n",
       "      <td>-0.061789</td>\n",
       "      <td>-1.180153</td>\n",
       "      <td>-0.740937</td>\n",
       "      <td>1.137477</td>\n",
       "      <td>-1.303873</td>\n",
       "      <td>-1.932118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>-0.011957</td>\n",
       "      <td>1.177972</td>\n",
       "      <td>-1.469779</td>\n",
       "      <td>0.358964</td>\n",
       "      <td>0.123565</td>\n",
       "      <td>-1.275866</td>\n",
       "      <td>0.530374</td>\n",
       "      <td>0.205258</td>\n",
       "      <td>0.090318</td>\n",
       "      <td>-0.546076</td>\n",
       "      <td>...</td>\n",
       "      <td>0.452588</td>\n",
       "      <td>-0.059148</td>\n",
       "      <td>1.406024</td>\n",
       "      <td>0.917376</td>\n",
       "      <td>0.503887</td>\n",
       "      <td>-0.659837</td>\n",
       "      <td>-0.509938</td>\n",
       "      <td>0.433887</td>\n",
       "      <td>-1.000301</td>\n",
       "      <td>-1.117185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>-0.340222</td>\n",
       "      <td>1.357012</td>\n",
       "      <td>-1.523113</td>\n",
       "      <td>0.502377</td>\n",
       "      <td>0.060938</td>\n",
       "      <td>-1.753876</td>\n",
       "      <td>0.659446</td>\n",
       "      <td>0.230910</td>\n",
       "      <td>-0.881364</td>\n",
       "      <td>0.177388</td>\n",
       "      <td>...</td>\n",
       "      <td>0.103967</td>\n",
       "      <td>0.121654</td>\n",
       "      <td>1.642471</td>\n",
       "      <td>1.606108</td>\n",
       "      <td>-0.014413</td>\n",
       "      <td>-1.191484</td>\n",
       "      <td>-0.614346</td>\n",
       "      <td>0.516197</td>\n",
       "      <td>-0.819865</td>\n",
       "      <td>-2.183703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>-0.019467</td>\n",
       "      <td>1.210692</td>\n",
       "      <td>-1.568132</td>\n",
       "      <td>0.393560</td>\n",
       "      <td>-0.014462</td>\n",
       "      <td>-1.470363</td>\n",
       "      <td>0.584515</td>\n",
       "      <td>0.087994</td>\n",
       "      <td>-0.297698</td>\n",
       "      <td>-0.283613</td>\n",
       "      <td>...</td>\n",
       "      <td>0.380700</td>\n",
       "      <td>-0.011865</td>\n",
       "      <td>1.472005</td>\n",
       "      <td>1.181135</td>\n",
       "      <td>0.225303</td>\n",
       "      <td>-0.900997</td>\n",
       "      <td>-0.526916</td>\n",
       "      <td>0.643913</td>\n",
       "      <td>-0.705494</td>\n",
       "      <td>-1.352570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>0.111306</td>\n",
       "      <td>1.124060</td>\n",
       "      <td>-1.694703</td>\n",
       "      <td>0.164811</td>\n",
       "      <td>-0.346663</td>\n",
       "      <td>-1.207019</td>\n",
       "      <td>0.250062</td>\n",
       "      <td>0.269831</td>\n",
       "      <td>-0.271186</td>\n",
       "      <td>-0.515764</td>\n",
       "      <td>...</td>\n",
       "      <td>0.168750</td>\n",
       "      <td>-0.106328</td>\n",
       "      <td>1.701809</td>\n",
       "      <td>0.609775</td>\n",
       "      <td>0.122307</td>\n",
       "      <td>-0.262924</td>\n",
       "      <td>-0.927522</td>\n",
       "      <td>0.902912</td>\n",
       "      <td>-0.926099</td>\n",
       "      <td>-2.200033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>-0.088883</td>\n",
       "      <td>-1.774067</td>\n",
       "      <td>1.874891</td>\n",
       "      <td>-0.489322</td>\n",
       "      <td>0.226948</td>\n",
       "      <td>1.142187</td>\n",
       "      <td>0.214877</td>\n",
       "      <td>1.060860</td>\n",
       "      <td>0.478415</td>\n",
       "      <td>-0.106989</td>\n",
       "      <td>...</td>\n",
       "      <td>0.272983</td>\n",
       "      <td>0.929613</td>\n",
       "      <td>0.511087</td>\n",
       "      <td>-0.438031</td>\n",
       "      <td>1.021506</td>\n",
       "      <td>1.205718</td>\n",
       "      <td>1.321888</td>\n",
       "      <td>-2.213758</td>\n",
       "      <td>1.547996</td>\n",
       "      <td>1.263341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>-0.340377</td>\n",
       "      <td>-0.471164</td>\n",
       "      <td>0.299343</td>\n",
       "      <td>-0.134967</td>\n",
       "      <td>0.183982</td>\n",
       "      <td>-0.128862</td>\n",
       "      <td>0.362449</td>\n",
       "      <td>0.787587</td>\n",
       "      <td>0.192776</td>\n",
       "      <td>-0.464600</td>\n",
       "      <td>...</td>\n",
       "      <td>0.337627</td>\n",
       "      <td>0.433550</td>\n",
       "      <td>1.064676</td>\n",
       "      <td>0.044629</td>\n",
       "      <td>0.733577</td>\n",
       "      <td>0.898704</td>\n",
       "      <td>-0.061830</td>\n",
       "      <td>-0.428644</td>\n",
       "      <td>0.276558</td>\n",
       "      <td>-0.203909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>0.151566</td>\n",
       "      <td>1.676675</td>\n",
       "      <td>-2.312684</td>\n",
       "      <td>0.419239</td>\n",
       "      <td>-0.235821</td>\n",
       "      <td>-1.675354</td>\n",
       "      <td>0.641119</td>\n",
       "      <td>-0.002728</td>\n",
       "      <td>-0.743059</td>\n",
       "      <td>-0.173250</td>\n",
       "      <td>...</td>\n",
       "      <td>0.490933</td>\n",
       "      <td>0.173515</td>\n",
       "      <td>1.670215</td>\n",
       "      <td>1.196685</td>\n",
       "      <td>-0.310122</td>\n",
       "      <td>-1.046623</td>\n",
       "      <td>-0.680096</td>\n",
       "      <td>1.103604</td>\n",
       "      <td>-0.866108</td>\n",
       "      <td>-1.836900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>-0.594441</td>\n",
       "      <td>-2.292743</td>\n",
       "      <td>2.226959</td>\n",
       "      <td>-0.719194</td>\n",
       "      <td>-0.014157</td>\n",
       "      <td>1.485396</td>\n",
       "      <td>0.350506</td>\n",
       "      <td>1.388205</td>\n",
       "      <td>0.783406</td>\n",
       "      <td>-0.191380</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.903835</td>\n",
       "      <td>0.526295</td>\n",
       "      <td>0.339855</td>\n",
       "      <td>-0.750133</td>\n",
       "      <td>2.159755</td>\n",
       "      <td>1.009761</td>\n",
       "      <td>1.114186</td>\n",
       "      <td>-2.431833</td>\n",
       "      <td>1.412724</td>\n",
       "      <td>1.952451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>0.392315</td>\n",
       "      <td>1.645563</td>\n",
       "      <td>-1.082820</td>\n",
       "      <td>0.101507</td>\n",
       "      <td>-0.302728</td>\n",
       "      <td>-0.203934</td>\n",
       "      <td>1.014961</td>\n",
       "      <td>0.716560</td>\n",
       "      <td>0.038579</td>\n",
       "      <td>-0.645978</td>\n",
       "      <td>...</td>\n",
       "      <td>0.143099</td>\n",
       "      <td>-0.283704</td>\n",
       "      <td>1.080616</td>\n",
       "      <td>0.039264</td>\n",
       "      <td>0.411999</td>\n",
       "      <td>-0.031855</td>\n",
       "      <td>-0.331281</td>\n",
       "      <td>1.311558</td>\n",
       "      <td>-1.536049</td>\n",
       "      <td>-1.380166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>0.789021</td>\n",
       "      <td>1.580475</td>\n",
       "      <td>-2.906554</td>\n",
       "      <td>0.379436</td>\n",
       "      <td>-0.971741</td>\n",
       "      <td>-1.724445</td>\n",
       "      <td>0.362805</td>\n",
       "      <td>-0.111940</td>\n",
       "      <td>-0.818864</td>\n",
       "      <td>-0.132889</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.013987</td>\n",
       "      <td>0.033021</td>\n",
       "      <td>2.188884</td>\n",
       "      <td>0.832239</td>\n",
       "      <td>-0.568331</td>\n",
       "      <td>-0.994217</td>\n",
       "      <td>-1.207963</td>\n",
       "      <td>1.280391</td>\n",
       "      <td>-0.705766</td>\n",
       "      <td>-2.485486</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25 rows × 100 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0         1         2         3         4         5         6   \\\n",
       "0  -0.341285 -1.307480  1.667347 -0.181938  0.678358  0.561251  0.245543   \n",
       "1  -0.353241  1.356478 -1.519977  0.489819  0.011351 -1.753141  0.638735   \n",
       "2  -1.027619 -1.867537  1.986852 -0.700268  0.068323  1.246489  0.360177   \n",
       "3   0.075040  1.481463 -2.178516  0.583901 -0.200483 -1.770640  0.561849   \n",
       "4   0.075354  1.427001 -2.124519  0.521798 -0.294279 -1.672292  0.562149   \n",
       "5   0.129941  1.277619 -1.844996  0.219384 -0.353005 -1.293824  0.321701   \n",
       "6  -0.989811 -2.985559  2.709233 -0.640486  0.232940  1.623222  0.477722   \n",
       "7  -0.951616 -2.665345  2.896326 -0.725836  0.373385  1.783689  0.521206   \n",
       "8   0.063258 -0.265686  0.631984 -0.196667  0.216028  0.246256  0.385285   \n",
       "9  -0.025258 -0.052589 -0.332469 -0.216357 -0.029103 -0.398063  0.019306   \n",
       "10 -0.774891 -0.759372  0.990898 -0.550337  0.525517  0.238583  0.201613   \n",
       "11 -0.162748 -0.904852  1.138327 -0.428986 -0.075081  0.706694  0.313898   \n",
       "12 -0.056239  0.051084 -0.491837 -0.239224 -0.017112 -0.360456  0.228465   \n",
       "13  0.253001  1.930428 -2.134980  0.379541 -0.377623 -1.514137  0.723177   \n",
       "14  0.306026  1.964842 -2.190711  0.419888 -0.347752 -1.582111  0.711217   \n",
       "15 -0.011957  1.177972 -1.469779  0.358964  0.123565 -1.275866  0.530374   \n",
       "16 -0.340222  1.357012 -1.523113  0.502377  0.060938 -1.753876  0.659446   \n",
       "17 -0.019467  1.210692 -1.568132  0.393560 -0.014462 -1.470363  0.584515   \n",
       "18  0.111306  1.124060 -1.694703  0.164811 -0.346663 -1.207019  0.250062   \n",
       "19 -0.088883 -1.774067  1.874891 -0.489322  0.226948  1.142187  0.214877   \n",
       "20 -0.340377 -0.471164  0.299343 -0.134967  0.183982 -0.128862  0.362449   \n",
       "21  0.151566  1.676675 -2.312684  0.419239 -0.235821 -1.675354  0.641119   \n",
       "22 -0.594441 -2.292743  2.226959 -0.719194 -0.014157  1.485396  0.350506   \n",
       "23  0.392315  1.645563 -1.082820  0.101507 -0.302728 -0.203934  1.014961   \n",
       "24  0.789021  1.580475 -2.906554  0.379436 -0.971741 -1.724445  0.362805   \n",
       "\n",
       "          7         8         9   ...        90        91        92        93  \\\n",
       "0   1.012197  0.799083 -0.332817  ...  0.411772  0.342497  0.645314 -0.009584   \n",
       "1   0.260151 -0.848185  0.125419  ...  0.086121  0.107420  1.613271  1.648110   \n",
       "2   1.379367  0.769427 -0.346021  ... -0.694683  0.318921  0.516075 -0.900386   \n",
       "3  -0.078156 -0.928566  0.127184  ...  0.284476  0.141210  1.831504  1.398362   \n",
       "4   0.005005 -0.853643  0.065581  ...  0.274580  0.118297  1.762440  1.379492   \n",
       "5   0.232485 -0.444988 -0.437443  ...  0.159732 -0.018819  1.726185  0.603282   \n",
       "6   1.724624  0.602666 -0.209730  ... -1.317823  0.606889  0.257215 -0.741542   \n",
       "7   1.706788  0.528568 -0.127032  ... -0.642154  0.893542  0.116415 -0.858817   \n",
       "8   0.877592  0.378182 -0.462380  ...  0.465767  0.268095  0.891215 -0.241772   \n",
       "9   0.645865  0.542800 -0.684887  ...  0.247554  0.156420  1.397167 -0.127547   \n",
       "10  0.957158  1.068931 -0.877320  ...  0.296142 -0.024803  0.626915 -0.168473   \n",
       "11  1.044261  0.480721 -0.311993  ...  0.155758  0.605894  0.799425 -0.595461   \n",
       "12  0.655526  0.524382 -0.745152  ...  0.324269  0.240862  1.238024 -0.215436   \n",
       "13  0.247370 -0.625146 -0.328814  ...  0.268433  0.023448  1.530563  1.068806   \n",
       "14  0.191082 -0.666909 -0.279569  ...  0.265846  0.062000  1.592789  1.070795   \n",
       "15  0.205258  0.090318 -0.546076  ...  0.452588 -0.059148  1.406024  0.917376   \n",
       "16  0.230910 -0.881364  0.177388  ...  0.103967  0.121654  1.642471  1.606108   \n",
       "17  0.087994 -0.297698 -0.283613  ...  0.380700 -0.011865  1.472005  1.181135   \n",
       "18  0.269831 -0.271186 -0.515764  ...  0.168750 -0.106328  1.701809  0.609775   \n",
       "19  1.060860  0.478415 -0.106989  ...  0.272983  0.929613  0.511087 -0.438031   \n",
       "20  0.787587  0.192776 -0.464600  ...  0.337627  0.433550  1.064676  0.044629   \n",
       "21 -0.002728 -0.743059 -0.173250  ...  0.490933  0.173515  1.670215  1.196685   \n",
       "22  1.388205  0.783406 -0.191380  ... -0.903835  0.526295  0.339855 -0.750133   \n",
       "23  0.716560  0.038579 -0.645978  ...  0.143099 -0.283704  1.080616  0.039264   \n",
       "24 -0.111940 -0.818864 -0.132889  ... -0.013987  0.033021  2.188884  0.832239   \n",
       "\n",
       "          94        95        96        97        98        99  \n",
       "0   1.261419  0.947777  1.005131 -1.612071  0.664969  0.637182  \n",
       "1   0.013657 -1.203047 -0.610382  0.529846 -0.837352 -2.192798  \n",
       "2   1.879338  1.581902  0.321573 -1.432031  0.769141  0.652904  \n",
       "3  -0.607934 -0.966220 -0.872207  0.923392 -0.848844 -2.399399  \n",
       "4  -0.546828 -0.996493 -0.796771  0.906181 -0.835940 -2.342209  \n",
       "5   0.019392 -0.371690 -0.961833  0.956784 -0.973640 -2.271333  \n",
       "6   2.227737  1.319617  1.197607 -2.712788  1.819405  2.038449  \n",
       "7   1.663370  1.752637  1.220060 -2.411797  1.615608  1.523981  \n",
       "8   1.045041  0.846471  0.344685 -0.573762 -0.044410 -0.331948  \n",
       "9   1.007807  0.455250 -0.290012 -0.409483 -0.254392 -0.798456  \n",
       "10  1.561563  0.982090  0.094608 -0.521890 -0.185521  0.074665  \n",
       "11  1.138482  0.979226  0.447821 -1.221070  0.699337  0.300184  \n",
       "12  0.819516  0.475223 -0.292477 -0.086527 -0.501598 -0.760479  \n",
       "13 -0.002943 -1.203657 -0.693088  1.118636 -1.292646 -1.889813  \n",
       "14 -0.061789 -1.180153 -0.740937  1.137477 -1.303873 -1.932118  \n",
       "15  0.503887 -0.659837 -0.509938  0.433887 -1.000301 -1.117185  \n",
       "16 -0.014413 -1.191484 -0.614346  0.516197 -0.819865 -2.183703  \n",
       "17  0.225303 -0.900997 -0.526916  0.643913 -0.705494 -1.352570  \n",
       "18  0.122307 -0.262924 -0.927522  0.902912 -0.926099 -2.200033  \n",
       "19  1.021506  1.205718  1.321888 -2.213758  1.547996  1.263341  \n",
       "20  0.733577  0.898704 -0.061830 -0.428644  0.276558 -0.203909  \n",
       "21 -0.310122 -1.046623 -0.680096  1.103604 -0.866108 -1.836900  \n",
       "22  2.159755  1.009761  1.114186 -2.431833  1.412724  1.952451  \n",
       "23  0.411999 -0.031855 -0.331281  1.311558 -1.536049 -1.380166  \n",
       "24 -0.568331 -0.994217 -1.207963  1.280391 -0.705766 -2.485486  \n",
       "\n",
       "[25 rows x 100 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mags_df = pd.DataFrame (list_of_mag_vectors)\n",
    "mags_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "signal-ethernet",
   "metadata": {},
   "source": [
    "## Data preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "indonesian-sequence",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[-0.43592917, -0.94533422,  1.1326052 , ..., -1.11089311,\n",
       "         0.86863874,  0.97492981],\n",
       "       [-0.46306863,  0.77474429, -0.64966773, ...,  0.54099618,\n",
       "        -0.65185799, -0.98656637],\n",
       "       [-1.99393027, -1.30695476,  1.31126462, ..., -0.97204231,\n",
       "         0.9740716 ,  0.98582712],\n",
       "       ...,\n",
       "       [-1.01060177, -1.58150372,  1.44552636, ..., -1.74310989,\n",
       "         1.62544066,  1.88656041],\n",
       "       [ 1.22936975,  0.96140213, -0.40522044, ...,  1.14386822,\n",
       "        -1.359008  , -0.42331992],\n",
       "       [ 2.12990612,  0.9193756 , -1.4250071 , ...,  1.11983122,\n",
       "        -0.5186798 , -1.18943218]])"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mag_scaler = preprocessing.StandardScaler()\n",
    "scaled_mags_df = mag_scaler.fit_transform(mags_df)\n",
    "\n",
    "scaled_mags_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "baking-training",
   "metadata": {},
   "source": [
    "## Clustering"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "foreign-probability",
   "metadata": {},
   "source": [
    "### 1. K-means"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "floral-alcohol",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>k_range</th>\n",
       "      <th>k_scores</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>-2500.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>-987.169696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>-635.763126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>-519.369088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>-412.674751</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>-337.076059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>-281.966621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>-243.332138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>-208.154294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>-173.791454</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>11</td>\n",
       "      <td>-143.398140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>12</td>\n",
       "      <td>-115.219760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>13</td>\n",
       "      <td>-90.519651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>14</td>\n",
       "      <td>-78.780359</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    k_range     k_scores\n",
       "0         1 -2500.000000\n",
       "1         2  -987.169696\n",
       "2         3  -635.763126\n",
       "3         4  -519.369088\n",
       "4         5  -412.674751\n",
       "5         6  -337.076059\n",
       "6         7  -281.966621\n",
       "7         8  -243.332138\n",
       "8         9  -208.154294\n",
       "9        10  -173.791454\n",
       "10       11  -143.398140\n",
       "11       12  -115.219760\n",
       "12       13   -90.519651\n",
       "13       14   -78.780359"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k_range_end = 15 # Usually it is sqrt(# of mags)\n",
    "\n",
    "k_range = range(1, k_range_end)\n",
    "\n",
    "k_mean_models = [KMeans (n_clusters = i, random_state = SEED) for i in k_range]\n",
    "k_scores = [k_mean_model.fit(scaled_mags_df).score(scaled_mags_df) for k_mean_model in k_mean_models]\n",
    "k_data = pd.DataFrame ({'k_range':k_range, 'k_scores':k_scores})\n",
    "k_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "backed-birmingham",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-0efb3bb0922542698504590355535b49\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-0efb3bb0922542698504590355535b49\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-0efb3bb0922542698504590355535b49\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.8.1?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function loadScript(lib) {\n",
       "      return new Promise(function(resolve, reject) {\n",
       "        var s = document.createElement('script');\n",
       "        s.src = paths[lib];\n",
       "        s.async = true;\n",
       "        s.onload = () => resolve(paths[lib]);\n",
       "        s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "        document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "      });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else if (typeof vegaEmbed === \"function\") {\n",
       "      displayChart(vegaEmbed);\n",
       "    } else {\n",
       "      loadScript(\"vega\")\n",
       "        .then(() => loadScript(\"vega-lite\"))\n",
       "        .then(() => loadScript(\"vega-embed\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}}, \"data\": {\"name\": \"data-5566ce6d37e13c28b3045b495bc11ef5\"}, \"mark\": \"line\", \"encoding\": {\"x\": {\"type\": \"quantitative\", \"field\": \"k_range\"}, \"y\": {\"type\": \"quantitative\", \"field\": \"k_scores\"}}, \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.8.1.json\", \"datasets\": {\"data-5566ce6d37e13c28b3045b495bc11ef5\": [{\"k_range\": 1, \"k_scores\": -2500.0000000000005}, {\"k_range\": 2, \"k_scores\": -987.1696961680369}, {\"k_range\": 3, \"k_scores\": -635.7631258600818}, {\"k_range\": 4, \"k_scores\": -519.369088431661}, {\"k_range\": 5, \"k_scores\": -412.6747511246273}, {\"k_range\": 6, \"k_scores\": -337.0760585505566}, {\"k_range\": 7, \"k_scores\": -281.9666208165285}, {\"k_range\": 8, \"k_scores\": -243.33213778055696}, {\"k_range\": 9, \"k_scores\": -208.15429440129233}, {\"k_range\": 10, \"k_scores\": -173.79145406062116}, {\"k_range\": 11, \"k_scores\": -143.39814006365418}, {\"k_range\": 12, \"k_scores\": -115.21976031446437}, {\"k_range\": 13, \"k_scores\": -90.51965076645182}, {\"k_range\": 14, \"k_scores\": -78.78035932523582}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.Chart(...)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k_num_chart = alt.Chart(data = k_data).mark_line().encode(\n",
    "    alt.X ('k_range:Q'),\n",
    "    alt.Y ('k_scores:Q')\n",
    ")\n",
    "\n",
    "k_num_chart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "happy-minimum",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1, 0, 3, 0, 0, 2, 3, 3, 4, 4, 4, 1, 4, 2, 2, 5, 0, 5, 2, 1, 4, 0,\n",
       "       3, 6, 2], dtype=int32)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# We can see from the chart above that 6 or 7 clusters are optimal for this task (where END = 25 MAGs)\n",
    "num_of_clusters = 7\n",
    "\n",
    "k_means_model = KMeans (n_clusters = num_of_clusters, random_state = SEED)\n",
    "k_means_predicted = k_means_model.fit_predict(scaled_mags_df)\n",
    "k_means_predicted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "identical-smile",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-b1ad98f2f8114e5ca0b4ec89d49586f4\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-b1ad98f2f8114e5ca0b4ec89d49586f4\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-b1ad98f2f8114e5ca0b4ec89d49586f4\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.8.1?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function loadScript(lib) {\n",
       "      return new Promise(function(resolve, reject) {\n",
       "        var s = document.createElement('script');\n",
       "        s.src = paths[lib];\n",
       "        s.async = true;\n",
       "        s.onload = () => resolve(paths[lib]);\n",
       "        s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "        document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "      });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else if (typeof vegaEmbed === \"function\") {\n",
       "      displayChart(vegaEmbed);\n",
       "    } else {\n",
       "      loadScript(\"vega\")\n",
       "        .then(() => loadScript(\"vega-lite\"))\n",
       "        .then(() => loadScript(\"vega-embed\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}}, \"layer\": [{\"data\": {\"name\": \"data-9fe3a9dde2bfb264c6a9c23aeb35dfad\"}, \"mark\": {\"type\": \"circle\", \"opacity\": 1}, \"encoding\": {\"color\": {\"type\": \"nominal\", \"field\": \"Labels\", \"legend\": {}}, \"x\": {\"type\": \"quantitative\", \"field\": \"PC_1\"}, \"y\": {\"type\": \"quantitative\", \"field\": \"PC_2\"}}}, {\"data\": {\"name\": \"data-d07752ad55d31e1c387780ee4f199682\"}, \"mark\": {\"type\": \"point\", \"color\": \"black\", \"opacity\": 0.7, \"shape\": \"diamond\", \"size\": 50}, \"encoding\": {\"x\": {\"type\": \"quantitative\", \"field\": \"PC_1\"}, \"y\": {\"type\": \"quantitative\", \"field\": \"PC_2\"}}}], \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.8.1.json\", \"datasets\": {\"data-9fe3a9dde2bfb264c6a9c23aeb35dfad\": [{\"PC_1\": 7.676253616139165, \"PC_2\": 6.443405176596787, \"Labels\": 1}, {\"PC_1\": -8.50469183576414, \"PC_2\": 1.7209173450327375, \"Labels\": 0}, {\"PC_1\": 10.824248687046426, \"PC_2\": -5.110219437734072, \"Labels\": 3}, {\"PC_1\": -9.80633885188385, \"PC_2\": 1.2041866623374056, \"Labels\": 0}, {\"PC_1\": -9.442347519453536, \"PC_2\": 1.0267784429715243, \"Labels\": 0}, {\"PC_1\": -6.430261739310041, \"PC_2\": -2.175666406065586, \"Labels\": 2}, {\"PC_1\": 16.02361892365011, \"PC_2\": -4.569272221832386, \"Labels\": 3}, {\"PC_1\": 14.307926781993281, \"PC_2\": -0.010167945813132004, \"Labels\": 3}, {\"PC_1\": 3.476445831112152, \"PC_2\": 2.6667062192963322, \"Labels\": 4}, {\"PC_1\": 1.9678988196289349, \"PC_2\": 0.4667960240696394, \"Labels\": 4}, {\"PC_1\": 5.975459567236973, \"PC_2\": -0.275921908725272, \"Labels\": 4}, {\"PC_1\": 6.916704987432156, \"PC_2\": 1.4237144507733028, \"Labels\": 1}, {\"PC_1\": 1.738644580544049, \"PC_2\": -0.4155862950046197, \"Labels\": 4}, {\"PC_1\": -8.02854221988149, \"PC_2\": -1.2848578715844472, \"Labels\": 2}, {\"PC_1\": -8.268914843331425, \"PC_2\": -1.200655944240077, \"Labels\": 2}, {\"PC_1\": -4.359284262212505, \"PC_2\": 1.7822336426377048, \"Labels\": 5}, {\"PC_1\": -8.53094914613286, \"PC_2\": 1.8170759916232007, \"Labels\": 0}, {\"PC_1\": -6.077660741212628, \"PC_2\": 1.7601980195118747, \"Labels\": 5}, {\"PC_1\": -5.888719266976936, \"PC_2\": -2.2701894230321975, \"Labels\": 2}, {\"PC_1\": 10.330686211511072, \"PC_2\": 6.894579300646601, \"Labels\": 1}, {\"PC_1\": 3.356815348194068, \"PC_2\": 1.5585446419929192, \"Labels\": 4}, {\"PC_1\": -8.644964845559736, \"PC_2\": 1.6079891095194094, \"Labels\": 0}, {\"PC_1\": 14.031192805436726, \"PC_2\": -2.565122787697914, \"Labels\": 3}, {\"PC_1\": -3.082430223075838, \"PC_2\": -6.614978743908136, \"Labels\": 6}, {\"PC_1\": -9.560790665130115, \"PC_2\": -3.8804860413716047, \"Labels\": 2}], \"data-d07752ad55d31e1c387780ee4f199682\": [{\"PC_1\": -8.715435225764914, \"PC_2\": -3.650784608943655}, {\"PC_1\": 8.512793419415909, \"PC_2\": -4.053385974963417}, {\"PC_1\": -7.693303919133629, \"PC_2\": 0.6576724470683691}, {\"PC_1\": 13.748935333720317, \"PC_2\": 1.8902959567571664}, {\"PC_1\": 3.1848619528527777, \"PC_2\": -0.35545417067703083}, {\"PC_1\": -5.269085884995281, \"PC_2\": -2.4406733513713483}, {\"PC_1\": -3.7687656760951707, \"PC_2\": 7.952329702129914}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.LayerChart(...)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k_means_chart = visualize_with_pca (scaled_mags_df, k_means_predicted, k_means_model.cluster_centers_)\n",
    "k_means_chart\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "exclusive-kinase",
   "metadata": {},
   "source": [
    "### 2. OPTICS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ranking-cambridge",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 0,  1,  2,  1,  1,  1,  2,  2,  0,  0,  0,  0,  0,  1,  1,  1,  1,\n",
       "        1,  1,  0,  0,  1,  2, -1,  1])"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MIN_SAMPLES = 3\n",
    "\n",
    "optics_model = OPTICS (min_samples = MIN_SAMPLES, n_jobs = num_of_workers)\n",
    "optics_predicted = optics_model.fit_predict (scaled_mags_df)\n",
    "optics_predicted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "injured-taylor",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-54e1d3848e5a46089c9acbcffd697657\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-54e1d3848e5a46089c9acbcffd697657\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-54e1d3848e5a46089c9acbcffd697657\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.8.1?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function loadScript(lib) {\n",
       "      return new Promise(function(resolve, reject) {\n",
       "        var s = document.createElement('script');\n",
       "        s.src = paths[lib];\n",
       "        s.async = true;\n",
       "        s.onload = () => resolve(paths[lib]);\n",
       "        s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "        document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "      });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else if (typeof vegaEmbed === \"function\") {\n",
       "      displayChart(vegaEmbed);\n",
       "    } else {\n",
       "      loadScript(\"vega\")\n",
       "        .then(() => loadScript(\"vega-lite\"))\n",
       "        .then(() => loadScript(\"vega-embed\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}}, \"data\": {\"name\": \"data-891f658abad68637c832eb07d9f23604\"}, \"mark\": {\"type\": \"circle\", \"opacity\": 1}, \"encoding\": {\"color\": {\"type\": \"nominal\", \"field\": \"Labels\", \"legend\": {}}, \"x\": {\"type\": \"quantitative\", \"field\": \"PC_1\"}, \"y\": {\"type\": \"quantitative\", \"field\": \"PC_2\"}}, \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.8.1.json\", \"datasets\": {\"data-891f658abad68637c832eb07d9f23604\": [{\"PC_1\": 7.676253616139165, \"PC_2\": 6.443405176596787, \"Labels\": 0}, {\"PC_1\": -8.50469183576414, \"PC_2\": 1.7209173450327375, \"Labels\": 1}, {\"PC_1\": 10.824248687046426, \"PC_2\": -5.110219437734072, \"Labels\": 2}, {\"PC_1\": -9.80633885188385, \"PC_2\": 1.2041866623374056, \"Labels\": 1}, {\"PC_1\": -9.442347519453536, \"PC_2\": 1.0267784429715243, \"Labels\": 1}, {\"PC_1\": -6.430261739310041, \"PC_2\": -2.175666406065586, \"Labels\": 1}, {\"PC_1\": 16.02361892365011, \"PC_2\": -4.569272221832386, \"Labels\": 2}, {\"PC_1\": 14.307926781993281, \"PC_2\": -0.010167945813132004, \"Labels\": 2}, {\"PC_1\": 3.476445831112152, \"PC_2\": 2.6667062192963322, \"Labels\": 0}, {\"PC_1\": 1.9678988196289349, \"PC_2\": 0.4667960240696394, \"Labels\": 0}, {\"PC_1\": 5.975459567236973, \"PC_2\": -0.275921908725272, \"Labels\": 0}, {\"PC_1\": 6.916704987432156, \"PC_2\": 1.4237144507733028, \"Labels\": 0}, {\"PC_1\": 1.738644580544049, \"PC_2\": -0.4155862950046197, \"Labels\": 0}, {\"PC_1\": -8.02854221988149, \"PC_2\": -1.2848578715844472, \"Labels\": 1}, {\"PC_1\": -8.268914843331425, \"PC_2\": -1.200655944240077, \"Labels\": 1}, {\"PC_1\": -4.359284262212505, \"PC_2\": 1.7822336426377048, \"Labels\": 1}, {\"PC_1\": -8.53094914613286, \"PC_2\": 1.8170759916232007, \"Labels\": 1}, {\"PC_1\": -6.077660741212628, \"PC_2\": 1.7601980195118747, \"Labels\": 1}, {\"PC_1\": -5.888719266976936, \"PC_2\": -2.2701894230321975, \"Labels\": 1}, {\"PC_1\": 10.330686211511072, \"PC_2\": 6.894579300646601, \"Labels\": 0}, {\"PC_1\": 3.356815348194068, \"PC_2\": 1.5585446419929192, \"Labels\": 0}, {\"PC_1\": -8.644964845559736, \"PC_2\": 1.6079891095194094, \"Labels\": 1}, {\"PC_1\": 14.031192805436726, \"PC_2\": -2.565122787697914, \"Labels\": 2}, {\"PC_1\": -3.082430223075838, \"PC_2\": -6.614978743908136, \"Labels\": -1}, {\"PC_1\": -9.560790665130115, \"PC_2\": -3.8804860413716047, \"Labels\": 1}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.Chart(...)"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Visualize clusters, since there are no centroids, we are sending bogus array\n",
    "optics_chart = visualize_with_pca (scaled_mags_df, optics_predicted, np.empty([optics_predicted.shape[0], 1], dtype=int))\n",
    "optics_chart\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "running-wilson",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-a108d665d2ba40f98e55f2d2ca561265\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-a108d665d2ba40f98e55f2d2ca561265\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-a108d665d2ba40f98e55f2d2ca561265\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.8.1?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function loadScript(lib) {\n",
       "      return new Promise(function(resolve, reject) {\n",
       "        var s = document.createElement('script');\n",
       "        s.src = paths[lib];\n",
       "        s.async = true;\n",
       "        s.onload = () => resolve(paths[lib]);\n",
       "        s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "        document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "      });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else if (typeof vegaEmbed === \"function\") {\n",
       "      displayChart(vegaEmbed);\n",
       "    } else {\n",
       "      loadScript(\"vega\")\n",
       "        .then(() => loadScript(\"vega-lite\"))\n",
       "        .then(() => loadScript(\"vega-embed\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}}, \"hconcat\": [{\"layer\": [{\"data\": {\"name\": \"data-9fe3a9dde2bfb264c6a9c23aeb35dfad\"}, \"mark\": {\"type\": \"circle\", \"opacity\": 1}, \"encoding\": {\"color\": {\"type\": \"nominal\", \"field\": \"Labels\", \"legend\": {}}, \"x\": {\"type\": \"quantitative\", \"field\": \"PC_1\"}, \"y\": {\"type\": \"quantitative\", \"field\": \"PC_2\"}}}, {\"data\": {\"name\": \"data-d07752ad55d31e1c387780ee4f199682\"}, \"mark\": {\"type\": \"point\", \"color\": \"black\", \"opacity\": 0.7, \"shape\": \"diamond\", \"size\": 50}, \"encoding\": {\"x\": {\"type\": \"quantitative\", \"field\": \"PC_1\"}, \"y\": {\"type\": \"quantitative\", \"field\": \"PC_2\"}}}]}, {\"data\": {\"name\": \"data-891f658abad68637c832eb07d9f23604\"}, \"mark\": {\"type\": \"circle\", \"opacity\": 1}, \"encoding\": {\"color\": {\"type\": \"nominal\", \"field\": \"Labels\", \"legend\": {}}, \"x\": {\"type\": \"quantitative\", \"field\": \"PC_1\"}, \"y\": {\"type\": \"quantitative\", \"field\": \"PC_2\"}}}], \"resolve\": {\"scale\": {\"color\": \"independent\"}}, \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.8.1.json\", \"datasets\": {\"data-9fe3a9dde2bfb264c6a9c23aeb35dfad\": [{\"PC_1\": 7.676253616139165, \"PC_2\": 6.443405176596787, \"Labels\": 1}, {\"PC_1\": -8.50469183576414, \"PC_2\": 1.7209173450327375, \"Labels\": 0}, {\"PC_1\": 10.824248687046426, \"PC_2\": -5.110219437734072, \"Labels\": 3}, {\"PC_1\": -9.80633885188385, \"PC_2\": 1.2041866623374056, \"Labels\": 0}, {\"PC_1\": -9.442347519453536, \"PC_2\": 1.0267784429715243, \"Labels\": 0}, {\"PC_1\": -6.430261739310041, \"PC_2\": -2.175666406065586, \"Labels\": 2}, {\"PC_1\": 16.02361892365011, \"PC_2\": -4.569272221832386, \"Labels\": 3}, {\"PC_1\": 14.307926781993281, \"PC_2\": -0.010167945813132004, \"Labels\": 3}, {\"PC_1\": 3.476445831112152, \"PC_2\": 2.6667062192963322, \"Labels\": 4}, {\"PC_1\": 1.9678988196289349, \"PC_2\": 0.4667960240696394, \"Labels\": 4}, {\"PC_1\": 5.975459567236973, \"PC_2\": -0.275921908725272, \"Labels\": 4}, {\"PC_1\": 6.916704987432156, \"PC_2\": 1.4237144507733028, \"Labels\": 1}, {\"PC_1\": 1.738644580544049, \"PC_2\": -0.4155862950046197, \"Labels\": 4}, {\"PC_1\": -8.02854221988149, \"PC_2\": -1.2848578715844472, \"Labels\": 2}, {\"PC_1\": -8.268914843331425, \"PC_2\": -1.200655944240077, \"Labels\": 2}, {\"PC_1\": -4.359284262212505, \"PC_2\": 1.7822336426377048, \"Labels\": 5}, {\"PC_1\": -8.53094914613286, \"PC_2\": 1.8170759916232007, \"Labels\": 0}, {\"PC_1\": -6.077660741212628, \"PC_2\": 1.7601980195118747, \"Labels\": 5}, {\"PC_1\": -5.888719266976936, \"PC_2\": -2.2701894230321975, \"Labels\": 2}, {\"PC_1\": 10.330686211511072, \"PC_2\": 6.894579300646601, \"Labels\": 1}, {\"PC_1\": 3.356815348194068, \"PC_2\": 1.5585446419929192, \"Labels\": 4}, {\"PC_1\": -8.644964845559736, \"PC_2\": 1.6079891095194094, \"Labels\": 0}, {\"PC_1\": 14.031192805436726, \"PC_2\": -2.565122787697914, \"Labels\": 3}, {\"PC_1\": -3.082430223075838, \"PC_2\": -6.614978743908136, \"Labels\": 6}, {\"PC_1\": -9.560790665130115, \"PC_2\": -3.8804860413716047, \"Labels\": 2}], \"data-d07752ad55d31e1c387780ee4f199682\": [{\"PC_1\": -8.715435225764914, \"PC_2\": -3.650784608943655}, {\"PC_1\": 8.512793419415909, \"PC_2\": -4.053385974963417}, {\"PC_1\": -7.693303919133629, \"PC_2\": 0.6576724470683691}, {\"PC_1\": 13.748935333720317, \"PC_2\": 1.8902959567571664}, {\"PC_1\": 3.1848619528527777, \"PC_2\": -0.35545417067703083}, {\"PC_1\": -5.269085884995281, \"PC_2\": -2.4406733513713483}, {\"PC_1\": -3.7687656760951707, \"PC_2\": 7.952329702129914}], \"data-891f658abad68637c832eb07d9f23604\": [{\"PC_1\": 7.676253616139165, \"PC_2\": 6.443405176596787, \"Labels\": 0}, {\"PC_1\": -8.50469183576414, \"PC_2\": 1.7209173450327375, \"Labels\": 1}, {\"PC_1\": 10.824248687046426, \"PC_2\": -5.110219437734072, \"Labels\": 2}, {\"PC_1\": -9.80633885188385, \"PC_2\": 1.2041866623374056, \"Labels\": 1}, {\"PC_1\": -9.442347519453536, \"PC_2\": 1.0267784429715243, \"Labels\": 1}, {\"PC_1\": -6.430261739310041, \"PC_2\": -2.175666406065586, \"Labels\": 1}, {\"PC_1\": 16.02361892365011, \"PC_2\": -4.569272221832386, \"Labels\": 2}, {\"PC_1\": 14.307926781993281, \"PC_2\": -0.010167945813132004, \"Labels\": 2}, {\"PC_1\": 3.476445831112152, \"PC_2\": 2.6667062192963322, \"Labels\": 0}, {\"PC_1\": 1.9678988196289349, \"PC_2\": 0.4667960240696394, \"Labels\": 0}, {\"PC_1\": 5.975459567236973, \"PC_2\": -0.275921908725272, \"Labels\": 0}, {\"PC_1\": 6.916704987432156, \"PC_2\": 1.4237144507733028, \"Labels\": 0}, {\"PC_1\": 1.738644580544049, \"PC_2\": -0.4155862950046197, \"Labels\": 0}, {\"PC_1\": -8.02854221988149, \"PC_2\": -1.2848578715844472, \"Labels\": 1}, {\"PC_1\": -8.268914843331425, \"PC_2\": -1.200655944240077, \"Labels\": 1}, {\"PC_1\": -4.359284262212505, \"PC_2\": 1.7822336426377048, \"Labels\": 1}, {\"PC_1\": -8.53094914613286, \"PC_2\": 1.8170759916232007, \"Labels\": 1}, {\"PC_1\": -6.077660741212628, \"PC_2\": 1.7601980195118747, \"Labels\": 1}, {\"PC_1\": -5.888719266976936, \"PC_2\": -2.2701894230321975, \"Labels\": 1}, {\"PC_1\": 10.330686211511072, \"PC_2\": 6.894579300646601, \"Labels\": 0}, {\"PC_1\": 3.356815348194068, \"PC_2\": 1.5585446419929192, \"Labels\": 0}, {\"PC_1\": -8.644964845559736, \"PC_2\": 1.6079891095194094, \"Labels\": 1}, {\"PC_1\": 14.031192805436726, \"PC_2\": -2.565122787697914, \"Labels\": 2}, {\"PC_1\": -3.082430223075838, \"PC_2\": -6.614978743908136, \"Labels\": -1}, {\"PC_1\": -9.560790665130115, \"PC_2\": -3.8804860413716047, \"Labels\": 1}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.HConcatChart(...)"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Side by side comparison\n",
    "alt.hconcat (k_means_chart, optics_chart).resolve_scale(color='independent')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "southwest-documentation",
   "metadata": {},
   "source": [
    "## Evaluation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "boring-blend",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Silhouette scores: [best = 1, worst = -1]\n",
      "\t1. K-means: 0.3284530838720724\n",
      "\t2. OPTICS: 0.4068072149015211\n"
     ]
    }
   ],
   "source": [
    "eval_k_means = metrics.silhouette_score (scaled_mags_df, k_means_predicted)\n",
    "eval_optics = metrics.silhouette_score (scaled_mags_df, optics_predicted)\n",
    "\n",
    "print ('Silhouette scores: [best = 1, worst = -1]')\n",
    "print ('\\t1. K-means:', eval_k_means)\n",
    "print ('\\t2. OPTICS:', eval_optics)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "horizontal-killing",
   "metadata": {},
   "source": [
    "---\n",
    "# METABOLOMIC ANALYSIS\n",
    "---\n",
    "## Importing Metabolomic data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "ambient-declaration",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Metabolite</th>\n",
       "      <th>tp</th>\n",
       "      <th>date</th>\n",
       "      <th>type</th>\n",
       "      <th>known_type</th>\n",
       "      <th>means</th>\n",
       "      <th>medians</th>\n",
       "      <th>sds</th>\n",
       "      <th>N</th>\n",
       "      <th>se</th>\n",
       "      <th>ci</th>\n",
       "      <th>type2</th>\n",
       "      <th>measurement</th>\n",
       "      <th>KEGG.Compound.ID</th>\n",
       "      <th>Chebi.Name</th>\n",
       "      <th>Chebi.Name_combined</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Phosphoric_acid_3TMS</td>\n",
       "      <td>D1</td>\n",
       "      <td>2012-04-17</td>\n",
       "      <td>snp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.757113</td>\n",
       "      <td>0.737943</td>\n",
       "      <td>0.098650</td>\n",
       "      <td>3</td>\n",
       "      <td>0.056956</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>C00009</td>\n",
       "      <td>phosphoric acid</td>\n",
       "      <td>phosphoric acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Phosphoric_acid_3TMS</td>\n",
       "      <td>D10</td>\n",
       "      <td>2012-01-25</td>\n",
       "      <td>snp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.720935</td>\n",
       "      <td>0.873509</td>\n",
       "      <td>0.266593</td>\n",
       "      <td>3</td>\n",
       "      <td>0.153918</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>C00009</td>\n",
       "      <td>phosphoric acid</td>\n",
       "      <td>phosphoric acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Phosphoric_acid_3TMS</td>\n",
       "      <td>D11</td>\n",
       "      <td>2012-03-22</td>\n",
       "      <td>snp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.625047</td>\n",
       "      <td>0.652268</td>\n",
       "      <td>0.057820</td>\n",
       "      <td>3</td>\n",
       "      <td>0.033382</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>C00009</td>\n",
       "      <td>phosphoric acid</td>\n",
       "      <td>phosphoric acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Phosphoric_acid_3TMS</td>\n",
       "      <td>D12</td>\n",
       "      <td>2012-01-19</td>\n",
       "      <td>snp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.964956</td>\n",
       "      <td>0.941878</td>\n",
       "      <td>0.053754</td>\n",
       "      <td>3</td>\n",
       "      <td>0.031035</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>C00009</td>\n",
       "      <td>phosphoric acid</td>\n",
       "      <td>phosphoric acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>!Phosphoric_acid_3TMS</td>\n",
       "      <td>D13</td>\n",
       "      <td>2011-05-13</td>\n",
       "      <td>snp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.963984</td>\n",
       "      <td>0.974479</td>\n",
       "      <td>0.071659</td>\n",
       "      <td>3</td>\n",
       "      <td>0.041373</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>C00009</td>\n",
       "      <td>phosphoric acid</td>\n",
       "      <td>phosphoric acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18865</th>\n",
       "      <td>Valine_2TMS</td>\n",
       "      <td>D51</td>\n",
       "      <td>2011-08-11</td>\n",
       "      <td>bp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.704821</td>\n",
       "      <td>0.478251</td>\n",
       "      <td>0.428368</td>\n",
       "      <td>3</td>\n",
       "      <td>0.247319</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C16436,C06417,C00183</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18866</th>\n",
       "      <td>Valine_2TMS</td>\n",
       "      <td>D6</td>\n",
       "      <td>2011-06-03</td>\n",
       "      <td>bp</td>\n",
       "      <td>known</td>\n",
       "      <td>1.085888</td>\n",
       "      <td>1.197471</td>\n",
       "      <td>0.487484</td>\n",
       "      <td>3</td>\n",
       "      <td>0.281449</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C16436,C06417,C00183</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18867</th>\n",
       "      <td>Valine_2TMS</td>\n",
       "      <td>D7</td>\n",
       "      <td>2012-04-04</td>\n",
       "      <td>bp</td>\n",
       "      <td>known</td>\n",
       "      <td>2.958023</td>\n",
       "      <td>2.813145</td>\n",
       "      <td>0.341213</td>\n",
       "      <td>3</td>\n",
       "      <td>0.197000</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C16436,C06417,C00183</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18868</th>\n",
       "      <td>Valine_2TMS</td>\n",
       "      <td>D8</td>\n",
       "      <td>2012-02-08</td>\n",
       "      <td>bp</td>\n",
       "      <td>known</td>\n",
       "      <td>2.515980</td>\n",
       "      <td>2.322248</td>\n",
       "      <td>0.618556</td>\n",
       "      <td>3</td>\n",
       "      <td>0.357123</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C16436,C06417,C00183</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18869</th>\n",
       "      <td>Valine_2TMS</td>\n",
       "      <td>D9</td>\n",
       "      <td>2011-12-28</td>\n",
       "      <td>bp</td>\n",
       "      <td>known</td>\n",
       "      <td>2.356989</td>\n",
       "      <td>2.280757</td>\n",
       "      <td>0.264918</td>\n",
       "      <td>3</td>\n",
       "      <td>0.152950</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C16436,C06417,C00183</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18870 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Metabolite   tp        date type known_type     means  \\\n",
       "0      !Phosphoric_acid_3TMS   D1  2012-04-17  snp      known  0.757113   \n",
       "1      !Phosphoric_acid_3TMS  D10  2012-01-25  snp      known  0.720935   \n",
       "2      !Phosphoric_acid_3TMS  D11  2012-03-22  snp      known  0.625047   \n",
       "3      !Phosphoric_acid_3TMS  D12  2012-01-19  snp      known  0.964956   \n",
       "4      !Phosphoric_acid_3TMS  D13  2011-05-13  snp      known  0.963984   \n",
       "...                      ...  ...         ...  ...        ...       ...   \n",
       "18865            Valine_2TMS  D51  2011-08-11   bp      known  0.704821   \n",
       "18866            Valine_2TMS   D6  2011-06-03   bp      known  1.085888   \n",
       "18867            Valine_2TMS   D7  2012-04-04   bp      known  2.958023   \n",
       "18868            Valine_2TMS   D8  2012-02-08   bp      known  2.515980   \n",
       "18869            Valine_2TMS   D9  2011-12-28   bp      known  2.356989   \n",
       "\n",
       "        medians       sds  N        se        ci          type2 measurement  \\\n",
       "0      0.737943  0.098650  3  0.056956  4.302653  extracellular    nonpolar   \n",
       "1      0.873509  0.266593  3  0.153918  4.302653  extracellular    nonpolar   \n",
       "2      0.652268  0.057820  3  0.033382  4.302653  extracellular    nonpolar   \n",
       "3      0.941878  0.053754  3  0.031035  4.302653  extracellular    nonpolar   \n",
       "4      0.974479  0.071659  3  0.041373  4.302653  extracellular    nonpolar   \n",
       "...         ...       ... ..       ...       ...            ...         ...   \n",
       "18865  0.478251  0.428368  3  0.247319  4.302653  intracellular       polar   \n",
       "18866  1.197471  0.487484  3  0.281449  4.302653  intracellular       polar   \n",
       "18867  2.813145  0.341213  3  0.197000  4.302653  intracellular       polar   \n",
       "18868  2.322248  0.618556  3  0.357123  4.302653  intracellular       polar   \n",
       "18869  2.280757  0.264918  3  0.152950  4.302653  intracellular       polar   \n",
       "\n",
       "           KEGG.Compound.ID                Chebi.Name  \\\n",
       "0                    C00009           phosphoric acid   \n",
       "1                    C00009           phosphoric acid   \n",
       "2                    C00009           phosphoric acid   \n",
       "3                    C00009           phosphoric acid   \n",
       "4                    C00009           phosphoric acid   \n",
       "...                     ...                       ...   \n",
       "18865  C16436,C06417,C00183  valine;D-valine;L-valine   \n",
       "18866  C16436,C06417,C00183  valine;D-valine;L-valine   \n",
       "18867  C16436,C06417,C00183  valine;D-valine;L-valine   \n",
       "18868  C16436,C06417,C00183  valine;D-valine;L-valine   \n",
       "18869  C16436,C06417,C00183  valine;D-valine;L-valine   \n",
       "\n",
       "            Chebi.Name_combined  \n",
       "0               phosphoric acid  \n",
       "1               phosphoric acid  \n",
       "2               phosphoric acid  \n",
       "3               phosphoric acid  \n",
       "4               phosphoric acid  \n",
       "...                         ...  \n",
       "18865  valine;D-valine;L-valine  \n",
       "18866  valine;D-valine;L-valine  \n",
       "18867  valine;D-valine;L-valine  \n",
       "18868  valine;D-valine;L-valine  \n",
       "18869  valine;D-valine;L-valine  \n",
       "\n",
       "[18870 rows x 16 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metabolomics_file_name = os.path.join(path_normalised_metabolomics, os.listdir(path_normalised_metabolomics)[0])\n",
    "metabolomics_df = pd.read_csv (metabolomics_file_name, delimiter = '\\t')\n",
    "metabolomics_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "sublime-testament",
   "metadata": {},
   "source": [
    "## Data preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "behind-stereo",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>date</th>\n",
       "      <th>Metabolite</th>\n",
       "      <th>tp</th>\n",
       "      <th>type</th>\n",
       "      <th>known_type</th>\n",
       "      <th>means</th>\n",
       "      <th>medians</th>\n",
       "      <th>sds</th>\n",
       "      <th>N</th>\n",
       "      <th>se</th>\n",
       "      <th>ci</th>\n",
       "      <th>type2</th>\n",
       "      <th>measurement</th>\n",
       "      <th>KEGG.Compound.ID</th>\n",
       "      <th>Chebi.Name</th>\n",
       "      <th>Chebi.Name_combined</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>Leucine_2TMS</td>\n",
       "      <td>D37</td>\n",
       "      <td>sp</td>\n",
       "      <td>known</td>\n",
       "      <td>1.440122</td>\n",
       "      <td>1.489979</td>\n",
       "      <td>0.146440</td>\n",
       "      <td>3</td>\n",
       "      <td>0.084547</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C16439,C00123,C01570</td>\n",
       "      <td>leucine;L-leucine;D-leucine</td>\n",
       "      <td>leucine;L-leucine;D-leucine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>No match: 1181.78_EM_SP_D1_1_223</td>\n",
       "      <td>D37</td>\n",
       "      <td>sp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>3.778469</td>\n",
       "      <td>3.801417</td>\n",
       "      <td>1.924377</td>\n",
       "      <td>3</td>\n",
       "      <td>1.111040</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>unknown#emu_WW_2144.25</td>\n",
       "      <td>D37</td>\n",
       "      <td>bnp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.949545</td>\n",
       "      <td>0.947209</td>\n",
       "      <td>0.065422</td>\n",
       "      <td>3</td>\n",
       "      <td>0.037771</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>Unknown#bth_pae_001</td>\n",
       "      <td>D37</td>\n",
       "      <td>bp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2.095116</td>\n",
       "      <td>1.809891</td>\n",
       "      <td>1.916896</td>\n",
       "      <td>3</td>\n",
       "      <td>1.106721</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>No match: 2220.34_EM_BP_D1_1_192</td>\n",
       "      <td>D37</td>\n",
       "      <td>bp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.631643</td>\n",
       "      <td>1.571125</td>\n",
       "      <td>0.168501</td>\n",
       "      <td>3</td>\n",
       "      <td>0.097284</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18865</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>Octadecanol_1TMS</td>\n",
       "      <td>D48</td>\n",
       "      <td>bnp</td>\n",
       "      <td>known</td>\n",
       "      <td>1.364105</td>\n",
       "      <td>1.457780</td>\n",
       "      <td>0.170311</td>\n",
       "      <td>3</td>\n",
       "      <td>0.098329</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>D01924</td>\n",
       "      <td>octadecan-1-ol</td>\n",
       "      <td>octadecan-1-ol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18866</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>No match: 1447.93_EM_BP_D2_1_57</td>\n",
       "      <td>D48</td>\n",
       "      <td>bp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>4.815774</td>\n",
       "      <td>1.160165</td>\n",
       "      <td>6.835474</td>\n",
       "      <td>3</td>\n",
       "      <td>3.946463</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18867</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>Lactulose_8TMS</td>\n",
       "      <td>D48</td>\n",
       "      <td>bp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.697489</td>\n",
       "      <td>0.684339</td>\n",
       "      <td>0.049632</td>\n",
       "      <td>3</td>\n",
       "      <td>0.028655</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C07064</td>\n",
       "      <td>lactulose</td>\n",
       "      <td>lactulose</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18868</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>No match: 2004.78_EM_BNP_D10_1_205</td>\n",
       "      <td>D48</td>\n",
       "      <td>bnp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.785996</td>\n",
       "      <td>0.744529</td>\n",
       "      <td>0.139901</td>\n",
       "      <td>3</td>\n",
       "      <td>0.080772</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18869</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>2-Oxobutanoic_acid_1_1MeOX_1TMS</td>\n",
       "      <td>D48</td>\n",
       "      <td>sp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.755565</td>\n",
       "      <td>0.709972</td>\n",
       "      <td>0.149214</td>\n",
       "      <td>3</td>\n",
       "      <td>0.086149</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C00109</td>\n",
       "      <td>2-oxobutanoic acid</td>\n",
       "      <td>2-oxobutanoic acid</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18870 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            date                          Metabolite   tp type known_type  \\\n",
       "0     2011-03-21                        Leucine_2TMS  D37   sp      known   \n",
       "1     2011-03-21    No match: 1181.78_EM_SP_D1_1_223  D37   sp    unknown   \n",
       "2     2011-03-21              unknown#emu_WW_2144.25  D37  bnp    unknown   \n",
       "3     2011-03-21                 Unknown#bth_pae_001  D37   bp    unknown   \n",
       "4     2011-03-21    No match: 2220.34_EM_BP_D1_1_192  D37   bp    unknown   \n",
       "...          ...                                 ...  ...  ...        ...   \n",
       "18865 2012-05-03                    Octadecanol_1TMS  D48  bnp      known   \n",
       "18866 2012-05-03     No match: 1447.93_EM_BP_D2_1_57  D48   bp    unknown   \n",
       "18867 2012-05-03                      Lactulose_8TMS  D48   bp      known   \n",
       "18868 2012-05-03  No match: 2004.78_EM_BNP_D10_1_205  D48  bnp    unknown   \n",
       "18869 2012-05-03     2-Oxobutanoic_acid_1_1MeOX_1TMS  D48   sp      known   \n",
       "\n",
       "          means   medians       sds  N        se        ci          type2  \\\n",
       "0      1.440122  1.489979  0.146440  3  0.084547  4.302653  extracellular   \n",
       "1      3.778469  3.801417  1.924377  3  1.111040  4.302653  extracellular   \n",
       "2      0.949545  0.947209  0.065422  3  0.037771  4.302653  intracellular   \n",
       "3      2.095116  1.809891  1.916896  3  1.106721  4.302653  intracellular   \n",
       "4      1.631643  1.571125  0.168501  3  0.097284  4.302653  intracellular   \n",
       "...         ...       ...       ... ..       ...       ...            ...   \n",
       "18865  1.364105  1.457780  0.170311  3  0.098329  4.302653  intracellular   \n",
       "18866  4.815774  1.160165  6.835474  3  3.946463  4.302653  intracellular   \n",
       "18867  0.697489  0.684339  0.049632  3  0.028655  4.302653  intracellular   \n",
       "18868  0.785996  0.744529  0.139901  3  0.080772  4.302653  intracellular   \n",
       "18869  0.755565  0.709972  0.149214  3  0.086149  4.302653  extracellular   \n",
       "\n",
       "      measurement      KEGG.Compound.ID                   Chebi.Name  \\\n",
       "0           polar  C16439,C00123,C01570  leucine;L-leucine;D-leucine   \n",
       "1           polar                   NaN                          NaN   \n",
       "2        nonpolar                   NaN                          NaN   \n",
       "3           polar                   NaN                          NaN   \n",
       "4           polar                   NaN                          NaN   \n",
       "...           ...                   ...                          ...   \n",
       "18865    nonpolar                D01924               octadecan-1-ol   \n",
       "18866       polar                   NaN                          NaN   \n",
       "18867       polar                C07064                    lactulose   \n",
       "18868    nonpolar                   NaN                          NaN   \n",
       "18869       polar                C00109           2-oxobutanoic acid   \n",
       "\n",
       "               Chebi.Name_combined  \n",
       "0      leucine;L-leucine;D-leucine  \n",
       "1                              NaN  \n",
       "2                              NaN  \n",
       "3                              NaN  \n",
       "4                              NaN  \n",
       "...                            ...  \n",
       "18865               octadecan-1-ol  \n",
       "18866                          NaN  \n",
       "18867                    lactulose  \n",
       "18868                          NaN  \n",
       "18869           2-oxobutanoic acid  \n",
       "\n",
       "[18870 rows x 16 columns]"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metabolomics_df['date'] = pd.to_datetime(metabolomics_df['date'])\n",
    "metabolomics_df.insert (0, 'date', metabolomics_df.pop('date'))\n",
    "metabolomics_df.sort_values ('date', inplace = True, ignore_index = True)\n",
    "metabolomics_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "exceptional-appearance",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dataset uniqueness:\n",
      "\t1. Metabolites: 313\n",
      "\t2. Types: 4\n",
      "\t3. Known types: 2\n",
      "\t2. Ns: 2\n",
      "\t2. Type 2s 2\n",
      "\t2. Measurements 2\n"
     ]
    }
   ],
   "source": [
    "print ('Dataset uniqueness:')\n",
    "print ('\\t1. Metabolites:', len(np.unique(metabolomics_df['Metabolite'].values)))\n",
    "print ('\\t2. Types:', len(np.unique(metabolomics_df['type'].values)))\n",
    "print ('\\t3. Known types:', len(np.unique(metabolomics_df['known_type'].values)))\n",
    "print ('\\t2. Ns:', len(np.unique(metabolomics_df['N'].values)))\n",
    "print ('\\t2. Type 2s', len(np.unique(metabolomics_df['type2'].values)))\n",
    "print ('\\t2. Measurements', len(np.unique(metabolomics_df['measurement'].values)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "vietnamese-framework",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>date</th>\n",
       "      <th>Metabolite</th>\n",
       "      <th>tp</th>\n",
       "      <th>type</th>\n",
       "      <th>known_type</th>\n",
       "      <th>means</th>\n",
       "      <th>medians</th>\n",
       "      <th>sds</th>\n",
       "      <th>N</th>\n",
       "      <th>se</th>\n",
       "      <th>ci</th>\n",
       "      <th>type2</th>\n",
       "      <th>measurement</th>\n",
       "      <th>KEGG.Compound.ID</th>\n",
       "      <th>Chebi.Name</th>\n",
       "      <th>Chebi.Name_combined</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>Leucine_2TMS</td>\n",
       "      <td>D37</td>\n",
       "      <td>sp</td>\n",
       "      <td>known</td>\n",
       "      <td>1.440122</td>\n",
       "      <td>1.489979</td>\n",
       "      <td>0.146440</td>\n",
       "      <td>3</td>\n",
       "      <td>0.084547</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C16439,C00123,C01570</td>\n",
       "      <td>leucine;L-leucine;D-leucine</td>\n",
       "      <td>leucine;L-leucine;D-leucine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>D37</td>\n",
       "      <td>sp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>3.778469</td>\n",
       "      <td>3.801417</td>\n",
       "      <td>1.924377</td>\n",
       "      <td>3</td>\n",
       "      <td>1.111040</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>D37</td>\n",
       "      <td>bnp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.949545</td>\n",
       "      <td>0.947209</td>\n",
       "      <td>0.065422</td>\n",
       "      <td>3</td>\n",
       "      <td>0.037771</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>D37</td>\n",
       "      <td>bp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2.095116</td>\n",
       "      <td>1.809891</td>\n",
       "      <td>1.916896</td>\n",
       "      <td>3</td>\n",
       "      <td>1.106721</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>D37</td>\n",
       "      <td>bp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.631643</td>\n",
       "      <td>1.571125</td>\n",
       "      <td>0.168501</td>\n",
       "      <td>3</td>\n",
       "      <td>0.097284</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18865</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>Octadecanol_1TMS</td>\n",
       "      <td>D48</td>\n",
       "      <td>bnp</td>\n",
       "      <td>known</td>\n",
       "      <td>1.364105</td>\n",
       "      <td>1.457780</td>\n",
       "      <td>0.170311</td>\n",
       "      <td>3</td>\n",
       "      <td>0.098329</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>D01924</td>\n",
       "      <td>octadecan-1-ol</td>\n",
       "      <td>octadecan-1-ol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18866</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>D48</td>\n",
       "      <td>bp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>4.815774</td>\n",
       "      <td>1.160165</td>\n",
       "      <td>6.835474</td>\n",
       "      <td>3</td>\n",
       "      <td>3.946463</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18867</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>Lactulose_8TMS</td>\n",
       "      <td>D48</td>\n",
       "      <td>bp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.697489</td>\n",
       "      <td>0.684339</td>\n",
       "      <td>0.049632</td>\n",
       "      <td>3</td>\n",
       "      <td>0.028655</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C07064</td>\n",
       "      <td>lactulose</td>\n",
       "      <td>lactulose</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18868</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>D48</td>\n",
       "      <td>bnp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.785996</td>\n",
       "      <td>0.744529</td>\n",
       "      <td>0.139901</td>\n",
       "      <td>3</td>\n",
       "      <td>0.080772</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18869</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>2-Oxobutanoic_acid_1_1MeOX_1TMS</td>\n",
       "      <td>D48</td>\n",
       "      <td>sp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.755565</td>\n",
       "      <td>0.709972</td>\n",
       "      <td>0.149214</td>\n",
       "      <td>3</td>\n",
       "      <td>0.086149</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C00109</td>\n",
       "      <td>2-oxobutanoic acid</td>\n",
       "      <td>2-oxobutanoic acid</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18870 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            date                       Metabolite   tp type known_type  \\\n",
       "0     2011-03-21                     Leucine_2TMS  D37   sp      known   \n",
       "1     2011-03-21                          Unknown  D37   sp    unknown   \n",
       "2     2011-03-21                          Unknown  D37  bnp    unknown   \n",
       "3     2011-03-21                          Unknown  D37   bp    unknown   \n",
       "4     2011-03-21                          Unknown  D37   bp    unknown   \n",
       "...          ...                              ...  ...  ...        ...   \n",
       "18865 2012-05-03                 Octadecanol_1TMS  D48  bnp      known   \n",
       "18866 2012-05-03                          Unknown  D48   bp    unknown   \n",
       "18867 2012-05-03                   Lactulose_8TMS  D48   bp      known   \n",
       "18868 2012-05-03                          Unknown  D48  bnp    unknown   \n",
       "18869 2012-05-03  2-Oxobutanoic_acid_1_1MeOX_1TMS  D48   sp      known   \n",
       "\n",
       "          means   medians       sds  N        se        ci          type2  \\\n",
       "0      1.440122  1.489979  0.146440  3  0.084547  4.302653  extracellular   \n",
       "1      3.778469  3.801417  1.924377  3  1.111040  4.302653  extracellular   \n",
       "2      0.949545  0.947209  0.065422  3  0.037771  4.302653  intracellular   \n",
       "3      2.095116  1.809891  1.916896  3  1.106721  4.302653  intracellular   \n",
       "4      1.631643  1.571125  0.168501  3  0.097284  4.302653  intracellular   \n",
       "...         ...       ...       ... ..       ...       ...            ...   \n",
       "18865  1.364105  1.457780  0.170311  3  0.098329  4.302653  intracellular   \n",
       "18866  4.815774  1.160165  6.835474  3  3.946463  4.302653  intracellular   \n",
       "18867  0.697489  0.684339  0.049632  3  0.028655  4.302653  intracellular   \n",
       "18868  0.785996  0.744529  0.139901  3  0.080772  4.302653  intracellular   \n",
       "18869  0.755565  0.709972  0.149214  3  0.086149  4.302653  extracellular   \n",
       "\n",
       "      measurement      KEGG.Compound.ID                   Chebi.Name  \\\n",
       "0           polar  C16439,C00123,C01570  leucine;L-leucine;D-leucine   \n",
       "1           polar                   NaN                          NaN   \n",
       "2        nonpolar                   NaN                          NaN   \n",
       "3           polar                   NaN                          NaN   \n",
       "4           polar                   NaN                          NaN   \n",
       "...           ...                   ...                          ...   \n",
       "18865    nonpolar                D01924               octadecan-1-ol   \n",
       "18866       polar                   NaN                          NaN   \n",
       "18867       polar                C07064                    lactulose   \n",
       "18868    nonpolar                   NaN                          NaN   \n",
       "18869       polar                C00109           2-oxobutanoic acid   \n",
       "\n",
       "               Chebi.Name_combined  \n",
       "0      leucine;L-leucine;D-leucine  \n",
       "1                              NaN  \n",
       "2                              NaN  \n",
       "3                              NaN  \n",
       "4                              NaN  \n",
       "...                            ...  \n",
       "18865               octadecan-1-ol  \n",
       "18866                          NaN  \n",
       "18867                    lactulose  \n",
       "18868                          NaN  \n",
       "18869           2-oxobutanoic acid  \n",
       "\n",
       "[18870 rows x 16 columns]"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Changing metabolite name if it is unknown\n",
    "metabolomics_df.loc[metabolomics_df['known_type'].eq('unknown'), 'Metabolite'] = 'Unknown'\n",
    "metabolomics_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "early-avenue",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>date</th>\n",
       "      <th>type</th>\n",
       "      <th>known_type</th>\n",
       "      <th>means</th>\n",
       "      <th>medians</th>\n",
       "      <th>sds</th>\n",
       "      <th>N</th>\n",
       "      <th>se</th>\n",
       "      <th>ci</th>\n",
       "      <th>type2</th>\n",
       "      <th>measurement</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>sp</td>\n",
       "      <td>known</td>\n",
       "      <td>1.440122</td>\n",
       "      <td>1.489979</td>\n",
       "      <td>0.146440</td>\n",
       "      <td>3</td>\n",
       "      <td>0.084547</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>polar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>sp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>3.778469</td>\n",
       "      <td>3.801417</td>\n",
       "      <td>1.924377</td>\n",
       "      <td>3</td>\n",
       "      <td>1.111040</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>polar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>bnp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.949545</td>\n",
       "      <td>0.947209</td>\n",
       "      <td>0.065422</td>\n",
       "      <td>3</td>\n",
       "      <td>0.037771</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>bp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2.095116</td>\n",
       "      <td>1.809891</td>\n",
       "      <td>1.916896</td>\n",
       "      <td>3</td>\n",
       "      <td>1.106721</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>bp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.631643</td>\n",
       "      <td>1.571125</td>\n",
       "      <td>0.168501</td>\n",
       "      <td>3</td>\n",
       "      <td>0.097284</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18865</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>bnp</td>\n",
       "      <td>known</td>\n",
       "      <td>1.364105</td>\n",
       "      <td>1.457780</td>\n",
       "      <td>0.170311</td>\n",
       "      <td>3</td>\n",
       "      <td>0.098329</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18866</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>bp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>4.815774</td>\n",
       "      <td>1.160165</td>\n",
       "      <td>6.835474</td>\n",
       "      <td>3</td>\n",
       "      <td>3.946463</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18867</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>bp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.697489</td>\n",
       "      <td>0.684339</td>\n",
       "      <td>0.049632</td>\n",
       "      <td>3</td>\n",
       "      <td>0.028655</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18868</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>bnp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.785996</td>\n",
       "      <td>0.744529</td>\n",
       "      <td>0.139901</td>\n",
       "      <td>3</td>\n",
       "      <td>0.080772</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18869</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>sp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.755565</td>\n",
       "      <td>0.709972</td>\n",
       "      <td>0.149214</td>\n",
       "      <td>3</td>\n",
       "      <td>0.086149</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>polar</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18870 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            date type known_type     means   medians       sds  N        se  \\\n",
       "0     2011-03-21   sp      known  1.440122  1.489979  0.146440  3  0.084547   \n",
       "1     2011-03-21   sp    unknown  3.778469  3.801417  1.924377  3  1.111040   \n",
       "2     2011-03-21  bnp    unknown  0.949545  0.947209  0.065422  3  0.037771   \n",
       "3     2011-03-21   bp    unknown  2.095116  1.809891  1.916896  3  1.106721   \n",
       "4     2011-03-21   bp    unknown  1.631643  1.571125  0.168501  3  0.097284   \n",
       "...          ...  ...        ...       ...       ...       ... ..       ...   \n",
       "18865 2012-05-03  bnp      known  1.364105  1.457780  0.170311  3  0.098329   \n",
       "18866 2012-05-03   bp    unknown  4.815774  1.160165  6.835474  3  3.946463   \n",
       "18867 2012-05-03   bp      known  0.697489  0.684339  0.049632  3  0.028655   \n",
       "18868 2012-05-03  bnp    unknown  0.785996  0.744529  0.139901  3  0.080772   \n",
       "18869 2012-05-03   sp      known  0.755565  0.709972  0.149214  3  0.086149   \n",
       "\n",
       "             ci          type2 measurement  \n",
       "0      4.302653  extracellular       polar  \n",
       "1      4.302653  extracellular       polar  \n",
       "2      4.302653  intracellular    nonpolar  \n",
       "3      4.302653  intracellular       polar  \n",
       "4      4.302653  intracellular       polar  \n",
       "...         ...            ...         ...  \n",
       "18865  4.302653  intracellular    nonpolar  \n",
       "18866  4.302653  intracellular       polar  \n",
       "18867  4.302653  intracellular       polar  \n",
       "18868  4.302653  intracellular    nonpolar  \n",
       "18869  4.302653  extracellular       polar  \n",
       "\n",
       "[18870 rows x 11 columns]"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Saving the name column and removing unnecessairy columns\n",
    "metabolite_names = metabolomics_df['Metabolite']\n",
    "metabolomics_df.drop(labels = ['Metabolite', 'tp', 'KEGG.Compound.ID', 'Chebi.Name', 'Chebi.Name_combined'], axis = 1, inplace = True)\n",
    "metabolomics_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "worldwide-scholar",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>date</th>\n",
       "      <th>means</th>\n",
       "      <th>medians</th>\n",
       "      <th>sds</th>\n",
       "      <th>se</th>\n",
       "      <th>ci</th>\n",
       "      <th>type_bnp</th>\n",
       "      <th>type_bp</th>\n",
       "      <th>type_snp</th>\n",
       "      <th>type_sp</th>\n",
       "      <th>known_type_known</th>\n",
       "      <th>known_type_unknown</th>\n",
       "      <th>N_2</th>\n",
       "      <th>N_3</th>\n",
       "      <th>type2_extracellular</th>\n",
       "      <th>type2_intracellular</th>\n",
       "      <th>measurement_nonpolar</th>\n",
       "      <th>measurement_polar</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>1.440122</td>\n",
       "      <td>1.489979</td>\n",
       "      <td>0.146440</td>\n",
       "      <td>0.084547</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>3.778469</td>\n",
       "      <td>3.801417</td>\n",
       "      <td>1.924377</td>\n",
       "      <td>1.111040</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>0.949545</td>\n",
       "      <td>0.947209</td>\n",
       "      <td>0.065422</td>\n",
       "      <td>0.037771</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>2.095116</td>\n",
       "      <td>1.809891</td>\n",
       "      <td>1.916896</td>\n",
       "      <td>1.106721</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>1.631643</td>\n",
       "      <td>1.571125</td>\n",
       "      <td>0.168501</td>\n",
       "      <td>0.097284</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18865</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>1.364105</td>\n",
       "      <td>1.457780</td>\n",
       "      <td>0.170311</td>\n",
       "      <td>0.098329</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18866</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>4.815774</td>\n",
       "      <td>1.160165</td>\n",
       "      <td>6.835474</td>\n",
       "      <td>3.946463</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18867</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>0.697489</td>\n",
       "      <td>0.684339</td>\n",
       "      <td>0.049632</td>\n",
       "      <td>0.028655</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18868</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>0.785996</td>\n",
       "      <td>0.744529</td>\n",
       "      <td>0.139901</td>\n",
       "      <td>0.080772</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18869</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>0.755565</td>\n",
       "      <td>0.709972</td>\n",
       "      <td>0.149214</td>\n",
       "      <td>0.086149</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18870 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            date     means   medians       sds        se        ci  type_bnp  \\\n",
       "0     2011-03-21  1.440122  1.489979  0.146440  0.084547  4.302653         0   \n",
       "1     2011-03-21  3.778469  3.801417  1.924377  1.111040  4.302653         0   \n",
       "2     2011-03-21  0.949545  0.947209  0.065422  0.037771  4.302653         1   \n",
       "3     2011-03-21  2.095116  1.809891  1.916896  1.106721  4.302653         0   \n",
       "4     2011-03-21  1.631643  1.571125  0.168501  0.097284  4.302653         0   \n",
       "...          ...       ...       ...       ...       ...       ...       ...   \n",
       "18865 2012-05-03  1.364105  1.457780  0.170311  0.098329  4.302653         1   \n",
       "18866 2012-05-03  4.815774  1.160165  6.835474  3.946463  4.302653         0   \n",
       "18867 2012-05-03  0.697489  0.684339  0.049632  0.028655  4.302653         0   \n",
       "18868 2012-05-03  0.785996  0.744529  0.139901  0.080772  4.302653         1   \n",
       "18869 2012-05-03  0.755565  0.709972  0.149214  0.086149  4.302653         0   \n",
       "\n",
       "       type_bp  type_snp  type_sp  known_type_known  known_type_unknown  N_2  \\\n",
       "0            0         0        1                 1                   0    0   \n",
       "1            0         0        1                 0                   1    0   \n",
       "2            0         0        0                 0                   1    0   \n",
       "3            1         0        0                 0                   1    0   \n",
       "4            1         0        0                 0                   1    0   \n",
       "...        ...       ...      ...               ...                 ...  ...   \n",
       "18865        0         0        0                 1                   0    0   \n",
       "18866        1         0        0                 0                   1    0   \n",
       "18867        1         0        0                 1                   0    0   \n",
       "18868        0         0        0                 0                   1    0   \n",
       "18869        0         0        1                 1                   0    0   \n",
       "\n",
       "       N_3  type2_extracellular  type2_intracellular  measurement_nonpolar  \\\n",
       "0        1                    1                    0                     0   \n",
       "1        1                    1                    0                     0   \n",
       "2        1                    0                    1                     1   \n",
       "3        1                    0                    1                     0   \n",
       "4        1                    0                    1                     0   \n",
       "...    ...                  ...                  ...                   ...   \n",
       "18865    1                    0                    1                     1   \n",
       "18866    1                    0                    1                     0   \n",
       "18867    1                    0                    1                     0   \n",
       "18868    1                    0                    1                     1   \n",
       "18869    1                    1                    0                     0   \n",
       "\n",
       "       measurement_polar  \n",
       "0                      1  \n",
       "1                      1  \n",
       "2                      0  \n",
       "3                      1  \n",
       "4                      1  \n",
       "...                  ...  \n",
       "18865                  0  \n",
       "18866                  1  \n",
       "18867                  1  \n",
       "18868                  0  \n",
       "18869                  1  \n",
       "\n",
       "[18870 rows x 18 columns]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Dummy eencoding categorical data\n",
    "metabolomics_df = pd.get_dummies(metabolomics_df, columns = ['type', 'known_type', 'N', 'type2', 'measurement'])\n",
    "metabolomics_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "narrow-resource",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>date</th>\n",
       "      <th>means</th>\n",
       "      <th>medians</th>\n",
       "      <th>sds</th>\n",
       "      <th>se</th>\n",
       "      <th>ci</th>\n",
       "      <th>type_bnp</th>\n",
       "      <th>type_bp</th>\n",
       "      <th>type_snp</th>\n",
       "      <th>type_sp</th>\n",
       "      <th>known_type_known</th>\n",
       "      <th>known_type_unknown</th>\n",
       "      <th>N_2</th>\n",
       "      <th>N_3</th>\n",
       "      <th>type2_extracellular</th>\n",
       "      <th>type2_intracellular</th>\n",
       "      <th>measurement_nonpolar</th>\n",
       "      <th>measurement_polar</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>-0.145146</td>\n",
       "      <td>-0.094812</td>\n",
       "      <td>-0.189837</td>\n",
       "      <td>-0.190047</td>\n",
       "      <td>-0.128175</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>0.148026</td>\n",
       "      <td>0.316083</td>\n",
       "      <td>-0.000023</td>\n",
       "      <td>-0.000270</td>\n",
       "      <td>-0.128175</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>-0.206653</td>\n",
       "      <td>-0.191297</td>\n",
       "      <td>-0.198486</td>\n",
       "      <td>-0.198695</td>\n",
       "      <td>-0.128175</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>-0.063026</td>\n",
       "      <td>-0.037942</td>\n",
       "      <td>-0.000822</td>\n",
       "      <td>-0.001068</td>\n",
       "      <td>-0.128175</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>-0.121134</td>\n",
       "      <td>-0.080386</td>\n",
       "      <td>-0.187482</td>\n",
       "      <td>-0.187692</td>\n",
       "      <td>-0.128175</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18865</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>-0.154677</td>\n",
       "      <td>-0.100535</td>\n",
       "      <td>-0.187288</td>\n",
       "      <td>-0.187499</td>\n",
       "      <td>-0.128175</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18866</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>0.278079</td>\n",
       "      <td>-0.153441</td>\n",
       "      <td>0.524288</td>\n",
       "      <td>0.523941</td>\n",
       "      <td>-0.128175</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18867</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>-0.238254</td>\n",
       "      <td>-0.238027</td>\n",
       "      <td>-0.200172</td>\n",
       "      <td>-0.200380</td>\n",
       "      <td>-0.128175</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18868</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>-0.227158</td>\n",
       "      <td>-0.227327</td>\n",
       "      <td>-0.190535</td>\n",
       "      <td>-0.190745</td>\n",
       "      <td>-0.128175</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18869</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>-0.230973</td>\n",
       "      <td>-0.233470</td>\n",
       "      <td>-0.189541</td>\n",
       "      <td>-0.189751</td>\n",
       "      <td>-0.128175</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18870 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            date     means   medians       sds        se        ci  type_bnp  \\\n",
       "0     2011-03-21 -0.145146 -0.094812 -0.189837 -0.190047 -0.128175         0   \n",
       "1     2011-03-21  0.148026  0.316083 -0.000023 -0.000270 -0.128175         0   \n",
       "2     2011-03-21 -0.206653 -0.191297 -0.198486 -0.198695 -0.128175         1   \n",
       "3     2011-03-21 -0.063026 -0.037942 -0.000822 -0.001068 -0.128175         0   \n",
       "4     2011-03-21 -0.121134 -0.080386 -0.187482 -0.187692 -0.128175         0   \n",
       "...          ...       ...       ...       ...       ...       ...       ...   \n",
       "18865 2012-05-03 -0.154677 -0.100535 -0.187288 -0.187499 -0.128175         1   \n",
       "18866 2012-05-03  0.278079 -0.153441  0.524288  0.523941 -0.128175         0   \n",
       "18867 2012-05-03 -0.238254 -0.238027 -0.200172 -0.200380 -0.128175         0   \n",
       "18868 2012-05-03 -0.227158 -0.227327 -0.190535 -0.190745 -0.128175         1   \n",
       "18869 2012-05-03 -0.230973 -0.233470 -0.189541 -0.189751 -0.128175         0   \n",
       "\n",
       "       type_bp  type_snp  type_sp  known_type_known  known_type_unknown  N_2  \\\n",
       "0            0         0        1                 1                   0    0   \n",
       "1            0         0        1                 0                   1    0   \n",
       "2            0         0        0                 0                   1    0   \n",
       "3            1         0        0                 0                   1    0   \n",
       "4            1         0        0                 0                   1    0   \n",
       "...        ...       ...      ...               ...                 ...  ...   \n",
       "18865        0         0        0                 1                   0    0   \n",
       "18866        1         0        0                 0                   1    0   \n",
       "18867        1         0        0                 1                   0    0   \n",
       "18868        0         0        0                 0                   1    0   \n",
       "18869        0         0        1                 1                   0    0   \n",
       "\n",
       "       N_3  type2_extracellular  type2_intracellular  measurement_nonpolar  \\\n",
       "0        1                    1                    0                     0   \n",
       "1        1                    1                    0                     0   \n",
       "2        1                    0                    1                     1   \n",
       "3        1                    0                    1                     0   \n",
       "4        1                    0                    1                     0   \n",
       "...    ...                  ...                  ...                   ...   \n",
       "18865    1                    0                    1                     1   \n",
       "18866    1                    0                    1                     0   \n",
       "18867    1                    0                    1                     0   \n",
       "18868    1                    0                    1                     1   \n",
       "18869    1                    1                    0                     0   \n",
       "\n",
       "       measurement_polar  \n",
       "0                      1  \n",
       "1                      1  \n",
       "2                      0  \n",
       "3                      1  \n",
       "4                      1  \n",
       "...                  ...  \n",
       "18865                  0  \n",
       "18866                  1  \n",
       "18867                  1  \n",
       "18868                  0  \n",
       "18869                  1  \n",
       "\n",
       "[18870 rows x 18 columns]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Standardizing data\n",
    "scaled_metabolomics_df = metabolomics_df\n",
    "metabolomics_scaler = preprocessing.StandardScaler()\n",
    "scaled_metabolomics_df[['means', 'medians', 'sds', 'se', 'ci']] = metabolomics_scaler.fit_transform(metabolomics_df[['means', 'medians', 'sds', 'se', 'ci']])\n",
    "\n",
    "scaled_metabolomics_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acceptable-sender",
   "metadata": {},
   "source": [
    "## Clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ethical-alexander",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Deep learning temporal clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "comfortable-northern",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
