{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "nonprofit-lightning",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import altair as alt\n",
    "import os, random\n",
    "from Bio import SeqIO\n",
    "import gensim \n",
    "from gensim.models import Word2Vec\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "reverse-hampshire",
   "metadata": {},
   "source": [
    "## Defining paths for each and every omic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "sound-benefit",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_root_data = os.path.join ('..', 'Data', 'Extracted', 'Databases')\n",
    "path_all_fasta = os.path.join (path_root_data, 'fasta_files', 'AllBins')\n",
    "path_normalised_metabolomics = os.path.join (path_root_data, 'Metabolomics', 'Normalised_Tables')\n",
    "path_model_save_root = 'Saved_models'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "collectible-tuition",
   "metadata": {},
   "outputs": [],
   "source": [
    "num_of_mags = len([i for i in os.listdir(path_all_fasta) if (i[-2:] == 'fa')])\n",
    "SEED = 42\n",
    "END = 25\n",
    "num_of_workers = 3\n",
    "random.seed(SEED)\n",
    "np.random.seed(SEED)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ranking-reform",
   "metadata": {},
   "source": [
    "---\n",
    "## Importing MAGs\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "alpha-geometry",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function that splits each genome into k-mers thus creating even longer sentence (MAG)\n",
    "# It returns tokenized genome i.e. [kmer, kmer,...]\n",
    "def split_genome (genome, k = 5):\n",
    "    new_genome = []\n",
    "    n = len(genome)\n",
    "    \n",
    "    if n-k <=0:\n",
    "        return genome\n",
    "    else:\n",
    "        for i in range(n-k):\n",
    "            new_genome.append(genome[i:i+k])\n",
    "        \n",
    "        return new_genome\n",
    "\n",
    "\n",
    "def vectorize_one_mag (one_mag, w2v_model):\n",
    "    \n",
    "    # We have to generate vectors for each word in one MAG and then create vector representation of that MAG\n",
    "    # by averaging vectors of its words\n",
    "    zero_vector = np.zeros(w2v_model.vector_size)\n",
    "    word_vectors = []\n",
    "    one_mag_vector = []\n",
    "    \n",
    "    for sentence in one_mag:\n",
    "        for word in sentence:\n",
    "            if word in w2v_model.wv:\n",
    "                try:\n",
    "                    word_vectors.append(w2v_model.wv[word])\n",
    "                except KeyError:\n",
    "                    print ('Key Error')\n",
    "                    continue\n",
    "    \n",
    "    if word_vectors:\n",
    "        word_vectors = np.asarray(word_vectors)\n",
    "        one_mag_vector = word_vectors.mean (axis=0)\n",
    "    \n",
    "    else:\n",
    "        one_mag_vector = zero_vector\n",
    "    \n",
    "    return one_mag_vector\n",
    "\n",
    "\n",
    "# Function that vectorizes a MAG (document) with a pretrained word2vec model. It returns vector representation of a given MAG\n",
    "# Vectorization is done by averaging word (k-mer) vectors for the whole document (MAG)\n",
    "def vectorize_mags (w2v_model, end = 25):\n",
    "    \n",
    "    print ('Vectorizing MAGs')\n",
    "    \n",
    "    fasta_files = [i for i in os.listdir(path_all_fasta) if (i[-2:] == 'fa')]\n",
    "    list_of_mag_vectors = []\n",
    "    \n",
    "    # This was done so that I could work with first 'end' FASTA files only. Otherwise, I should just remove: i, and enumerate\n",
    "    for i, fasta_file_name in enumerate(fasta_files):\n",
    "        \n",
    "        if i == end:\n",
    "            break\n",
    "        \n",
    "        else:\n",
    "            with open(os.path.join(path_all_fasta, fasta_file_name), 'r') as input_file:\n",
    "                \n",
    "                one_mag = []\n",
    "                for fasta_string in SeqIO.parse(input_file, \"fasta\"):\n",
    "                    \n",
    "                    # Get kmers of a genome and create a sentence (list of words)\n",
    "                    temp_kmers = split_genome (str(fasta_string.seq))\n",
    "                    \n",
    "                    # Create a document (list of sentences)\n",
    "                    one_mag.append(temp_kmers)\n",
    "                    \n",
    "                # Vectorizing MAGs one by one\n",
    "                list_of_mag_vectors.append (vectorize_one_mag (one_mag, w2v_model))\n",
    "    \n",
    "    print ('Finished vectorizing')\n",
    "    \n",
    "    return list_of_mag_vectors\n",
    "    \n",
    "    \n",
    "\n",
    "# If one wants to import MAGs in order to vectorize them, one should use start argument in order to skip first 'start' MAGs\n",
    "# If one wants to import MAGs to train word2vec model, one should use only end argument, so that first 'end' MAGs are used for training\n",
    "# Todo: Implement randomisation in picking MAGs for training, and don't use first 'start' MAGs for training\n",
    "# (create list of indexes from 0 to 1364, use sklearn split train test, then traverse directory and use only MAGs with indexes in train for training w2v)\n",
    "def import_mags_and_build_model (end = 25, path_all_fasta = path_all_fasta):\n",
    "    \n",
    "    print ('Importing MAGs')\n",
    "    \n",
    "    # There are 1364 MAGs enclosed in FASTA files\n",
    "    # I have to traverse every FASTA file, and in each file every sequence\n",
    "    \n",
    "    fasta_files = [i for i in os.listdir(path_all_fasta) if (i[-2:] == 'fa')]\n",
    "    \n",
    "    # This was done so that I could work with first 100 FASTA files only. Otherwise, I should just remove: i, and enumerate\n",
    "    for i, fasta_file_name in enumerate(fasta_files):\n",
    "        \n",
    "        if i == end:\n",
    "            break\n",
    "        \n",
    "        else:\n",
    "            with open(os.path.join(path_all_fasta, fasta_file_name), 'r') as input_file:\n",
    "                \n",
    "                one_mag = []\n",
    "                for fasta_string in SeqIO.parse(input_file, \"fasta\"):\n",
    "                    \n",
    "                    # Get kmers of a genome and create a sentence (list of words)\n",
    "                    temp_kmers = split_genome (str(fasta_string.seq))\n",
    "                    \n",
    "                    # Create a document (list of sentences)\n",
    "                    one_mag.append(temp_kmers)\n",
    "                    \n",
    "                # If we do not have a model, we build one\n",
    "                if i == 0:\n",
    "                    print ('Building w2v model')\n",
    "                    # We build our model on the first MAG\n",
    "                    w2v_model = Word2Vec (sentences = one_mag, size = 100, workers = num_of_workers, seed=SEED)\n",
    "                \n",
    "                # Else we just expand its vocabulary\n",
    "                else:\n",
    "                    # Now we expand our vocabulary\n",
    "                    w2v_model.build_vocab (one_mag, update = True)\n",
    "                    \n",
    "    print ('Finished building')\n",
    "    \n",
    "    return w2v_model\n",
    "\n",
    "\n",
    "def train_model (w2v_model, epochs, end = 25):\n",
    "    \n",
    "    print ('Starting model training')\n",
    "    \n",
    "    # There are 1364 MAGs enclosed in FASTA files\n",
    "    # I have to traverse every FASTA file, and in each file every sequence\n",
    "    \n",
    "    fasta_files = [i for i in os.listdir(path_all_fasta) if (i[-2:] == 'fa')]\n",
    "    \n",
    "    # This was done so that I could work with first 100 FASTA files only. Otherwise, I should just remove: i, and enumerate\n",
    "    for i, fasta_file_name in enumerate(fasta_files):\n",
    "        \n",
    "        if i == end:\n",
    "            break\n",
    "        \n",
    "        else:\n",
    "            with open(os.path.join(path_all_fasta, fasta_file_name), 'r') as input_file:\n",
    "                \n",
    "                one_mag = []\n",
    "                for fasta_string in SeqIO.parse(input_file, \"fasta\"):\n",
    "                    \n",
    "                    # Get kmers of a genome and create a sentence (list of words)\n",
    "                    temp_kmers = split_genome (str(fasta_string.seq))\n",
    "                    \n",
    "                    # Create a document (list of sentences)\n",
    "                    one_mag.append(temp_kmers)\n",
    "                    \n",
    "                \n",
    "                w2v_model.train (one_mag, total_examples = w2v_model.corpus_count, epochs = epochs)\n",
    "                    \n",
    "    print ('Model training finished')\n",
    "    \n",
    "    return w2v_model\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "developmental-broadcasting",
   "metadata": {},
   "source": [
    "---\n",
    "## Importing Metabolomic data\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "hollow-macro",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Metabolite</th>\n",
       "      <th>tp</th>\n",
       "      <th>date</th>\n",
       "      <th>type</th>\n",
       "      <th>known_type</th>\n",
       "      <th>means</th>\n",
       "      <th>medians</th>\n",
       "      <th>sds</th>\n",
       "      <th>N</th>\n",
       "      <th>se</th>\n",
       "      <th>ci</th>\n",
       "      <th>type2</th>\n",
       "      <th>measurement</th>\n",
       "      <th>KEGG.Compound.ID</th>\n",
       "      <th>Chebi.Name</th>\n",
       "      <th>Chebi.Name_combined</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Phosphoric_acid_3TMS</td>\n",
       "      <td>D1</td>\n",
       "      <td>2012-04-17</td>\n",
       "      <td>snp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.757113</td>\n",
       "      <td>0.737943</td>\n",
       "      <td>0.098650</td>\n",
       "      <td>3</td>\n",
       "      <td>0.056956</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>C00009</td>\n",
       "      <td>phosphoric acid</td>\n",
       "      <td>phosphoric acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Phosphoric_acid_3TMS</td>\n",
       "      <td>D10</td>\n",
       "      <td>2012-01-25</td>\n",
       "      <td>snp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.720935</td>\n",
       "      <td>0.873509</td>\n",
       "      <td>0.266593</td>\n",
       "      <td>3</td>\n",
       "      <td>0.153918</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>C00009</td>\n",
       "      <td>phosphoric acid</td>\n",
       "      <td>phosphoric acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Phosphoric_acid_3TMS</td>\n",
       "      <td>D11</td>\n",
       "      <td>2012-03-22</td>\n",
       "      <td>snp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.625047</td>\n",
       "      <td>0.652268</td>\n",
       "      <td>0.057820</td>\n",
       "      <td>3</td>\n",
       "      <td>0.033382</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>C00009</td>\n",
       "      <td>phosphoric acid</td>\n",
       "      <td>phosphoric acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Phosphoric_acid_3TMS</td>\n",
       "      <td>D12</td>\n",
       "      <td>2012-01-19</td>\n",
       "      <td>snp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.964956</td>\n",
       "      <td>0.941878</td>\n",
       "      <td>0.053754</td>\n",
       "      <td>3</td>\n",
       "      <td>0.031035</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>C00009</td>\n",
       "      <td>phosphoric acid</td>\n",
       "      <td>phosphoric acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>!Phosphoric_acid_3TMS</td>\n",
       "      <td>D13</td>\n",
       "      <td>2011-05-13</td>\n",
       "      <td>snp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.963984</td>\n",
       "      <td>0.974479</td>\n",
       "      <td>0.071659</td>\n",
       "      <td>3</td>\n",
       "      <td>0.041373</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>C00009</td>\n",
       "      <td>phosphoric acid</td>\n",
       "      <td>phosphoric acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18865</th>\n",
       "      <td>Valine_2TMS</td>\n",
       "      <td>D51</td>\n",
       "      <td>2011-08-11</td>\n",
       "      <td>bp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.704821</td>\n",
       "      <td>0.478251</td>\n",
       "      <td>0.428368</td>\n",
       "      <td>3</td>\n",
       "      <td>0.247319</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C16436,C06417,C00183</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18866</th>\n",
       "      <td>Valine_2TMS</td>\n",
       "      <td>D6</td>\n",
       "      <td>2011-06-03</td>\n",
       "      <td>bp</td>\n",
       "      <td>known</td>\n",
       "      <td>1.085888</td>\n",
       "      <td>1.197471</td>\n",
       "      <td>0.487484</td>\n",
       "      <td>3</td>\n",
       "      <td>0.281449</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C16436,C06417,C00183</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18867</th>\n",
       "      <td>Valine_2TMS</td>\n",
       "      <td>D7</td>\n",
       "      <td>2012-04-04</td>\n",
       "      <td>bp</td>\n",
       "      <td>known</td>\n",
       "      <td>2.958023</td>\n",
       "      <td>2.813145</td>\n",
       "      <td>0.341213</td>\n",
       "      <td>3</td>\n",
       "      <td>0.197000</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C16436,C06417,C00183</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18868</th>\n",
       "      <td>Valine_2TMS</td>\n",
       "      <td>D8</td>\n",
       "      <td>2012-02-08</td>\n",
       "      <td>bp</td>\n",
       "      <td>known</td>\n",
       "      <td>2.515980</td>\n",
       "      <td>2.322248</td>\n",
       "      <td>0.618556</td>\n",
       "      <td>3</td>\n",
       "      <td>0.357123</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C16436,C06417,C00183</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18869</th>\n",
       "      <td>Valine_2TMS</td>\n",
       "      <td>D9</td>\n",
       "      <td>2011-12-28</td>\n",
       "      <td>bp</td>\n",
       "      <td>known</td>\n",
       "      <td>2.356989</td>\n",
       "      <td>2.280757</td>\n",
       "      <td>0.264918</td>\n",
       "      <td>3</td>\n",
       "      <td>0.152950</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C16436,C06417,C00183</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18870 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Metabolite   tp        date type known_type     means  \\\n",
       "0      !Phosphoric_acid_3TMS   D1  2012-04-17  snp      known  0.757113   \n",
       "1      !Phosphoric_acid_3TMS  D10  2012-01-25  snp      known  0.720935   \n",
       "2      !Phosphoric_acid_3TMS  D11  2012-03-22  snp      known  0.625047   \n",
       "3      !Phosphoric_acid_3TMS  D12  2012-01-19  snp      known  0.964956   \n",
       "4      !Phosphoric_acid_3TMS  D13  2011-05-13  snp      known  0.963984   \n",
       "...                      ...  ...         ...  ...        ...       ...   \n",
       "18865            Valine_2TMS  D51  2011-08-11   bp      known  0.704821   \n",
       "18866            Valine_2TMS   D6  2011-06-03   bp      known  1.085888   \n",
       "18867            Valine_2TMS   D7  2012-04-04   bp      known  2.958023   \n",
       "18868            Valine_2TMS   D8  2012-02-08   bp      known  2.515980   \n",
       "18869            Valine_2TMS   D9  2011-12-28   bp      known  2.356989   \n",
       "\n",
       "        medians       sds  N        se        ci          type2 measurement  \\\n",
       "0      0.737943  0.098650  3  0.056956  4.302653  extracellular    nonpolar   \n",
       "1      0.873509  0.266593  3  0.153918  4.302653  extracellular    nonpolar   \n",
       "2      0.652268  0.057820  3  0.033382  4.302653  extracellular    nonpolar   \n",
       "3      0.941878  0.053754  3  0.031035  4.302653  extracellular    nonpolar   \n",
       "4      0.974479  0.071659  3  0.041373  4.302653  extracellular    nonpolar   \n",
       "...         ...       ... ..       ...       ...            ...         ...   \n",
       "18865  0.478251  0.428368  3  0.247319  4.302653  intracellular       polar   \n",
       "18866  1.197471  0.487484  3  0.281449  4.302653  intracellular       polar   \n",
       "18867  2.813145  0.341213  3  0.197000  4.302653  intracellular       polar   \n",
       "18868  2.322248  0.618556  3  0.357123  4.302653  intracellular       polar   \n",
       "18869  2.280757  0.264918  3  0.152950  4.302653  intracellular       polar   \n",
       "\n",
       "           KEGG.Compound.ID                Chebi.Name  \\\n",
       "0                    C00009           phosphoric acid   \n",
       "1                    C00009           phosphoric acid   \n",
       "2                    C00009           phosphoric acid   \n",
       "3                    C00009           phosphoric acid   \n",
       "4                    C00009           phosphoric acid   \n",
       "...                     ...                       ...   \n",
       "18865  C16436,C06417,C00183  valine;D-valine;L-valine   \n",
       "18866  C16436,C06417,C00183  valine;D-valine;L-valine   \n",
       "18867  C16436,C06417,C00183  valine;D-valine;L-valine   \n",
       "18868  C16436,C06417,C00183  valine;D-valine;L-valine   \n",
       "18869  C16436,C06417,C00183  valine;D-valine;L-valine   \n",
       "\n",
       "            Chebi.Name_combined  \n",
       "0               phosphoric acid  \n",
       "1               phosphoric acid  \n",
       "2               phosphoric acid  \n",
       "3               phosphoric acid  \n",
       "4               phosphoric acid  \n",
       "...                         ...  \n",
       "18865  valine;D-valine;L-valine  \n",
       "18866  valine;D-valine;L-valine  \n",
       "18867  valine;D-valine;L-valine  \n",
       "18868  valine;D-valine;L-valine  \n",
       "18869  valine;D-valine;L-valine  \n",
       "\n",
       "[18870 rows x 16 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metabolomics_file_name = os.path.join(path_normalised_metabolomics, os.listdir(path_normalised_metabolomics)[0])\n",
    "metabolomics_df = pd.read_csv (metabolomics_file_name, delimiter = '\\t')\n",
    "metabolomics_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "furnished-bandwidth",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>date</th>\n",
       "      <th>Metabolite</th>\n",
       "      <th>tp</th>\n",
       "      <th>type</th>\n",
       "      <th>known_type</th>\n",
       "      <th>means</th>\n",
       "      <th>medians</th>\n",
       "      <th>sds</th>\n",
       "      <th>N</th>\n",
       "      <th>se</th>\n",
       "      <th>ci</th>\n",
       "      <th>type2</th>\n",
       "      <th>measurement</th>\n",
       "      <th>KEGG.Compound.ID</th>\n",
       "      <th>Chebi.Name</th>\n",
       "      <th>Chebi.Name_combined</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>Leucine_2TMS</td>\n",
       "      <td>D37</td>\n",
       "      <td>sp</td>\n",
       "      <td>known</td>\n",
       "      <td>1.440122</td>\n",
       "      <td>1.489979</td>\n",
       "      <td>0.146440</td>\n",
       "      <td>3</td>\n",
       "      <td>0.084547</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C16439,C00123,C01570</td>\n",
       "      <td>leucine;L-leucine;D-leucine</td>\n",
       "      <td>leucine;L-leucine;D-leucine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>No match: 1181.78_EM_SP_D1_1_223</td>\n",
       "      <td>D37</td>\n",
       "      <td>sp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>3.778469</td>\n",
       "      <td>3.801417</td>\n",
       "      <td>1.924377</td>\n",
       "      <td>3</td>\n",
       "      <td>1.111040</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>unknown#emu_WW_2144.25</td>\n",
       "      <td>D37</td>\n",
       "      <td>bnp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.949545</td>\n",
       "      <td>0.947209</td>\n",
       "      <td>0.065422</td>\n",
       "      <td>3</td>\n",
       "      <td>0.037771</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>Unknown#bth_pae_001</td>\n",
       "      <td>D37</td>\n",
       "      <td>bp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2.095116</td>\n",
       "      <td>1.809891</td>\n",
       "      <td>1.916896</td>\n",
       "      <td>3</td>\n",
       "      <td>1.106721</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>No match: 2220.34_EM_BP_D1_1_192</td>\n",
       "      <td>D37</td>\n",
       "      <td>bp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.631643</td>\n",
       "      <td>1.571125</td>\n",
       "      <td>0.168501</td>\n",
       "      <td>3</td>\n",
       "      <td>0.097284</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18865</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>Octadecanol_1TMS</td>\n",
       "      <td>D48</td>\n",
       "      <td>bnp</td>\n",
       "      <td>known</td>\n",
       "      <td>1.364105</td>\n",
       "      <td>1.457780</td>\n",
       "      <td>0.170311</td>\n",
       "      <td>3</td>\n",
       "      <td>0.098329</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>D01924</td>\n",
       "      <td>octadecan-1-ol</td>\n",
       "      <td>octadecan-1-ol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18866</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>No match: 1447.93_EM_BP_D2_1_57</td>\n",
       "      <td>D48</td>\n",
       "      <td>bp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>4.815774</td>\n",
       "      <td>1.160165</td>\n",
       "      <td>6.835474</td>\n",
       "      <td>3</td>\n",
       "      <td>3.946463</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18867</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>Lactulose_8TMS</td>\n",
       "      <td>D48</td>\n",
       "      <td>bp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.697489</td>\n",
       "      <td>0.684339</td>\n",
       "      <td>0.049632</td>\n",
       "      <td>3</td>\n",
       "      <td>0.028655</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C07064</td>\n",
       "      <td>lactulose</td>\n",
       "      <td>lactulose</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18868</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>No match: 2004.78_EM_BNP_D10_1_205</td>\n",
       "      <td>D48</td>\n",
       "      <td>bnp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.785996</td>\n",
       "      <td>0.744529</td>\n",
       "      <td>0.139901</td>\n",
       "      <td>3</td>\n",
       "      <td>0.080772</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18869</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>2-Oxobutanoic_acid_1_1MeOX_1TMS</td>\n",
       "      <td>D48</td>\n",
       "      <td>sp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.755565</td>\n",
       "      <td>0.709972</td>\n",
       "      <td>0.149214</td>\n",
       "      <td>3</td>\n",
       "      <td>0.086149</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C00109</td>\n",
       "      <td>2-oxobutanoic acid</td>\n",
       "      <td>2-oxobutanoic acid</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18870 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            date                          Metabolite   tp type known_type  \\\n",
       "0     2011-03-21                        Leucine_2TMS  D37   sp      known   \n",
       "1     2011-03-21    No match: 1181.78_EM_SP_D1_1_223  D37   sp    unknown   \n",
       "2     2011-03-21              unknown#emu_WW_2144.25  D37  bnp    unknown   \n",
       "3     2011-03-21                 Unknown#bth_pae_001  D37   bp    unknown   \n",
       "4     2011-03-21    No match: 2220.34_EM_BP_D1_1_192  D37   bp    unknown   \n",
       "...          ...                                 ...  ...  ...        ...   \n",
       "18865 2012-05-03                    Octadecanol_1TMS  D48  bnp      known   \n",
       "18866 2012-05-03     No match: 1447.93_EM_BP_D2_1_57  D48   bp    unknown   \n",
       "18867 2012-05-03                      Lactulose_8TMS  D48   bp      known   \n",
       "18868 2012-05-03  No match: 2004.78_EM_BNP_D10_1_205  D48  bnp    unknown   \n",
       "18869 2012-05-03     2-Oxobutanoic_acid_1_1MeOX_1TMS  D48   sp      known   \n",
       "\n",
       "          means   medians       sds  N        se        ci          type2  \\\n",
       "0      1.440122  1.489979  0.146440  3  0.084547  4.302653  extracellular   \n",
       "1      3.778469  3.801417  1.924377  3  1.111040  4.302653  extracellular   \n",
       "2      0.949545  0.947209  0.065422  3  0.037771  4.302653  intracellular   \n",
       "3      2.095116  1.809891  1.916896  3  1.106721  4.302653  intracellular   \n",
       "4      1.631643  1.571125  0.168501  3  0.097284  4.302653  intracellular   \n",
       "...         ...       ...       ... ..       ...       ...            ...   \n",
       "18865  1.364105  1.457780  0.170311  3  0.098329  4.302653  intracellular   \n",
       "18866  4.815774  1.160165  6.835474  3  3.946463  4.302653  intracellular   \n",
       "18867  0.697489  0.684339  0.049632  3  0.028655  4.302653  intracellular   \n",
       "18868  0.785996  0.744529  0.139901  3  0.080772  4.302653  intracellular   \n",
       "18869  0.755565  0.709972  0.149214  3  0.086149  4.302653  extracellular   \n",
       "\n",
       "      measurement      KEGG.Compound.ID                   Chebi.Name  \\\n",
       "0           polar  C16439,C00123,C01570  leucine;L-leucine;D-leucine   \n",
       "1           polar                   NaN                          NaN   \n",
       "2        nonpolar                   NaN                          NaN   \n",
       "3           polar                   NaN                          NaN   \n",
       "4           polar                   NaN                          NaN   \n",
       "...           ...                   ...                          ...   \n",
       "18865    nonpolar                D01924               octadecan-1-ol   \n",
       "18866       polar                   NaN                          NaN   \n",
       "18867       polar                C07064                    lactulose   \n",
       "18868    nonpolar                   NaN                          NaN   \n",
       "18869       polar                C00109           2-oxobutanoic acid   \n",
       "\n",
       "               Chebi.Name_combined  \n",
       "0      leucine;L-leucine;D-leucine  \n",
       "1                              NaN  \n",
       "2                              NaN  \n",
       "3                              NaN  \n",
       "4                              NaN  \n",
       "...                            ...  \n",
       "18865               octadecan-1-ol  \n",
       "18866                          NaN  \n",
       "18867                    lactulose  \n",
       "18868                          NaN  \n",
       "18869           2-oxobutanoic acid  \n",
       "\n",
       "[18870 rows x 16 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metabolomics_df['date'] = pd.to_datetime(metabolomics_df['date'])\n",
    "metabolomics_df.insert (0, 'date', metabolomics_df.pop('date'))\n",
    "metabolomics_df.sort_values ('date', inplace = True, ignore_index = True)\n",
    "metabolomics_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "supported-phone",
   "metadata": {},
   "source": [
    "## MAG examination"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "sticky-stanford",
   "metadata": {},
   "source": [
    "---\n",
    "Important info:\n",
    "\n",
    "> We can look at metagenome as a sentence which consists of words (that are represented with genomes). Therefore, we can create metagenomes as strings where words are seperated with space character (' '). This is one way of tackling this problem"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "initial-rates",
   "metadata": {},
   "source": [
    "To do:\n",
    "1. Make kmers of some length, out of every word in every sentence\n",
    "2. Encode every sentence with Word2Vec and CBOW (because it's faster than skip-gram)\n",
    "3. Reduce dimensionality with PCA or with autoencoder\n",
    "    > Maybe it is good to do the first 3 steps while importing, for each and every sentence. That way, I might use less memory\n",
    "4. Cluster similar sentences (MAGs) to get rMAGs\n",
    "\n",
    "> Important: this is not temporal data, since there are no timestamps!\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "coordinate-local",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing MAGs\n",
      "Building w2v model\n",
      "Finished building\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# FOR CLUSTERING I SHOULD CREATE A DATAFRAME WITH MAGs INDEXES AND THEIR VECTOR REPRESENTATIONS\n",
    "\n",
    "final_model = import_mags_and_build_model (end = END, path_all_fasta = path_all_fasta)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "funny-workplace",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('CCGGC', 0.5800316333770752),\n",
       " ('TCGGC', 0.5560053586959839),\n",
       " ('ACGGA', 0.5357427597045898),\n",
       " ('GCGGC', 0.5319786667823792),\n",
       " ('ACGGG', 0.5190925598144531),\n",
       " ('ACGGT', 0.4692608118057251),\n",
       " ('CTACG', 0.30135977268218994),\n",
       " ('ACGAC', 0.2825249433517456),\n",
       " ('TACGA', 0.2640616297721863),\n",
       " ('AACGA', 0.2519393265247345)]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_model.wv.most_similar('ACGGC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "peaceful-bracket",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting model training\n",
      "Model training finished\n"
     ]
    }
   ],
   "source": [
    "# Train model\n",
    "epochs = 10\n",
    "\n",
    "final_model = train_model (final_model, epochs = epochs, end = END)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "fitted-trust",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_model.wv.save_word2vec_format(os.path.join (path_model_save_root, 'model_25.bin'), binary=True) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "realistic-tuning",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "powerful-trader",
   "metadata": {},
   "source": [
    "Now I should vectorize documents with this model. For further use, I could save this model's weights, and use it to vectorize all mags. That would take a lot, but every MAG will have its vector representation\n",
    "> This could be done by importing one MAG at a time, then tokenizing it (like before), then getting vector representations of that MAG's sentences (genomes) and then finding the vector representation of the whole MAG (document). If I do that for one MAG at a time, There is no need to worry about memory\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "thrown-coordinator",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Vectorizing MAGs\n"
     ]
    }
   ],
   "source": [
    "list_of_mag_vectors = vectorize_mags (final_model, end = END)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "moved-market",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[array([ 1.2695041 , -1.3016177 ,  0.65977436,  0.80642366,  1.7741423 ,\n",
       "        -0.34343222, -0.7791513 , -0.21512581, -0.18608665, -1.4746015 ,\n",
       "        -0.83035123, -0.11485589,  0.8619527 ,  0.07787611, -1.0242358 ,\n",
       "        -0.85770607, -1.024094  ,  0.46239102,  2.9101133 ,  0.3759784 ,\n",
       "        -0.42531765,  1.3783201 , -0.08274465, -0.16104594, -1.5054408 ,\n",
       "        -1.2894931 , -0.99346983,  0.01654907,  0.80869704, -1.9024653 ,\n",
       "         1.0351242 ,  0.71593   ,  1.1183119 , -0.6273121 , -1.2179893 ,\n",
       "         0.6693737 ,  1.6966379 ,  0.66082835,  0.5160004 ,  0.3739503 ,\n",
       "         1.4906927 ,  1.2522779 ,  0.9794797 , -0.5130619 , -0.5718369 ,\n",
       "         1.9109504 ,  1.1832289 , -1.4082754 , -1.2452774 , -1.1157957 ,\n",
       "        -0.54767996,  1.4872566 , -0.76184076, -1.5441176 , -2.2053912 ,\n",
       "        -0.8439441 ,  0.82276267,  1.8204472 ,  0.45814872,  0.12408362,\n",
       "        -1.2742198 , -0.01376653,  0.1029577 , -2.7019987 ,  1.4996148 ,\n",
       "        -0.6873514 , -0.39449534,  2.92373   ,  0.7501656 , -1.3751278 ,\n",
       "         1.8498234 ,  0.01898357,  0.8074912 ,  0.8665609 ,  0.82386124,\n",
       "        -0.02354039,  1.772538  ,  0.13172135, -0.17890362,  0.18128195,\n",
       "        -0.41221955, -1.2307491 , -2.1118433 , -0.20580591,  0.5122297 ,\n",
       "        -1.1426629 ,  0.19766866,  0.03300591, -0.16566567,  0.19262098,\n",
       "        -1.6931012 , -1.6358544 , -1.9607099 ,  0.33910325, -0.02559898,\n",
       "         0.2606834 ,  0.16334978,  1.2534832 ,  0.5585448 , -1.4541376 ],\n",
       "       dtype=float32),\n",
       " array([-1.2193525 ,  0.8589669 , -2.20915   , -0.58468676, -1.2852927 ,\n",
       "         0.30707875, -0.01081399,  1.042626  , -0.40956217,  1.6552488 ,\n",
       "         2.1096427 , -0.20100251, -1.3928843 , -0.34211087, -0.04543225,\n",
       "         0.03091494,  0.9814241 ,  1.1176212 , -1.9237609 , -0.436577  ,\n",
       "        -0.44206476, -2.0611515 , -0.16409905,  1.255576  ,  1.4155343 ,\n",
       "        -0.93562317,  0.83154494, -0.7217269 , -1.131977  ,  0.00415279,\n",
       "         0.05433305, -1.3424077 ,  0.1904471 ,  0.52477396,  1.5866417 ,\n",
       "        -1.2858078 , -2.0590494 , -0.9300009 , -1.0959525 ,  0.0588738 ,\n",
       "        -1.304373  , -1.4754282 , -0.0942607 , -0.29581097,  1.010519  ,\n",
       "        -2.136267  , -1.1964653 ,  1.4954319 ,  1.1682458 , -0.31216508,\n",
       "         0.16609518,  1.1967148 ,  1.2930214 ,  1.4752198 ,  1.6045095 ,\n",
       "         0.8923253 ,  0.47111955, -0.94730425, -0.57773286,  0.41931093,\n",
       "         1.1543385 , -0.48197785, -0.980275  ,  0.9405091 , -1.6298665 ,\n",
       "         1.6535296 , -0.06740531, -0.78906554, -0.12881367, -0.12768953,\n",
       "        -1.541576  ,  1.1664783 , -0.97644866,  0.6305173 , -0.17012301,\n",
       "        -0.61246717, -2.1187108 ,  0.3764007 , -0.27338734, -0.8108769 ,\n",
       "        -0.4114229 ,  1.7671143 ,  1.5077313 ,  1.8906677 ,  0.33942553,\n",
       "        -0.7720714 ,  0.8053655 , -1.8276353 ,  0.6971252 , -0.64094937,\n",
       "         1.1117618 ,  1.085674  ,  1.1863314 , -1.5794499 , -0.3555805 ,\n",
       "         1.3301547 ,  0.14252223,  0.4145582 ,  2.009863  ,  0.8349676 ],\n",
       "       dtype=float32)]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_of_mag_vectors[0:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "driving-engineering",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>90</th>\n",
       "      <th>91</th>\n",
       "      <th>92</th>\n",
       "      <th>93</th>\n",
       "      <th>94</th>\n",
       "      <th>95</th>\n",
       "      <th>96</th>\n",
       "      <th>97</th>\n",
       "      <th>98</th>\n",
       "      <th>99</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.269504</td>\n",
       "      <td>-1.301618</td>\n",
       "      <td>0.659774</td>\n",
       "      <td>0.806424</td>\n",
       "      <td>1.774142</td>\n",
       "      <td>-0.343432</td>\n",
       "      <td>-0.779151</td>\n",
       "      <td>-0.215126</td>\n",
       "      <td>-0.186087</td>\n",
       "      <td>-1.474602</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.693101</td>\n",
       "      <td>-1.635854</td>\n",
       "      <td>-1.960710</td>\n",
       "      <td>0.339103</td>\n",
       "      <td>-0.025599</td>\n",
       "      <td>0.260683</td>\n",
       "      <td>0.163350</td>\n",
       "      <td>1.253483</td>\n",
       "      <td>0.558545</td>\n",
       "      <td>-1.454138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-1.219352</td>\n",
       "      <td>0.858967</td>\n",
       "      <td>-2.209150</td>\n",
       "      <td>-0.584687</td>\n",
       "      <td>-1.285293</td>\n",
       "      <td>0.307079</td>\n",
       "      <td>-0.010814</td>\n",
       "      <td>1.042626</td>\n",
       "      <td>-0.409562</td>\n",
       "      <td>1.655249</td>\n",
       "      <td>...</td>\n",
       "      <td>1.111762</td>\n",
       "      <td>1.085674</td>\n",
       "      <td>1.186331</td>\n",
       "      <td>-1.579450</td>\n",
       "      <td>-0.355581</td>\n",
       "      <td>1.330155</td>\n",
       "      <td>0.142522</td>\n",
       "      <td>0.414558</td>\n",
       "      <td>2.009863</td>\n",
       "      <td>0.834968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.455337</td>\n",
       "      <td>-0.348460</td>\n",
       "      <td>2.492809</td>\n",
       "      <td>-0.002663</td>\n",
       "      <td>1.328323</td>\n",
       "      <td>-0.874232</td>\n",
       "      <td>-0.157595</td>\n",
       "      <td>0.838077</td>\n",
       "      <td>0.904898</td>\n",
       "      <td>-1.876874</td>\n",
       "      <td>...</td>\n",
       "      <td>-3.277988</td>\n",
       "      <td>-1.223678</td>\n",
       "      <td>-2.357131</td>\n",
       "      <td>0.819557</td>\n",
       "      <td>-0.908406</td>\n",
       "      <td>0.079156</td>\n",
       "      <td>1.186641</td>\n",
       "      <td>0.651305</td>\n",
       "      <td>-0.033716</td>\n",
       "      <td>-1.122361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-1.128447</td>\n",
       "      <td>0.772527</td>\n",
       "      <td>-2.491378</td>\n",
       "      <td>-0.651029</td>\n",
       "      <td>-1.623185</td>\n",
       "      <td>0.663864</td>\n",
       "      <td>0.125967</td>\n",
       "      <td>1.104853</td>\n",
       "      <td>-0.519641</td>\n",
       "      <td>2.120744</td>\n",
       "      <td>...</td>\n",
       "      <td>1.073344</td>\n",
       "      <td>1.364996</td>\n",
       "      <td>1.529100</td>\n",
       "      <td>-1.807320</td>\n",
       "      <td>-0.537589</td>\n",
       "      <td>1.110650</td>\n",
       "      <td>-0.042950</td>\n",
       "      <td>0.446592</td>\n",
       "      <td>1.589758</td>\n",
       "      <td>0.847480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-1.036939</td>\n",
       "      <td>0.809364</td>\n",
       "      <td>-2.464699</td>\n",
       "      <td>-0.687017</td>\n",
       "      <td>-1.576593</td>\n",
       "      <td>0.638427</td>\n",
       "      <td>0.118440</td>\n",
       "      <td>1.119194</td>\n",
       "      <td>-0.525833</td>\n",
       "      <td>2.067393</td>\n",
       "      <td>...</td>\n",
       "      <td>1.060543</td>\n",
       "      <td>1.332533</td>\n",
       "      <td>1.429077</td>\n",
       "      <td>-1.721097</td>\n",
       "      <td>-0.570125</td>\n",
       "      <td>1.051956</td>\n",
       "      <td>-0.089572</td>\n",
       "      <td>0.433013</td>\n",
       "      <td>1.584785</td>\n",
       "      <td>0.764304</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>-0.547059</td>\n",
       "      <td>0.894858</td>\n",
       "      <td>-1.755465</td>\n",
       "      <td>-0.490597</td>\n",
       "      <td>-1.380145</td>\n",
       "      <td>0.359310</td>\n",
       "      <td>-0.060178</td>\n",
       "      <td>0.771358</td>\n",
       "      <td>-0.472385</td>\n",
       "      <td>1.670432</td>\n",
       "      <td>...</td>\n",
       "      <td>0.283317</td>\n",
       "      <td>1.162495</td>\n",
       "      <td>1.423480</td>\n",
       "      <td>-1.511162</td>\n",
       "      <td>-0.741800</td>\n",
       "      <td>0.205500</td>\n",
       "      <td>0.125913</td>\n",
       "      <td>0.088043</td>\n",
       "      <td>0.769423</td>\n",
       "      <td>0.282571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1.733710</td>\n",
       "      <td>-0.566986</td>\n",
       "      <td>2.949093</td>\n",
       "      <td>0.390996</td>\n",
       "      <td>2.418676</td>\n",
       "      <td>-1.398353</td>\n",
       "      <td>-0.696868</td>\n",
       "      <td>0.764553</td>\n",
       "      <td>1.415131</td>\n",
       "      <td>-3.074588</td>\n",
       "      <td>...</td>\n",
       "      <td>-3.647666</td>\n",
       "      <td>-2.091958</td>\n",
       "      <td>-3.698048</td>\n",
       "      <td>1.739185</td>\n",
       "      <td>-0.828084</td>\n",
       "      <td>0.024867</td>\n",
       "      <td>1.245571</td>\n",
       "      <td>0.942139</td>\n",
       "      <td>-0.019091</td>\n",
       "      <td>-1.623251</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1.902613</td>\n",
       "      <td>-1.295641</td>\n",
       "      <td>2.932941</td>\n",
       "      <td>0.292558</td>\n",
       "      <td>2.542777</td>\n",
       "      <td>-0.784014</td>\n",
       "      <td>-0.564508</td>\n",
       "      <td>0.691168</td>\n",
       "      <td>0.872062</td>\n",
       "      <td>-2.692172</td>\n",
       "      <td>...</td>\n",
       "      <td>-3.710976</td>\n",
       "      <td>-2.149990</td>\n",
       "      <td>-3.701959</td>\n",
       "      <td>1.104722</td>\n",
       "      <td>-0.588351</td>\n",
       "      <td>0.193837</td>\n",
       "      <td>1.070350</td>\n",
       "      <td>1.491803</td>\n",
       "      <td>-0.170504</td>\n",
       "      <td>-1.562872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.986385</td>\n",
       "      <td>-0.628855</td>\n",
       "      <td>0.235144</td>\n",
       "      <td>0.294113</td>\n",
       "      <td>0.755645</td>\n",
       "      <td>-0.287721</td>\n",
       "      <td>-0.565301</td>\n",
       "      <td>-0.137433</td>\n",
       "      <td>-0.318865</td>\n",
       "      <td>-0.364859</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.473106</td>\n",
       "      <td>-0.788323</td>\n",
       "      <td>-0.883935</td>\n",
       "      <td>-0.450523</td>\n",
       "      <td>-0.389556</td>\n",
       "      <td>-0.204728</td>\n",
       "      <td>0.018448</td>\n",
       "      <td>0.864524</td>\n",
       "      <td>0.220383</td>\n",
       "      <td>-1.094242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0.685016</td>\n",
       "      <td>-0.085557</td>\n",
       "      <td>-0.214265</td>\n",
       "      <td>0.292677</td>\n",
       "      <td>0.134052</td>\n",
       "      <td>-0.131870</td>\n",
       "      <td>-0.285285</td>\n",
       "      <td>0.141165</td>\n",
       "      <td>-0.274307</td>\n",
       "      <td>-0.183934</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.833542</td>\n",
       "      <td>-0.349754</td>\n",
       "      <td>-0.144961</td>\n",
       "      <td>-0.388498</td>\n",
       "      <td>-0.842064</td>\n",
       "      <td>-0.071577</td>\n",
       "      <td>0.339474</td>\n",
       "      <td>0.374968</td>\n",
       "      <td>0.558935</td>\n",
       "      <td>-0.894052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>0.701596</td>\n",
       "      <td>-0.450739</td>\n",
       "      <td>1.009886</td>\n",
       "      <td>0.350183</td>\n",
       "      <td>0.902692</td>\n",
       "      <td>-0.374130</td>\n",
       "      <td>-0.330576</td>\n",
       "      <td>0.256810</td>\n",
       "      <td>0.090173</td>\n",
       "      <td>-0.911043</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.842188</td>\n",
       "      <td>-0.739546</td>\n",
       "      <td>-0.848902</td>\n",
       "      <td>-0.062544</td>\n",
       "      <td>-0.576259</td>\n",
       "      <td>-0.004731</td>\n",
       "      <td>0.499465</td>\n",
       "      <td>0.434306</td>\n",
       "      <td>0.375023</td>\n",
       "      <td>-0.781382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1.683621</td>\n",
       "      <td>-0.727743</td>\n",
       "      <td>0.919457</td>\n",
       "      <td>0.141144</td>\n",
       "      <td>1.054881</td>\n",
       "      <td>-0.339080</td>\n",
       "      <td>-0.291974</td>\n",
       "      <td>0.352025</td>\n",
       "      <td>0.022157</td>\n",
       "      <td>-1.230969</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.081283</td>\n",
       "      <td>-1.231950</td>\n",
       "      <td>-1.862873</td>\n",
       "      <td>0.450388</td>\n",
       "      <td>-0.862843</td>\n",
       "      <td>0.106902</td>\n",
       "      <td>0.379381</td>\n",
       "      <td>0.846408</td>\n",
       "      <td>-0.022321</td>\n",
       "      <td>-1.240623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>0.466126</td>\n",
       "      <td>0.101711</td>\n",
       "      <td>-0.077127</td>\n",
       "      <td>0.155005</td>\n",
       "      <td>0.125790</td>\n",
       "      <td>-0.041057</td>\n",
       "      <td>-0.276507</td>\n",
       "      <td>0.292252</td>\n",
       "      <td>-0.257001</td>\n",
       "      <td>-0.115536</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.799583</td>\n",
       "      <td>-0.130002</td>\n",
       "      <td>-0.045846</td>\n",
       "      <td>-0.405901</td>\n",
       "      <td>-0.820496</td>\n",
       "      <td>-0.442367</td>\n",
       "      <td>0.338477</td>\n",
       "      <td>0.264230</td>\n",
       "      <td>0.435244</td>\n",
       "      <td>-0.790947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>-0.759412</td>\n",
       "      <td>1.024231</td>\n",
       "      <td>-2.187070</td>\n",
       "      <td>-0.540236</td>\n",
       "      <td>-1.619746</td>\n",
       "      <td>0.164110</td>\n",
       "      <td>-0.068278</td>\n",
       "      <td>0.645948</td>\n",
       "      <td>-0.407626</td>\n",
       "      <td>1.856634</td>\n",
       "      <td>...</td>\n",
       "      <td>0.856082</td>\n",
       "      <td>1.134467</td>\n",
       "      <td>1.595821</td>\n",
       "      <td>-1.509817</td>\n",
       "      <td>-0.782409</td>\n",
       "      <td>0.228485</td>\n",
       "      <td>-0.154000</td>\n",
       "      <td>-0.065237</td>\n",
       "      <td>1.118699</td>\n",
       "      <td>0.502978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>-0.754947</td>\n",
       "      <td>1.020849</td>\n",
       "      <td>-2.226209</td>\n",
       "      <td>-0.547058</td>\n",
       "      <td>-1.670198</td>\n",
       "      <td>0.195537</td>\n",
       "      <td>-0.056791</td>\n",
       "      <td>0.645053</td>\n",
       "      <td>-0.396648</td>\n",
       "      <td>1.872752</td>\n",
       "      <td>...</td>\n",
       "      <td>0.898750</td>\n",
       "      <td>1.148711</td>\n",
       "      <td>1.644138</td>\n",
       "      <td>-1.553703</td>\n",
       "      <td>-0.814907</td>\n",
       "      <td>0.271635</td>\n",
       "      <td>-0.134622</td>\n",
       "      <td>-0.072735</td>\n",
       "      <td>1.117311</td>\n",
       "      <td>0.530235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>-0.411977</td>\n",
       "      <td>0.362871</td>\n",
       "      <td>-1.834723</td>\n",
       "      <td>0.049433</td>\n",
       "      <td>-0.509645</td>\n",
       "      <td>0.276153</td>\n",
       "      <td>-0.247823</td>\n",
       "      <td>0.221101</td>\n",
       "      <td>-0.502014</td>\n",
       "      <td>1.184999</td>\n",
       "      <td>...</td>\n",
       "      <td>0.493495</td>\n",
       "      <td>0.502407</td>\n",
       "      <td>1.243561</td>\n",
       "      <td>-1.015062</td>\n",
       "      <td>-0.413656</td>\n",
       "      <td>0.168762</td>\n",
       "      <td>-0.128181</td>\n",
       "      <td>0.226103</td>\n",
       "      <td>1.303812</td>\n",
       "      <td>-0.115343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>-1.218965</td>\n",
       "      <td>0.830500</td>\n",
       "      <td>-2.185841</td>\n",
       "      <td>-0.570560</td>\n",
       "      <td>-1.280202</td>\n",
       "      <td>0.319064</td>\n",
       "      <td>0.003466</td>\n",
       "      <td>1.010240</td>\n",
       "      <td>-0.428807</td>\n",
       "      <td>1.630356</td>\n",
       "      <td>...</td>\n",
       "      <td>1.075938</td>\n",
       "      <td>1.085889</td>\n",
       "      <td>1.208442</td>\n",
       "      <td>-1.600671</td>\n",
       "      <td>-0.350294</td>\n",
       "      <td>1.333407</td>\n",
       "      <td>0.171035</td>\n",
       "      <td>0.425677</td>\n",
       "      <td>1.997774</td>\n",
       "      <td>0.812689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>-0.555137</td>\n",
       "      <td>0.558694</td>\n",
       "      <td>-2.143589</td>\n",
       "      <td>-0.260143</td>\n",
       "      <td>-0.811713</td>\n",
       "      <td>0.420748</td>\n",
       "      <td>-0.149957</td>\n",
       "      <td>0.586930</td>\n",
       "      <td>-0.516998</td>\n",
       "      <td>1.409124</td>\n",
       "      <td>...</td>\n",
       "      <td>0.678266</td>\n",
       "      <td>0.758758</td>\n",
       "      <td>1.219930</td>\n",
       "      <td>-1.169663</td>\n",
       "      <td>-0.401978</td>\n",
       "      <td>0.716644</td>\n",
       "      <td>-0.240001</td>\n",
       "      <td>0.351458</td>\n",
       "      <td>1.469989</td>\n",
       "      <td>0.357084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>-0.497441</td>\n",
       "      <td>0.900854</td>\n",
       "      <td>-1.739998</td>\n",
       "      <td>-0.500726</td>\n",
       "      <td>-1.275969</td>\n",
       "      <td>0.361475</td>\n",
       "      <td>-0.109821</td>\n",
       "      <td>0.769593</td>\n",
       "      <td>-0.474697</td>\n",
       "      <td>1.584012</td>\n",
       "      <td>...</td>\n",
       "      <td>0.191692</td>\n",
       "      <td>1.092978</td>\n",
       "      <td>1.367714</td>\n",
       "      <td>-1.470395</td>\n",
       "      <td>-0.727541</td>\n",
       "      <td>0.222102</td>\n",
       "      <td>0.114576</td>\n",
       "      <td>0.101187</td>\n",
       "      <td>0.752133</td>\n",
       "      <td>0.215249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>2.161353</td>\n",
       "      <td>-1.558242</td>\n",
       "      <td>1.166957</td>\n",
       "      <td>0.620180</td>\n",
       "      <td>2.168992</td>\n",
       "      <td>-0.391184</td>\n",
       "      <td>-0.669115</td>\n",
       "      <td>-0.061056</td>\n",
       "      <td>-0.099860</td>\n",
       "      <td>-2.276891</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.160515</td>\n",
       "      <td>-2.118650</td>\n",
       "      <td>-3.155472</td>\n",
       "      <td>0.915395</td>\n",
       "      <td>-0.566094</td>\n",
       "      <td>0.313201</td>\n",
       "      <td>0.265942</td>\n",
       "      <td>1.462074</td>\n",
       "      <td>0.200964</td>\n",
       "      <td>-1.775188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>0.648648</td>\n",
       "      <td>-0.581670</td>\n",
       "      <td>0.295617</td>\n",
       "      <td>0.400857</td>\n",
       "      <td>0.573323</td>\n",
       "      <td>-0.187852</td>\n",
       "      <td>-0.302794</td>\n",
       "      <td>0.610208</td>\n",
       "      <td>-0.122874</td>\n",
       "      <td>-0.504728</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.344615</td>\n",
       "      <td>-0.578575</td>\n",
       "      <td>-0.853332</td>\n",
       "      <td>-0.059915</td>\n",
       "      <td>-0.600465</td>\n",
       "      <td>0.220790</td>\n",
       "      <td>0.144245</td>\n",
       "      <td>0.723104</td>\n",
       "      <td>0.151220</td>\n",
       "      <td>-0.371426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>-0.760844</td>\n",
       "      <td>0.710512</td>\n",
       "      <td>-2.481286</td>\n",
       "      <td>-0.437427</td>\n",
       "      <td>-1.308108</td>\n",
       "      <td>0.591027</td>\n",
       "      <td>0.056683</td>\n",
       "      <td>0.763948</td>\n",
       "      <td>-0.719780</td>\n",
       "      <td>1.900450</td>\n",
       "      <td>...</td>\n",
       "      <td>1.156037</td>\n",
       "      <td>1.171746</td>\n",
       "      <td>1.545236</td>\n",
       "      <td>-1.502601</td>\n",
       "      <td>-0.571165</td>\n",
       "      <td>0.619852</td>\n",
       "      <td>-0.324282</td>\n",
       "      <td>0.352138</td>\n",
       "      <td>1.448214</td>\n",
       "      <td>0.588982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>2.295607</td>\n",
       "      <td>-0.580548</td>\n",
       "      <td>2.293776</td>\n",
       "      <td>0.177691</td>\n",
       "      <td>2.017256</td>\n",
       "      <td>-1.174565</td>\n",
       "      <td>-0.612211</td>\n",
       "      <td>0.374687</td>\n",
       "      <td>0.971726</td>\n",
       "      <td>-2.776167</td>\n",
       "      <td>...</td>\n",
       "      <td>-3.149711</td>\n",
       "      <td>-2.009171</td>\n",
       "      <td>-3.409827</td>\n",
       "      <td>1.543367</td>\n",
       "      <td>-0.957289</td>\n",
       "      <td>0.129220</td>\n",
       "      <td>0.979079</td>\n",
       "      <td>0.772927</td>\n",
       "      <td>0.111955</td>\n",
       "      <td>-1.813775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>-0.271630</td>\n",
       "      <td>0.657064</td>\n",
       "      <td>-0.642705</td>\n",
       "      <td>-0.644728</td>\n",
       "      <td>-1.180068</td>\n",
       "      <td>-0.188553</td>\n",
       "      <td>-0.175109</td>\n",
       "      <td>0.399957</td>\n",
       "      <td>0.031009</td>\n",
       "      <td>0.968563</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.919186</td>\n",
       "      <td>0.778575</td>\n",
       "      <td>1.060992</td>\n",
       "      <td>-1.556234</td>\n",
       "      <td>-0.915651</td>\n",
       "      <td>-1.105132</td>\n",
       "      <td>0.062578</td>\n",
       "      <td>-0.155323</td>\n",
       "      <td>0.038430</td>\n",
       "      <td>-0.088435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>-0.536143</td>\n",
       "      <td>1.546998</td>\n",
       "      <td>-2.851274</td>\n",
       "      <td>-0.864680</td>\n",
       "      <td>-2.135705</td>\n",
       "      <td>0.428020</td>\n",
       "      <td>-0.005084</td>\n",
       "      <td>1.061637</td>\n",
       "      <td>-0.446020</td>\n",
       "      <td>2.200855</td>\n",
       "      <td>...</td>\n",
       "      <td>0.847694</td>\n",
       "      <td>1.534841</td>\n",
       "      <td>1.929033</td>\n",
       "      <td>-1.666917</td>\n",
       "      <td>-1.111955</td>\n",
       "      <td>0.419459</td>\n",
       "      <td>-0.133787</td>\n",
       "      <td>-0.126385</td>\n",
       "      <td>0.805780</td>\n",
       "      <td>0.639809</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25 rows × 100 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0         1         2         3         4         5         6   \\\n",
       "0   1.269504 -1.301618  0.659774  0.806424  1.774142 -0.343432 -0.779151   \n",
       "1  -1.219352  0.858967 -2.209150 -0.584687 -1.285293  0.307079 -0.010814   \n",
       "2   1.455337 -0.348460  2.492809 -0.002663  1.328323 -0.874232 -0.157595   \n",
       "3  -1.128447  0.772527 -2.491378 -0.651029 -1.623185  0.663864  0.125967   \n",
       "4  -1.036939  0.809364 -2.464699 -0.687017 -1.576593  0.638427  0.118440   \n",
       "5  -0.547059  0.894858 -1.755465 -0.490597 -1.380145  0.359310 -0.060178   \n",
       "6   1.733710 -0.566986  2.949093  0.390996  2.418676 -1.398353 -0.696868   \n",
       "7   1.902613 -1.295641  2.932941  0.292558  2.542777 -0.784014 -0.564508   \n",
       "8   0.986385 -0.628855  0.235144  0.294113  0.755645 -0.287721 -0.565301   \n",
       "9   0.685016 -0.085557 -0.214265  0.292677  0.134052 -0.131870 -0.285285   \n",
       "10  0.701596 -0.450739  1.009886  0.350183  0.902692 -0.374130 -0.330576   \n",
       "11  1.683621 -0.727743  0.919457  0.141144  1.054881 -0.339080 -0.291974   \n",
       "12  0.466126  0.101711 -0.077127  0.155005  0.125790 -0.041057 -0.276507   \n",
       "13 -0.759412  1.024231 -2.187070 -0.540236 -1.619746  0.164110 -0.068278   \n",
       "14 -0.754947  1.020849 -2.226209 -0.547058 -1.670198  0.195537 -0.056791   \n",
       "15 -0.411977  0.362871 -1.834723  0.049433 -0.509645  0.276153 -0.247823   \n",
       "16 -1.218965  0.830500 -2.185841 -0.570560 -1.280202  0.319064  0.003466   \n",
       "17 -0.555137  0.558694 -2.143589 -0.260143 -0.811713  0.420748 -0.149957   \n",
       "18 -0.497441  0.900854 -1.739998 -0.500726 -1.275969  0.361475 -0.109821   \n",
       "19  2.161353 -1.558242  1.166957  0.620180  2.168992 -0.391184 -0.669115   \n",
       "20  0.648648 -0.581670  0.295617  0.400857  0.573323 -0.187852 -0.302794   \n",
       "21 -0.760844  0.710512 -2.481286 -0.437427 -1.308108  0.591027  0.056683   \n",
       "22  2.295607 -0.580548  2.293776  0.177691  2.017256 -1.174565 -0.612211   \n",
       "23 -0.271630  0.657064 -0.642705 -0.644728 -1.180068 -0.188553 -0.175109   \n",
       "24 -0.536143  1.546998 -2.851274 -0.864680 -2.135705  0.428020 -0.005084   \n",
       "\n",
       "          7         8         9   ...        90        91        92        93  \\\n",
       "0  -0.215126 -0.186087 -1.474602  ... -1.693101 -1.635854 -1.960710  0.339103   \n",
       "1   1.042626 -0.409562  1.655249  ...  1.111762  1.085674  1.186331 -1.579450   \n",
       "2   0.838077  0.904898 -1.876874  ... -3.277988 -1.223678 -2.357131  0.819557   \n",
       "3   1.104853 -0.519641  2.120744  ...  1.073344  1.364996  1.529100 -1.807320   \n",
       "4   1.119194 -0.525833  2.067393  ...  1.060543  1.332533  1.429077 -1.721097   \n",
       "5   0.771358 -0.472385  1.670432  ...  0.283317  1.162495  1.423480 -1.511162   \n",
       "6   0.764553  1.415131 -3.074588  ... -3.647666 -2.091958 -3.698048  1.739185   \n",
       "7   0.691168  0.872062 -2.692172  ... -3.710976 -2.149990 -3.701959  1.104722   \n",
       "8  -0.137433 -0.318865 -0.364859  ... -1.473106 -0.788323 -0.883935 -0.450523   \n",
       "9   0.141165 -0.274307 -0.183934  ... -0.833542 -0.349754 -0.144961 -0.388498   \n",
       "10  0.256810  0.090173 -0.911043  ... -1.842188 -0.739546 -0.848902 -0.062544   \n",
       "11  0.352025  0.022157 -1.230969  ... -2.081283 -1.231950 -1.862873  0.450388   \n",
       "12  0.292252 -0.257001 -0.115536  ... -0.799583 -0.130002 -0.045846 -0.405901   \n",
       "13  0.645948 -0.407626  1.856634  ...  0.856082  1.134467  1.595821 -1.509817   \n",
       "14  0.645053 -0.396648  1.872752  ...  0.898750  1.148711  1.644138 -1.553703   \n",
       "15  0.221101 -0.502014  1.184999  ...  0.493495  0.502407  1.243561 -1.015062   \n",
       "16  1.010240 -0.428807  1.630356  ...  1.075938  1.085889  1.208442 -1.600671   \n",
       "17  0.586930 -0.516998  1.409124  ...  0.678266  0.758758  1.219930 -1.169663   \n",
       "18  0.769593 -0.474697  1.584012  ...  0.191692  1.092978  1.367714 -1.470395   \n",
       "19 -0.061056 -0.099860 -2.276891  ... -2.160515 -2.118650 -3.155472  0.915395   \n",
       "20  0.610208 -0.122874 -0.504728  ... -1.344615 -0.578575 -0.853332 -0.059915   \n",
       "21  0.763948 -0.719780  1.900450  ...  1.156037  1.171746  1.545236 -1.502601   \n",
       "22  0.374687  0.971726 -2.776167  ... -3.149711 -2.009171 -3.409827  1.543367   \n",
       "23  0.399957  0.031009  0.968563  ... -0.919186  0.778575  1.060992 -1.556234   \n",
       "24  1.061637 -0.446020  2.200855  ...  0.847694  1.534841  1.929033 -1.666917   \n",
       "\n",
       "          94        95        96        97        98        99  \n",
       "0  -0.025599  0.260683  0.163350  1.253483  0.558545 -1.454138  \n",
       "1  -0.355581  1.330155  0.142522  0.414558  2.009863  0.834968  \n",
       "2  -0.908406  0.079156  1.186641  0.651305 -0.033716 -1.122361  \n",
       "3  -0.537589  1.110650 -0.042950  0.446592  1.589758  0.847480  \n",
       "4  -0.570125  1.051956 -0.089572  0.433013  1.584785  0.764304  \n",
       "5  -0.741800  0.205500  0.125913  0.088043  0.769423  0.282571  \n",
       "6  -0.828084  0.024867  1.245571  0.942139 -0.019091 -1.623251  \n",
       "7  -0.588351  0.193837  1.070350  1.491803 -0.170504 -1.562872  \n",
       "8  -0.389556 -0.204728  0.018448  0.864524  0.220383 -1.094242  \n",
       "9  -0.842064 -0.071577  0.339474  0.374968  0.558935 -0.894052  \n",
       "10 -0.576259 -0.004731  0.499465  0.434306  0.375023 -0.781382  \n",
       "11 -0.862843  0.106902  0.379381  0.846408 -0.022321 -1.240623  \n",
       "12 -0.820496 -0.442367  0.338477  0.264230  0.435244 -0.790947  \n",
       "13 -0.782409  0.228485 -0.154000 -0.065237  1.118699  0.502978  \n",
       "14 -0.814907  0.271635 -0.134622 -0.072735  1.117311  0.530235  \n",
       "15 -0.413656  0.168762 -0.128181  0.226103  1.303812 -0.115343  \n",
       "16 -0.350294  1.333407  0.171035  0.425677  1.997774  0.812689  \n",
       "17 -0.401978  0.716644 -0.240001  0.351458  1.469989  0.357084  \n",
       "18 -0.727541  0.222102  0.114576  0.101187  0.752133  0.215249  \n",
       "19 -0.566094  0.313201  0.265942  1.462074  0.200964 -1.775188  \n",
       "20 -0.600465  0.220790  0.144245  0.723104  0.151220 -0.371426  \n",
       "21 -0.571165  0.619852 -0.324282  0.352138  1.448214  0.588982  \n",
       "22 -0.957289  0.129220  0.979079  0.772927  0.111955 -1.813775  \n",
       "23 -0.915651 -1.105132  0.062578 -0.155323  0.038430 -0.088435  \n",
       "24 -1.111955  0.419459 -0.133787 -0.126385  0.805780  0.639809  \n",
       "\n",
       "[25 rows x 100 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mags_df = pd.DataFrame (list_of_mag_vectors)\n",
    "mags_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "neither-credits",
   "metadata": {},
   "source": [
    "---\n",
    "## Clustering\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "valued-cylinder",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "familiar-ownership",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "grateful-fancy",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "alert-muscle",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "authentic-teddy",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
