{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "handed-austin",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import altair as alt\n",
    "import os, random\n",
    "from Bio import SeqIO\n",
    "import gensim\n",
    "from gensim.models import Word2Vec\n",
    "from sklearn.cluster import KMeans\n",
    "import altair as alt\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "native-substitute",
   "metadata": {},
   "source": [
    "## Defining paths for each and every omic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "dramatic-norfolk",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_root_data = os.path.join ('..', 'Data', 'Extracted', 'Databases')\n",
    "path_all_fasta = os.path.join (path_root_data, 'fasta_files', 'AllBins')\n",
    "path_normalised_metabolomics = os.path.join (path_root_data, 'Metabolomics', 'Normalised_Tables')\n",
    "path_model_save_root = 'Saved_models'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "beautiful-pulse",
   "metadata": {},
   "outputs": [],
   "source": [
    "num_of_mags = len([i for i in os.listdir(path_all_fasta) if (i[-2:] == 'fa')])\n",
    "SEED = 42\n",
    "END = 25\n",
    "num_of_workers = 3\n",
    "random.seed(SEED)\n",
    "np.random.seed(SEED)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "speaking-investor",
   "metadata": {},
   "source": [
    "---\n",
    "## MAG-related functions\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "retained-prize",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function that splits each genome into k-mers thus creating even longer sentence (MAG)\n",
    "# It returns tokenized genome i.e. [kmer, kmer,...]\n",
    "def split_genome (genome, k = 5):\n",
    "    new_genome = []\n",
    "    n = len(genome)\n",
    "    \n",
    "    if n-k <=0:\n",
    "        return genome\n",
    "    else:\n",
    "        for i in range(n-k):\n",
    "            new_genome.append(genome[i:i+k])\n",
    "        \n",
    "        return new_genome\n",
    "\n",
    "\n",
    "def vectorize_one_mag (one_mag, w2v_model):\n",
    "    \n",
    "    # We have to generate vectors for each word in one MAG and then create vector representation of that MAG\n",
    "    # by averaging vectors of its words\n",
    "    zero_vector = np.zeros(w2v_model.vector_size)\n",
    "    word_vectors = []\n",
    "    one_mag_vector = []\n",
    "    \n",
    "    for sentence in one_mag:\n",
    "        for word in sentence:\n",
    "            if word in w2v_model.wv:\n",
    "                try:\n",
    "                    word_vectors.append(w2v_model.wv[word])\n",
    "                except KeyError:\n",
    "                    print ('Key Error')\n",
    "                    continue\n",
    "    \n",
    "    if word_vectors:\n",
    "        word_vectors = np.asarray(word_vectors)\n",
    "        one_mag_vector = word_vectors.mean (axis=0)\n",
    "    \n",
    "    else:\n",
    "        one_mag_vector = zero_vector\n",
    "    \n",
    "    return one_mag_vector\n",
    "\n",
    "\n",
    "# Function that vectorizes a MAG (document) with a pretrained word2vec model. It returns vector representation of a given MAG\n",
    "# Vectorization is done by averaging word (k-mer) vectors for the whole document (MAG)\n",
    "def vectorize_mags (w2v_model, end = 25):\n",
    "    \n",
    "    print ('Vectorizing MAGs')\n",
    "    \n",
    "    fasta_files = [i for i in os.listdir(path_all_fasta) if (i[-2:] == 'fa')]\n",
    "    list_of_mag_vectors = []\n",
    "    \n",
    "    # This was done so that I could work with first 'end' FASTA files only. Otherwise, I should just remove: i, and enumerate\n",
    "    for i, fasta_file_name in enumerate(fasta_files):\n",
    "        \n",
    "        if i == end:\n",
    "            break\n",
    "        \n",
    "        else:\n",
    "            with open(os.path.join(path_all_fasta, fasta_file_name), 'r') as input_file:\n",
    "                \n",
    "                one_mag = []\n",
    "                for fasta_string in SeqIO.parse(input_file, \"fasta\"):\n",
    "                    \n",
    "                    # Get kmers of a genome and create a sentence (list of words)\n",
    "                    temp_kmers = split_genome (str(fasta_string.seq))\n",
    "                    \n",
    "                    # Create a document (list of sentences)\n",
    "                    one_mag.append(temp_kmers)\n",
    "                    \n",
    "                # Vectorizing MAGs one by one\n",
    "                list_of_mag_vectors.append (vectorize_one_mag (one_mag, w2v_model))\n",
    "    \n",
    "    print ('Finished vectorizing')\n",
    "    \n",
    "    return list_of_mag_vectors\n",
    "    \n",
    "    \n",
    "\n",
    "# If one wants to import MAGs in order to vectorize them, one should use start argument in order to skip first 'start' MAGs\n",
    "# If one wants to import MAGs to train word2vec model, one should use only end argument, so that first 'end' MAGs are used for training\n",
    "# Todo: Implement randomisation in picking MAGs for training, and don't use first 'start' MAGs for training\n",
    "# (create list of indexes from 0 to 1364, use sklearn split train test, then traverse directory and use only MAGs with indexes in train for training w2v)\n",
    "def import_mags_and_build_model (end = 25, path_all_fasta = path_all_fasta):\n",
    "    \n",
    "    print ('Importing MAGs')\n",
    "    \n",
    "    # There are 1364 MAGs enclosed in FASTA files\n",
    "    # I have to traverse every FASTA file, and in each file every sequence\n",
    "    \n",
    "    fasta_files = [i for i in os.listdir(path_all_fasta) if (i[-2:] == 'fa')]\n",
    "    \n",
    "    # This was done so that I could work with first 100 FASTA files only. Otherwise, I should just remove: i, and enumerate\n",
    "    for i, fasta_file_name in enumerate(fasta_files):\n",
    "        \n",
    "        if i == end:\n",
    "            break\n",
    "        \n",
    "        else:\n",
    "            with open(os.path.join(path_all_fasta, fasta_file_name), 'r') as input_file:\n",
    "                \n",
    "                one_mag = []\n",
    "                for fasta_string in SeqIO.parse(input_file, \"fasta\"):\n",
    "                    \n",
    "                    # Get kmers of a genome and create a sentence (list of words)\n",
    "                    temp_kmers = split_genome (str(fasta_string.seq))\n",
    "                    \n",
    "                    # Create a document (list of sentences)\n",
    "                    one_mag.append(temp_kmers)\n",
    "                    \n",
    "                # If we do not have a model, we build one\n",
    "                if i == 0:\n",
    "                    print ('Building w2v model')\n",
    "                    # We build our model on the first MAG\n",
    "                    w2v_model = Word2Vec (sentences = one_mag, size = 100, workers = num_of_workers, seed=SEED)\n",
    "                \n",
    "                # Else we just expand its vocabulary\n",
    "                else:\n",
    "                    # Now we expand our vocabulary\n",
    "                    w2v_model.build_vocab (one_mag, update = True)\n",
    "                    \n",
    "    print ('Finished building')\n",
    "    \n",
    "    return w2v_model\n",
    "\n",
    "\n",
    "def train_model (w2v_model, epochs, end = 25):\n",
    "    \n",
    "    print ('Starting model training')\n",
    "    \n",
    "    # There are 1364 MAGs enclosed in FASTA files\n",
    "    # I have to traverse every FASTA file, and in each file every sequence\n",
    "    \n",
    "    fasta_files = [i for i in os.listdir(path_all_fasta) if (i[-2:] == 'fa')]\n",
    "    \n",
    "    # This was done so that I could work with first 100 FASTA files only. Otherwise, I should just remove: i, and enumerate\n",
    "    for i, fasta_file_name in enumerate(fasta_files):\n",
    "        \n",
    "        if i == end:\n",
    "            break\n",
    "        \n",
    "        else:\n",
    "            with open(os.path.join(path_all_fasta, fasta_file_name), 'r') as input_file:\n",
    "                \n",
    "                one_mag = []\n",
    "                for fasta_string in SeqIO.parse(input_file, \"fasta\"):\n",
    "                    \n",
    "                    # Get kmers of a genome and create a sentence (list of words)\n",
    "                    temp_kmers = split_genome (str(fasta_string.seq))\n",
    "                    \n",
    "                    # Create a document (list of sentences)\n",
    "                    one_mag.append(temp_kmers)\n",
    "                    \n",
    "                \n",
    "                w2v_model.train (one_mag, total_examples = w2v_model.corpus_count, epochs = epochs)\n",
    "                    \n",
    "    print ('Model training finished')\n",
    "    \n",
    "    return w2v_model\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "subject-uzbekistan",
   "metadata": {},
   "source": [
    "---\n",
    "## Importing Metabolomic data\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "applied-option",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Metabolite</th>\n",
       "      <th>tp</th>\n",
       "      <th>date</th>\n",
       "      <th>type</th>\n",
       "      <th>known_type</th>\n",
       "      <th>means</th>\n",
       "      <th>medians</th>\n",
       "      <th>sds</th>\n",
       "      <th>N</th>\n",
       "      <th>se</th>\n",
       "      <th>ci</th>\n",
       "      <th>type2</th>\n",
       "      <th>measurement</th>\n",
       "      <th>KEGG.Compound.ID</th>\n",
       "      <th>Chebi.Name</th>\n",
       "      <th>Chebi.Name_combined</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Phosphoric_acid_3TMS</td>\n",
       "      <td>D1</td>\n",
       "      <td>2012-04-17</td>\n",
       "      <td>snp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.757113</td>\n",
       "      <td>0.737943</td>\n",
       "      <td>0.098650</td>\n",
       "      <td>3</td>\n",
       "      <td>0.056956</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>C00009</td>\n",
       "      <td>phosphoric acid</td>\n",
       "      <td>phosphoric acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Phosphoric_acid_3TMS</td>\n",
       "      <td>D10</td>\n",
       "      <td>2012-01-25</td>\n",
       "      <td>snp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.720935</td>\n",
       "      <td>0.873509</td>\n",
       "      <td>0.266593</td>\n",
       "      <td>3</td>\n",
       "      <td>0.153918</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>C00009</td>\n",
       "      <td>phosphoric acid</td>\n",
       "      <td>phosphoric acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Phosphoric_acid_3TMS</td>\n",
       "      <td>D11</td>\n",
       "      <td>2012-03-22</td>\n",
       "      <td>snp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.625047</td>\n",
       "      <td>0.652268</td>\n",
       "      <td>0.057820</td>\n",
       "      <td>3</td>\n",
       "      <td>0.033382</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>C00009</td>\n",
       "      <td>phosphoric acid</td>\n",
       "      <td>phosphoric acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Phosphoric_acid_3TMS</td>\n",
       "      <td>D12</td>\n",
       "      <td>2012-01-19</td>\n",
       "      <td>snp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.964956</td>\n",
       "      <td>0.941878</td>\n",
       "      <td>0.053754</td>\n",
       "      <td>3</td>\n",
       "      <td>0.031035</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>C00009</td>\n",
       "      <td>phosphoric acid</td>\n",
       "      <td>phosphoric acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>!Phosphoric_acid_3TMS</td>\n",
       "      <td>D13</td>\n",
       "      <td>2011-05-13</td>\n",
       "      <td>snp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.963984</td>\n",
       "      <td>0.974479</td>\n",
       "      <td>0.071659</td>\n",
       "      <td>3</td>\n",
       "      <td>0.041373</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>C00009</td>\n",
       "      <td>phosphoric acid</td>\n",
       "      <td>phosphoric acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18865</th>\n",
       "      <td>Valine_2TMS</td>\n",
       "      <td>D51</td>\n",
       "      <td>2011-08-11</td>\n",
       "      <td>bp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.704821</td>\n",
       "      <td>0.478251</td>\n",
       "      <td>0.428368</td>\n",
       "      <td>3</td>\n",
       "      <td>0.247319</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C16436,C06417,C00183</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18866</th>\n",
       "      <td>Valine_2TMS</td>\n",
       "      <td>D6</td>\n",
       "      <td>2011-06-03</td>\n",
       "      <td>bp</td>\n",
       "      <td>known</td>\n",
       "      <td>1.085888</td>\n",
       "      <td>1.197471</td>\n",
       "      <td>0.487484</td>\n",
       "      <td>3</td>\n",
       "      <td>0.281449</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C16436,C06417,C00183</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18867</th>\n",
       "      <td>Valine_2TMS</td>\n",
       "      <td>D7</td>\n",
       "      <td>2012-04-04</td>\n",
       "      <td>bp</td>\n",
       "      <td>known</td>\n",
       "      <td>2.958023</td>\n",
       "      <td>2.813145</td>\n",
       "      <td>0.341213</td>\n",
       "      <td>3</td>\n",
       "      <td>0.197000</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C16436,C06417,C00183</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18868</th>\n",
       "      <td>Valine_2TMS</td>\n",
       "      <td>D8</td>\n",
       "      <td>2012-02-08</td>\n",
       "      <td>bp</td>\n",
       "      <td>known</td>\n",
       "      <td>2.515980</td>\n",
       "      <td>2.322248</td>\n",
       "      <td>0.618556</td>\n",
       "      <td>3</td>\n",
       "      <td>0.357123</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C16436,C06417,C00183</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18869</th>\n",
       "      <td>Valine_2TMS</td>\n",
       "      <td>D9</td>\n",
       "      <td>2011-12-28</td>\n",
       "      <td>bp</td>\n",
       "      <td>known</td>\n",
       "      <td>2.356989</td>\n",
       "      <td>2.280757</td>\n",
       "      <td>0.264918</td>\n",
       "      <td>3</td>\n",
       "      <td>0.152950</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C16436,C06417,C00183</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18870 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Metabolite   tp        date type known_type     means  \\\n",
       "0      !Phosphoric_acid_3TMS   D1  2012-04-17  snp      known  0.757113   \n",
       "1      !Phosphoric_acid_3TMS  D10  2012-01-25  snp      known  0.720935   \n",
       "2      !Phosphoric_acid_3TMS  D11  2012-03-22  snp      known  0.625047   \n",
       "3      !Phosphoric_acid_3TMS  D12  2012-01-19  snp      known  0.964956   \n",
       "4      !Phosphoric_acid_3TMS  D13  2011-05-13  snp      known  0.963984   \n",
       "...                      ...  ...         ...  ...        ...       ...   \n",
       "18865            Valine_2TMS  D51  2011-08-11   bp      known  0.704821   \n",
       "18866            Valine_2TMS   D6  2011-06-03   bp      known  1.085888   \n",
       "18867            Valine_2TMS   D7  2012-04-04   bp      known  2.958023   \n",
       "18868            Valine_2TMS   D8  2012-02-08   bp      known  2.515980   \n",
       "18869            Valine_2TMS   D9  2011-12-28   bp      known  2.356989   \n",
       "\n",
       "        medians       sds  N        se        ci          type2 measurement  \\\n",
       "0      0.737943  0.098650  3  0.056956  4.302653  extracellular    nonpolar   \n",
       "1      0.873509  0.266593  3  0.153918  4.302653  extracellular    nonpolar   \n",
       "2      0.652268  0.057820  3  0.033382  4.302653  extracellular    nonpolar   \n",
       "3      0.941878  0.053754  3  0.031035  4.302653  extracellular    nonpolar   \n",
       "4      0.974479  0.071659  3  0.041373  4.302653  extracellular    nonpolar   \n",
       "...         ...       ... ..       ...       ...            ...         ...   \n",
       "18865  0.478251  0.428368  3  0.247319  4.302653  intracellular       polar   \n",
       "18866  1.197471  0.487484  3  0.281449  4.302653  intracellular       polar   \n",
       "18867  2.813145  0.341213  3  0.197000  4.302653  intracellular       polar   \n",
       "18868  2.322248  0.618556  3  0.357123  4.302653  intracellular       polar   \n",
       "18869  2.280757  0.264918  3  0.152950  4.302653  intracellular       polar   \n",
       "\n",
       "           KEGG.Compound.ID                Chebi.Name  \\\n",
       "0                    C00009           phosphoric acid   \n",
       "1                    C00009           phosphoric acid   \n",
       "2                    C00009           phosphoric acid   \n",
       "3                    C00009           phosphoric acid   \n",
       "4                    C00009           phosphoric acid   \n",
       "...                     ...                       ...   \n",
       "18865  C16436,C06417,C00183  valine;D-valine;L-valine   \n",
       "18866  C16436,C06417,C00183  valine;D-valine;L-valine   \n",
       "18867  C16436,C06417,C00183  valine;D-valine;L-valine   \n",
       "18868  C16436,C06417,C00183  valine;D-valine;L-valine   \n",
       "18869  C16436,C06417,C00183  valine;D-valine;L-valine   \n",
       "\n",
       "            Chebi.Name_combined  \n",
       "0               phosphoric acid  \n",
       "1               phosphoric acid  \n",
       "2               phosphoric acid  \n",
       "3               phosphoric acid  \n",
       "4               phosphoric acid  \n",
       "...                         ...  \n",
       "18865  valine;D-valine;L-valine  \n",
       "18866  valine;D-valine;L-valine  \n",
       "18867  valine;D-valine;L-valine  \n",
       "18868  valine;D-valine;L-valine  \n",
       "18869  valine;D-valine;L-valine  \n",
       "\n",
       "[18870 rows x 16 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metabolomics_file_name = os.path.join(path_normalised_metabolomics, os.listdir(path_normalised_metabolomics)[0])\n",
    "metabolomics_df = pd.read_csv (metabolomics_file_name, delimiter = '\\t')\n",
    "metabolomics_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "revolutionary-segment",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>date</th>\n",
       "      <th>Metabolite</th>\n",
       "      <th>tp</th>\n",
       "      <th>type</th>\n",
       "      <th>known_type</th>\n",
       "      <th>means</th>\n",
       "      <th>medians</th>\n",
       "      <th>sds</th>\n",
       "      <th>N</th>\n",
       "      <th>se</th>\n",
       "      <th>ci</th>\n",
       "      <th>type2</th>\n",
       "      <th>measurement</th>\n",
       "      <th>KEGG.Compound.ID</th>\n",
       "      <th>Chebi.Name</th>\n",
       "      <th>Chebi.Name_combined</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>Leucine_2TMS</td>\n",
       "      <td>D37</td>\n",
       "      <td>sp</td>\n",
       "      <td>known</td>\n",
       "      <td>1.440122</td>\n",
       "      <td>1.489979</td>\n",
       "      <td>0.146440</td>\n",
       "      <td>3</td>\n",
       "      <td>0.084547</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C16439,C00123,C01570</td>\n",
       "      <td>leucine;L-leucine;D-leucine</td>\n",
       "      <td>leucine;L-leucine;D-leucine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>No match: 1181.78_EM_SP_D1_1_223</td>\n",
       "      <td>D37</td>\n",
       "      <td>sp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>3.778469</td>\n",
       "      <td>3.801417</td>\n",
       "      <td>1.924377</td>\n",
       "      <td>3</td>\n",
       "      <td>1.111040</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>unknown#emu_WW_2144.25</td>\n",
       "      <td>D37</td>\n",
       "      <td>bnp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.949545</td>\n",
       "      <td>0.947209</td>\n",
       "      <td>0.065422</td>\n",
       "      <td>3</td>\n",
       "      <td>0.037771</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>Unknown#bth_pae_001</td>\n",
       "      <td>D37</td>\n",
       "      <td>bp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2.095116</td>\n",
       "      <td>1.809891</td>\n",
       "      <td>1.916896</td>\n",
       "      <td>3</td>\n",
       "      <td>1.106721</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>No match: 2220.34_EM_BP_D1_1_192</td>\n",
       "      <td>D37</td>\n",
       "      <td>bp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.631643</td>\n",
       "      <td>1.571125</td>\n",
       "      <td>0.168501</td>\n",
       "      <td>3</td>\n",
       "      <td>0.097284</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18865</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>Octadecanol_1TMS</td>\n",
       "      <td>D48</td>\n",
       "      <td>bnp</td>\n",
       "      <td>known</td>\n",
       "      <td>1.364105</td>\n",
       "      <td>1.457780</td>\n",
       "      <td>0.170311</td>\n",
       "      <td>3</td>\n",
       "      <td>0.098329</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>D01924</td>\n",
       "      <td>octadecan-1-ol</td>\n",
       "      <td>octadecan-1-ol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18866</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>No match: 1447.93_EM_BP_D2_1_57</td>\n",
       "      <td>D48</td>\n",
       "      <td>bp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>4.815774</td>\n",
       "      <td>1.160165</td>\n",
       "      <td>6.835474</td>\n",
       "      <td>3</td>\n",
       "      <td>3.946463</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18867</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>Lactulose_8TMS</td>\n",
       "      <td>D48</td>\n",
       "      <td>bp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.697489</td>\n",
       "      <td>0.684339</td>\n",
       "      <td>0.049632</td>\n",
       "      <td>3</td>\n",
       "      <td>0.028655</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C07064</td>\n",
       "      <td>lactulose</td>\n",
       "      <td>lactulose</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18868</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>No match: 2004.78_EM_BNP_D10_1_205</td>\n",
       "      <td>D48</td>\n",
       "      <td>bnp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.785996</td>\n",
       "      <td>0.744529</td>\n",
       "      <td>0.139901</td>\n",
       "      <td>3</td>\n",
       "      <td>0.080772</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18869</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>2-Oxobutanoic_acid_1_1MeOX_1TMS</td>\n",
       "      <td>D48</td>\n",
       "      <td>sp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.755565</td>\n",
       "      <td>0.709972</td>\n",
       "      <td>0.149214</td>\n",
       "      <td>3</td>\n",
       "      <td>0.086149</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C00109</td>\n",
       "      <td>2-oxobutanoic acid</td>\n",
       "      <td>2-oxobutanoic acid</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18870 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            date                          Metabolite   tp type known_type  \\\n",
       "0     2011-03-21                        Leucine_2TMS  D37   sp      known   \n",
       "1     2011-03-21    No match: 1181.78_EM_SP_D1_1_223  D37   sp    unknown   \n",
       "2     2011-03-21              unknown#emu_WW_2144.25  D37  bnp    unknown   \n",
       "3     2011-03-21                 Unknown#bth_pae_001  D37   bp    unknown   \n",
       "4     2011-03-21    No match: 2220.34_EM_BP_D1_1_192  D37   bp    unknown   \n",
       "...          ...                                 ...  ...  ...        ...   \n",
       "18865 2012-05-03                    Octadecanol_1TMS  D48  bnp      known   \n",
       "18866 2012-05-03     No match: 1447.93_EM_BP_D2_1_57  D48   bp    unknown   \n",
       "18867 2012-05-03                      Lactulose_8TMS  D48   bp      known   \n",
       "18868 2012-05-03  No match: 2004.78_EM_BNP_D10_1_205  D48  bnp    unknown   \n",
       "18869 2012-05-03     2-Oxobutanoic_acid_1_1MeOX_1TMS  D48   sp      known   \n",
       "\n",
       "          means   medians       sds  N        se        ci          type2  \\\n",
       "0      1.440122  1.489979  0.146440  3  0.084547  4.302653  extracellular   \n",
       "1      3.778469  3.801417  1.924377  3  1.111040  4.302653  extracellular   \n",
       "2      0.949545  0.947209  0.065422  3  0.037771  4.302653  intracellular   \n",
       "3      2.095116  1.809891  1.916896  3  1.106721  4.302653  intracellular   \n",
       "4      1.631643  1.571125  0.168501  3  0.097284  4.302653  intracellular   \n",
       "...         ...       ...       ... ..       ...       ...            ...   \n",
       "18865  1.364105  1.457780  0.170311  3  0.098329  4.302653  intracellular   \n",
       "18866  4.815774  1.160165  6.835474  3  3.946463  4.302653  intracellular   \n",
       "18867  0.697489  0.684339  0.049632  3  0.028655  4.302653  intracellular   \n",
       "18868  0.785996  0.744529  0.139901  3  0.080772  4.302653  intracellular   \n",
       "18869  0.755565  0.709972  0.149214  3  0.086149  4.302653  extracellular   \n",
       "\n",
       "      measurement      KEGG.Compound.ID                   Chebi.Name  \\\n",
       "0           polar  C16439,C00123,C01570  leucine;L-leucine;D-leucine   \n",
       "1           polar                   NaN                          NaN   \n",
       "2        nonpolar                   NaN                          NaN   \n",
       "3           polar                   NaN                          NaN   \n",
       "4           polar                   NaN                          NaN   \n",
       "...           ...                   ...                          ...   \n",
       "18865    nonpolar                D01924               octadecan-1-ol   \n",
       "18866       polar                   NaN                          NaN   \n",
       "18867       polar                C07064                    lactulose   \n",
       "18868    nonpolar                   NaN                          NaN   \n",
       "18869       polar                C00109           2-oxobutanoic acid   \n",
       "\n",
       "               Chebi.Name_combined  \n",
       "0      leucine;L-leucine;D-leucine  \n",
       "1                              NaN  \n",
       "2                              NaN  \n",
       "3                              NaN  \n",
       "4                              NaN  \n",
       "...                            ...  \n",
       "18865               octadecan-1-ol  \n",
       "18866                          NaN  \n",
       "18867                    lactulose  \n",
       "18868                          NaN  \n",
       "18869           2-oxobutanoic acid  \n",
       "\n",
       "[18870 rows x 16 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metabolomics_df['date'] = pd.to_datetime(metabolomics_df['date'])\n",
    "metabolomics_df.insert (0, 'date', metabolomics_df.pop('date'))\n",
    "metabolomics_df.sort_values ('date', inplace = True, ignore_index = True)\n",
    "metabolomics_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "champion-rogers",
   "metadata": {},
   "source": [
    "## MAG examination"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "similar-court",
   "metadata": {},
   "source": [
    "---\n",
    "Important info:\n",
    "\n",
    "> We can look at metagenome as a sentence which consists of words (that are represented with genomes). Therefore, we can create metagenomes as strings where words are seperated with space character (' '). This is one way of tackling this problem"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "preceding-climate",
   "metadata": {},
   "source": [
    "To do:\n",
    "1. Make kmers of some length, out of every word in every sentence\n",
    "2. Encode every sentence with Word2Vec and CBOW (because it's faster than skip-gram)\n",
    "3. Reduce dimensionality with PCA or with autoencoder\n",
    "    > Maybe it is good to do the first 3 steps while importing, for each and every sentence. That way, I might use less memory\n",
    "4. Cluster similar sentences (MAGs) to get rMAGs\n",
    "\n",
    "> Important: this is not temporal data, since there are no timestamps!\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "virtual-setting",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing MAGs\n",
      "Building w2v model\n",
      "Finished building\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# FOR CLUSTERING I SHOULD CREATE A DATAFRAME WITH MAGs INDEXES AND THEIR VECTOR REPRESENTATIONS\n",
    "\n",
    "final_model = import_mags_and_build_model (end = END, path_all_fasta = path_all_fasta)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "wrapped-notification",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('CCGGC', 0.5894356966018677),\n",
       " ('TCGGC', 0.5622138977050781),\n",
       " ('ACGGA', 0.5432533025741577),\n",
       " ('ACGGG', 0.5209246873855591),\n",
       " ('GCGGC', 0.5119935274124146),\n",
       " ('ACGGT', 0.4680752158164978),\n",
       " ('ACGAT', 0.2959875166416168),\n",
       " ('ACGAC', 0.2835376560688019),\n",
       " ('TACGA', 0.28112727403640747),\n",
       " ('TACGC', 0.2611392140388489)]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_model.wv.most_similar('ACGGC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "exact-madrid",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting model training\n",
      "Model training finished\n"
     ]
    }
   ],
   "source": [
    "# Train model. It tooks ~10 minutes for END = 25 amount of MAGs\n",
    "epochs = 10\n",
    "\n",
    "final_model = train_model (final_model, epochs = epochs, end = END)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "italic-bride",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_model.wv.save_word2vec_format(os.path.join (path_model_save_root, 'model_25.bin'), binary=True) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "subjective-invasion",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "leading-authority",
   "metadata": {},
   "source": [
    "Now I should vectorize documents with this model. For further use, I could save this model's weights, and use it to vectorize all mags. That would take a lot, but every MAG will have its vector representation\n",
    "> This could be done by importing one MAG at a time, then tokenizing it (like before), then getting vector representations of that MAG's sentences (genomes) and then finding the vector representation of the whole MAG (document). If I do that for one MAG at a time, There is no need to worry about memory\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "authorized-queens",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Vectorizing MAGs\n",
      "Finished vectorizing\n"
     ]
    }
   ],
   "source": [
    "list_of_mag_vectors = vectorize_mags (final_model, end = END)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "visible-enforcement",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[array([-1.1965473e+00, -3.4466343e+00,  1.1225539e+00,  2.6312834e-01,\n",
       "        -2.5361535e-01,  1.8168186e+00,  1.3943011e+00,  5.9793311e-01,\n",
       "        -2.0670781e+00, -2.0657931e-01,  2.3056791e+00, -1.0977945e+00,\n",
       "         1.2557960e+00, -9.9643773e-01,  1.1149993e+00,  1.8037397e+00,\n",
       "        -2.9583761e-01, -4.4174930e-01, -8.8848591e-01,  1.2939855e+00,\n",
       "        -4.4530161e-02, -1.0593354e+00, -1.1706048e+00, -2.1949111e-01,\n",
       "        -8.0128688e-01, -6.9839859e-01, -1.0369163e-01,  2.2203580e-01,\n",
       "         1.0739380e+00, -1.4342694e+00, -1.4528227e-01,  2.1896055e-01,\n",
       "         4.8607826e-01,  1.1318611e+00,  1.5093282e+00,  2.1629584e+00,\n",
       "         7.7518100e-01, -6.0288894e-01,  4.9686894e-01,  7.2373733e-02,\n",
       "        -1.9086123e+00, -6.7348307e-01, -1.8191979e+00,  2.6915163e-01,\n",
       "         7.9759979e-01,  1.7312871e+00, -2.3905370e-01, -3.9395061e-01,\n",
       "         1.1420479e+00, -1.6216646e+00,  3.4147367e-01, -3.3497950e-01,\n",
       "        -2.4591829e-03, -2.2679090e+00,  2.6078584e+00,  1.7288103e+00,\n",
       "        -7.9421496e-01, -2.2722781e-01,  1.3172734e+00,  1.0351394e+00,\n",
       "         4.6868160e-01, -1.8194504e-01,  3.6719728e-01,  2.3028682e-01,\n",
       "        -9.6427530e-01, -3.0429411e-01, -2.1593340e-01, -9.0763278e-02,\n",
       "        -6.4876276e-01,  1.8181096e-01, -1.1641691e-01, -1.0461124e+00,\n",
       "         1.1803749e+00,  2.1974628e+00,  2.8706485e-01, -1.4310737e-01,\n",
       "         8.3780617e-01,  8.8564473e-01,  4.4147503e-01, -1.2050735e+00,\n",
       "         8.5391057e-01,  1.7718740e+00, -5.5195868e-01,  5.2406383e-01,\n",
       "        -6.0771060e-01,  1.6487300e+00, -1.3435346e+00, -1.4417949e+00,\n",
       "         1.6463741e+00,  9.8803200e-02, -1.5863578e-01,  1.4259063e+00,\n",
       "         1.0251006e+00, -2.8354146e-02, -4.8041087e-02,  1.6294100e+00,\n",
       "        -9.5364356e-01, -2.8359219e-01, -7.0035350e-01, -1.0901330e+00],\n",
       "       dtype=float32),\n",
       " array([ 0.0490588 ,  2.8238797 ,  0.36262918, -1.5082027 , -0.32531515,\n",
       "        -1.4649022 , -0.79279613,  1.2461046 ,  1.0363489 ,  1.3318874 ,\n",
       "        -0.989357  , -0.7983788 , -0.8009588 ,  0.21222457, -0.0036513 ,\n",
       "        -1.3409563 ,  1.620286  , -0.60548735,  0.18047385, -1.8051909 ,\n",
       "        -0.7442775 , -0.39012802, -0.0636607 ,  2.017091  ,  0.10923091,\n",
       "        -0.5722706 , -1.1342194 ,  0.32641417, -2.6835063 ,  1.2183248 ,\n",
       "         0.05183235,  0.6307464 ,  0.20472166,  0.9847831 , -1.3527107 ,\n",
       "        -0.8584051 ,  0.39097103,  0.8982126 , -1.3853167 , -0.7767684 ,\n",
       "         0.4445061 , -1.6512398 ,  2.3961842 ,  0.3070077 , -1.265948  ,\n",
       "        -2.3706782 , -0.4835135 , -0.07411972,  0.6009345 ,  1.7791187 ,\n",
       "         0.04154263,  0.05471612,  0.7632086 ,  1.9861615 , -1.6657532 ,\n",
       "        -1.5594969 ,  0.05618905, -0.15973862, -1.2863268 , -0.27211648,\n",
       "        -0.12915374,  0.02516208,  0.5202215 , -0.61697584,  0.8625217 ,\n",
       "        -1.0377291 ,  0.14354588, -0.8988974 ,  0.6445315 , -0.9132908 ,\n",
       "        -0.16702668,  0.6269215 , -0.8265134 , -0.6423112 ,  0.4261871 ,\n",
       "         0.42267442, -0.7851802 ,  1.2930673 , -0.2113053 ,  0.71132493,\n",
       "         1.4617265 , -0.6709841 , -0.5596241 , -0.3509214 , -0.07466502,\n",
       "        -2.3158476 ,  1.09259   ,  1.5409704 , -0.5497132 , -0.1827418 ,\n",
       "        -0.887114  , -1.5875286 , -1.2643818 ,  0.9739708 , -0.8452433 ,\n",
       "        -0.6871335 ,  1.4822772 ,  0.22602493,  1.3908288 ,  1.9599937 ],\n",
       "       dtype=float32)]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_of_mag_vectors[0:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "excellent-retirement",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>90</th>\n",
       "      <th>91</th>\n",
       "      <th>92</th>\n",
       "      <th>93</th>\n",
       "      <th>94</th>\n",
       "      <th>95</th>\n",
       "      <th>96</th>\n",
       "      <th>97</th>\n",
       "      <th>98</th>\n",
       "      <th>99</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-1.196547</td>\n",
       "      <td>-3.446634</td>\n",
       "      <td>1.122554</td>\n",
       "      <td>0.263128</td>\n",
       "      <td>-0.253615</td>\n",
       "      <td>1.816819</td>\n",
       "      <td>1.394301</td>\n",
       "      <td>0.597933</td>\n",
       "      <td>-2.067078</td>\n",
       "      <td>-0.206579</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.158636</td>\n",
       "      <td>1.425906</td>\n",
       "      <td>1.025101</td>\n",
       "      <td>-0.028354</td>\n",
       "      <td>-0.048041</td>\n",
       "      <td>1.629410</td>\n",
       "      <td>-0.953644</td>\n",
       "      <td>-0.283592</td>\n",
       "      <td>-0.700354</td>\n",
       "      <td>-1.090133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.049059</td>\n",
       "      <td>2.823880</td>\n",
       "      <td>0.362629</td>\n",
       "      <td>-1.508203</td>\n",
       "      <td>-0.325315</td>\n",
       "      <td>-1.464902</td>\n",
       "      <td>-0.792796</td>\n",
       "      <td>1.246105</td>\n",
       "      <td>1.036349</td>\n",
       "      <td>1.331887</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.887114</td>\n",
       "      <td>-1.587529</td>\n",
       "      <td>-1.264382</td>\n",
       "      <td>0.973971</td>\n",
       "      <td>-0.845243</td>\n",
       "      <td>-0.687133</td>\n",
       "      <td>1.482277</td>\n",
       "      <td>0.226025</td>\n",
       "      <td>1.390829</td>\n",
       "      <td>1.959994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-1.076538</td>\n",
       "      <td>-4.783527</td>\n",
       "      <td>-0.096637</td>\n",
       "      <td>0.892442</td>\n",
       "      <td>0.856198</td>\n",
       "      <td>2.473337</td>\n",
       "      <td>0.901526</td>\n",
       "      <td>0.557027</td>\n",
       "      <td>-1.793551</td>\n",
       "      <td>-0.557365</td>\n",
       "      <td>...</td>\n",
       "      <td>0.062814</td>\n",
       "      <td>0.682193</td>\n",
       "      <td>0.660506</td>\n",
       "      <td>-0.592617</td>\n",
       "      <td>-0.380943</td>\n",
       "      <td>0.848278</td>\n",
       "      <td>-1.756477</td>\n",
       "      <td>-0.635433</td>\n",
       "      <td>-2.427694</td>\n",
       "      <td>-1.583026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.164475</td>\n",
       "      <td>3.217961</td>\n",
       "      <td>-0.096349</td>\n",
       "      <td>-1.096770</td>\n",
       "      <td>-0.300357</td>\n",
       "      <td>-1.441947</td>\n",
       "      <td>-0.763520</td>\n",
       "      <td>1.692813</td>\n",
       "      <td>1.136963</td>\n",
       "      <td>1.400865</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.791861</td>\n",
       "      <td>-1.958941</td>\n",
       "      <td>-1.440642</td>\n",
       "      <td>1.214189</td>\n",
       "      <td>-0.765646</td>\n",
       "      <td>-0.853262</td>\n",
       "      <td>1.519032</td>\n",
       "      <td>0.188515</td>\n",
       "      <td>1.302754</td>\n",
       "      <td>2.157268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.246169</td>\n",
       "      <td>3.115447</td>\n",
       "      <td>-0.082878</td>\n",
       "      <td>-1.079046</td>\n",
       "      <td>-0.321819</td>\n",
       "      <td>-1.298257</td>\n",
       "      <td>-0.760366</td>\n",
       "      <td>1.661492</td>\n",
       "      <td>1.121160</td>\n",
       "      <td>1.410003</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.782524</td>\n",
       "      <td>-1.796969</td>\n",
       "      <td>-1.363379</td>\n",
       "      <td>1.163899</td>\n",
       "      <td>-0.732312</td>\n",
       "      <td>-0.764388</td>\n",
       "      <td>1.450671</td>\n",
       "      <td>0.191647</td>\n",
       "      <td>1.245304</td>\n",
       "      <td>2.095642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>-0.073989</td>\n",
       "      <td>1.920999</td>\n",
       "      <td>-0.485796</td>\n",
       "      <td>-0.798457</td>\n",
       "      <td>-0.299321</td>\n",
       "      <td>-0.918790</td>\n",
       "      <td>-0.671724</td>\n",
       "      <td>1.400493</td>\n",
       "      <td>0.766859</td>\n",
       "      <td>0.836819</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.702894</td>\n",
       "      <td>-1.193122</td>\n",
       "      <td>-0.917276</td>\n",
       "      <td>0.615969</td>\n",
       "      <td>-0.763657</td>\n",
       "      <td>-0.400789</td>\n",
       "      <td>0.823787</td>\n",
       "      <td>0.114617</td>\n",
       "      <td>0.536050</td>\n",
       "      <td>1.900726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>-1.539677</td>\n",
       "      <td>-6.411491</td>\n",
       "      <td>0.716679</td>\n",
       "      <td>0.576411</td>\n",
       "      <td>0.459347</td>\n",
       "      <td>3.469873</td>\n",
       "      <td>1.477935</td>\n",
       "      <td>-0.333925</td>\n",
       "      <td>-2.537221</td>\n",
       "      <td>-1.127924</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.147849</td>\n",
       "      <td>1.183525</td>\n",
       "      <td>0.890112</td>\n",
       "      <td>-1.368696</td>\n",
       "      <td>-0.020944</td>\n",
       "      <td>1.508202</td>\n",
       "      <td>-2.503601</td>\n",
       "      <td>-0.887476</td>\n",
       "      <td>-3.253759</td>\n",
       "      <td>-2.831345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>-1.284740</td>\n",
       "      <td>-6.045468</td>\n",
       "      <td>0.443491</td>\n",
       "      <td>0.987065</td>\n",
       "      <td>0.897312</td>\n",
       "      <td>3.389049</td>\n",
       "      <td>1.879721</td>\n",
       "      <td>0.610316</td>\n",
       "      <td>-2.420880</td>\n",
       "      <td>-0.271895</td>\n",
       "      <td>...</td>\n",
       "      <td>0.297817</td>\n",
       "      <td>1.334344</td>\n",
       "      <td>1.391637</td>\n",
       "      <td>-0.642675</td>\n",
       "      <td>-0.048700</td>\n",
       "      <td>1.547720</td>\n",
       "      <td>-2.346980</td>\n",
       "      <td>-0.859433</td>\n",
       "      <td>-2.758810</td>\n",
       "      <td>-2.664072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>-0.579216</td>\n",
       "      <td>-1.839221</td>\n",
       "      <td>0.276878</td>\n",
       "      <td>0.045237</td>\n",
       "      <td>-0.073658</td>\n",
       "      <td>1.040149</td>\n",
       "      <td>0.755374</td>\n",
       "      <td>1.006982</td>\n",
       "      <td>-1.031035</td>\n",
       "      <td>0.293702</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.229055</td>\n",
       "      <td>0.767412</td>\n",
       "      <td>0.814296</td>\n",
       "      <td>0.126623</td>\n",
       "      <td>-0.252945</td>\n",
       "      <td>1.046583</td>\n",
       "      <td>-0.636697</td>\n",
       "      <td>0.065987</td>\n",
       "      <td>-0.687458</td>\n",
       "      <td>-0.310260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>-0.945494</td>\n",
       "      <td>-1.383347</td>\n",
       "      <td>0.255619</td>\n",
       "      <td>-0.045844</td>\n",
       "      <td>-0.424141</td>\n",
       "      <td>0.829961</td>\n",
       "      <td>0.247029</td>\n",
       "      <td>0.974351</td>\n",
       "      <td>-0.841089</td>\n",
       "      <td>0.088883</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.719211</td>\n",
       "      <td>0.411402</td>\n",
       "      <td>0.037168</td>\n",
       "      <td>-0.058658</td>\n",
       "      <td>-0.365759</td>\n",
       "      <td>0.596159</td>\n",
       "      <td>-0.516978</td>\n",
       "      <td>-0.328708</td>\n",
       "      <td>-0.495169</td>\n",
       "      <td>0.442289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>-1.077694</td>\n",
       "      <td>-2.567544</td>\n",
       "      <td>0.071874</td>\n",
       "      <td>0.335944</td>\n",
       "      <td>0.411825</td>\n",
       "      <td>1.065646</td>\n",
       "      <td>0.433874</td>\n",
       "      <td>0.451725</td>\n",
       "      <td>-1.429661</td>\n",
       "      <td>-0.734585</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.212799</td>\n",
       "      <td>0.951852</td>\n",
       "      <td>0.293330</td>\n",
       "      <td>0.033528</td>\n",
       "      <td>-0.382728</td>\n",
       "      <td>0.969915</td>\n",
       "      <td>-0.843015</td>\n",
       "      <td>-0.186675</td>\n",
       "      <td>-0.711314</td>\n",
       "      <td>-0.193649</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>-0.709038</td>\n",
       "      <td>-3.270481</td>\n",
       "      <td>0.778171</td>\n",
       "      <td>0.400224</td>\n",
       "      <td>0.020348</td>\n",
       "      <td>1.992440</td>\n",
       "      <td>1.182444</td>\n",
       "      <td>1.078757</td>\n",
       "      <td>-1.577477</td>\n",
       "      <td>0.218275</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.066839</td>\n",
       "      <td>0.956308</td>\n",
       "      <td>0.824618</td>\n",
       "      <td>-0.272232</td>\n",
       "      <td>-0.050337</td>\n",
       "      <td>1.046628</td>\n",
       "      <td>-1.434512</td>\n",
       "      <td>-0.442222</td>\n",
       "      <td>-1.568484</td>\n",
       "      <td>-0.932112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>-0.837098</td>\n",
       "      <td>-1.091814</td>\n",
       "      <td>0.095694</td>\n",
       "      <td>-0.007199</td>\n",
       "      <td>-0.297777</td>\n",
       "      <td>0.851025</td>\n",
       "      <td>0.173454</td>\n",
       "      <td>0.892706</td>\n",
       "      <td>-0.683448</td>\n",
       "      <td>-0.051315</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.609853</td>\n",
       "      <td>0.336005</td>\n",
       "      <td>-0.025942</td>\n",
       "      <td>-0.009587</td>\n",
       "      <td>-0.428027</td>\n",
       "      <td>0.578746</td>\n",
       "      <td>-0.477153</td>\n",
       "      <td>-0.132794</td>\n",
       "      <td>-0.557401</td>\n",
       "      <td>0.500838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>0.411740</td>\n",
       "      <td>3.174164</td>\n",
       "      <td>-0.081461</td>\n",
       "      <td>-1.476627</td>\n",
       "      <td>-0.512489</td>\n",
       "      <td>-1.472918</td>\n",
       "      <td>-1.016073</td>\n",
       "      <td>1.085818</td>\n",
       "      <td>1.003715</td>\n",
       "      <td>0.850834</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.875578</td>\n",
       "      <td>-1.198156</td>\n",
       "      <td>-1.054554</td>\n",
       "      <td>0.446105</td>\n",
       "      <td>-0.545685</td>\n",
       "      <td>-0.420433</td>\n",
       "      <td>1.150592</td>\n",
       "      <td>0.610831</td>\n",
       "      <td>0.766962</td>\n",
       "      <td>2.025311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>0.377875</td>\n",
       "      <td>3.228982</td>\n",
       "      <td>-0.070239</td>\n",
       "      <td>-1.497753</td>\n",
       "      <td>-0.535361</td>\n",
       "      <td>-1.550593</td>\n",
       "      <td>-0.985662</td>\n",
       "      <td>1.124721</td>\n",
       "      <td>1.015387</td>\n",
       "      <td>0.874665</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.888018</td>\n",
       "      <td>-1.305566</td>\n",
       "      <td>-1.087319</td>\n",
       "      <td>0.468503</td>\n",
       "      <td>-0.544103</td>\n",
       "      <td>-0.467280</td>\n",
       "      <td>1.177980</td>\n",
       "      <td>0.612449</td>\n",
       "      <td>0.777439</td>\n",
       "      <td>2.049372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>-0.524236</td>\n",
       "      <td>1.690124</td>\n",
       "      <td>0.360152</td>\n",
       "      <td>-1.021231</td>\n",
       "      <td>-0.809366</td>\n",
       "      <td>-0.742110</td>\n",
       "      <td>-0.560704</td>\n",
       "      <td>0.723358</td>\n",
       "      <td>0.072640</td>\n",
       "      <td>0.204775</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.898124</td>\n",
       "      <td>-0.534893</td>\n",
       "      <td>-0.827300</td>\n",
       "      <td>0.460882</td>\n",
       "      <td>-0.457188</td>\n",
       "      <td>0.073724</td>\n",
       "      <td>0.887954</td>\n",
       "      <td>0.485289</td>\n",
       "      <td>0.745222</td>\n",
       "      <td>1.501954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>0.022214</td>\n",
       "      <td>2.791778</td>\n",
       "      <td>0.391767</td>\n",
       "      <td>-1.505518</td>\n",
       "      <td>-0.314841</td>\n",
       "      <td>-1.483677</td>\n",
       "      <td>-0.768949</td>\n",
       "      <td>1.277467</td>\n",
       "      <td>1.043308</td>\n",
       "      <td>1.300657</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.908197</td>\n",
       "      <td>-1.612261</td>\n",
       "      <td>-1.272619</td>\n",
       "      <td>1.000879</td>\n",
       "      <td>-0.858920</td>\n",
       "      <td>-0.716763</td>\n",
       "      <td>1.478209</td>\n",
       "      <td>0.240117</td>\n",
       "      <td>1.382134</td>\n",
       "      <td>1.949388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>-0.332514</td>\n",
       "      <td>2.240826</td>\n",
       "      <td>0.499190</td>\n",
       "      <td>-1.198200</td>\n",
       "      <td>-0.647167</td>\n",
       "      <td>-1.076311</td>\n",
       "      <td>-0.553736</td>\n",
       "      <td>0.912188</td>\n",
       "      <td>0.467460</td>\n",
       "      <td>0.556399</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.754629</td>\n",
       "      <td>-0.877227</td>\n",
       "      <td>-1.142292</td>\n",
       "      <td>0.813646</td>\n",
       "      <td>-0.516773</td>\n",
       "      <td>-0.240872</td>\n",
       "      <td>1.053816</td>\n",
       "      <td>0.426998</td>\n",
       "      <td>1.019562</td>\n",
       "      <td>1.634114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>-0.150348</td>\n",
       "      <td>1.719013</td>\n",
       "      <td>-0.500162</td>\n",
       "      <td>-0.752633</td>\n",
       "      <td>-0.276743</td>\n",
       "      <td>-0.826216</td>\n",
       "      <td>-0.671714</td>\n",
       "      <td>1.360025</td>\n",
       "      <td>0.699547</td>\n",
       "      <td>0.722576</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.699954</td>\n",
       "      <td>-1.021949</td>\n",
       "      <td>-0.865052</td>\n",
       "      <td>0.617698</td>\n",
       "      <td>-0.728692</td>\n",
       "      <td>-0.289909</td>\n",
       "      <td>0.769596</td>\n",
       "      <td>0.075830</td>\n",
       "      <td>0.536395</td>\n",
       "      <td>1.838766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>-1.064339</td>\n",
       "      <td>-4.642395</td>\n",
       "      <td>1.728962</td>\n",
       "      <td>0.447515</td>\n",
       "      <td>-0.189246</td>\n",
       "      <td>2.685462</td>\n",
       "      <td>2.132598</td>\n",
       "      <td>0.912721</td>\n",
       "      <td>-2.282389</td>\n",
       "      <td>0.163796</td>\n",
       "      <td>...</td>\n",
       "      <td>0.050407</td>\n",
       "      <td>1.718530</td>\n",
       "      <td>1.492898</td>\n",
       "      <td>-0.410896</td>\n",
       "      <td>0.204248</td>\n",
       "      <td>1.757781</td>\n",
       "      <td>-1.839840</td>\n",
       "      <td>-0.585769</td>\n",
       "      <td>-1.710074</td>\n",
       "      <td>-1.876389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>-1.126585</td>\n",
       "      <td>-1.723158</td>\n",
       "      <td>0.350172</td>\n",
       "      <td>0.296101</td>\n",
       "      <td>-0.025916</td>\n",
       "      <td>0.996593</td>\n",
       "      <td>0.718516</td>\n",
       "      <td>0.763507</td>\n",
       "      <td>-1.208054</td>\n",
       "      <td>0.197199</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.055121</td>\n",
       "      <td>0.178722</td>\n",
       "      <td>0.090902</td>\n",
       "      <td>0.094516</td>\n",
       "      <td>-0.227029</td>\n",
       "      <td>0.668818</td>\n",
       "      <td>-0.542119</td>\n",
       "      <td>-0.446755</td>\n",
       "      <td>-0.715341</td>\n",
       "      <td>0.088875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>0.076902</td>\n",
       "      <td>3.143089</td>\n",
       "      <td>0.232344</td>\n",
       "      <td>-1.177991</td>\n",
       "      <td>-0.591331</td>\n",
       "      <td>-1.297317</td>\n",
       "      <td>-0.641267</td>\n",
       "      <td>1.326268</td>\n",
       "      <td>0.965242</td>\n",
       "      <td>1.159404</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.716265</td>\n",
       "      <td>-1.378925</td>\n",
       "      <td>-1.262173</td>\n",
       "      <td>0.997633</td>\n",
       "      <td>-0.603737</td>\n",
       "      <td>-0.562393</td>\n",
       "      <td>1.270813</td>\n",
       "      <td>0.461659</td>\n",
       "      <td>1.141418</td>\n",
       "      <td>2.057800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>-1.170744</td>\n",
       "      <td>-5.721959</td>\n",
       "      <td>1.088727</td>\n",
       "      <td>0.241799</td>\n",
       "      <td>0.041649</td>\n",
       "      <td>2.969879</td>\n",
       "      <td>1.494480</td>\n",
       "      <td>0.028380</td>\n",
       "      <td>-2.512938</td>\n",
       "      <td>-0.834894</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.162391</td>\n",
       "      <td>1.428777</td>\n",
       "      <td>0.985188</td>\n",
       "      <td>-1.260356</td>\n",
       "      <td>0.100895</td>\n",
       "      <td>1.585065</td>\n",
       "      <td>-2.331019</td>\n",
       "      <td>-0.538911</td>\n",
       "      <td>-2.809221</td>\n",
       "      <td>-2.523257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>0.267196</td>\n",
       "      <td>1.715368</td>\n",
       "      <td>-1.195532</td>\n",
       "      <td>-1.015345</td>\n",
       "      <td>0.556674</td>\n",
       "      <td>-0.647347</td>\n",
       "      <td>-1.205126</td>\n",
       "      <td>0.573093</td>\n",
       "      <td>0.610161</td>\n",
       "      <td>-0.088773</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.565490</td>\n",
       "      <td>-0.325405</td>\n",
       "      <td>-0.162700</td>\n",
       "      <td>-0.039037</td>\n",
       "      <td>-0.389480</td>\n",
       "      <td>0.093041</td>\n",
       "      <td>0.176263</td>\n",
       "      <td>0.844495</td>\n",
       "      <td>-0.485920</td>\n",
       "      <td>1.027265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>0.233661</td>\n",
       "      <td>3.211500</td>\n",
       "      <td>-0.156633</td>\n",
       "      <td>-1.358121</td>\n",
       "      <td>-0.902594</td>\n",
       "      <td>-1.440895</td>\n",
       "      <td>-0.853459</td>\n",
       "      <td>1.595041</td>\n",
       "      <td>1.413270</td>\n",
       "      <td>1.073089</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.990036</td>\n",
       "      <td>-1.728213</td>\n",
       "      <td>-1.393406</td>\n",
       "      <td>0.556439</td>\n",
       "      <td>-0.518007</td>\n",
       "      <td>-0.775438</td>\n",
       "      <td>1.138622</td>\n",
       "      <td>0.140067</td>\n",
       "      <td>0.596830</td>\n",
       "      <td>2.305128</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25 rows × 100 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0         1         2         3         4         5         6   \\\n",
       "0  -1.196547 -3.446634  1.122554  0.263128 -0.253615  1.816819  1.394301   \n",
       "1   0.049059  2.823880  0.362629 -1.508203 -0.325315 -1.464902 -0.792796   \n",
       "2  -1.076538 -4.783527 -0.096637  0.892442  0.856198  2.473337  0.901526   \n",
       "3   0.164475  3.217961 -0.096349 -1.096770 -0.300357 -1.441947 -0.763520   \n",
       "4   0.246169  3.115447 -0.082878 -1.079046 -0.321819 -1.298257 -0.760366   \n",
       "5  -0.073989  1.920999 -0.485796 -0.798457 -0.299321 -0.918790 -0.671724   \n",
       "6  -1.539677 -6.411491  0.716679  0.576411  0.459347  3.469873  1.477935   \n",
       "7  -1.284740 -6.045468  0.443491  0.987065  0.897312  3.389049  1.879721   \n",
       "8  -0.579216 -1.839221  0.276878  0.045237 -0.073658  1.040149  0.755374   \n",
       "9  -0.945494 -1.383347  0.255619 -0.045844 -0.424141  0.829961  0.247029   \n",
       "10 -1.077694 -2.567544  0.071874  0.335944  0.411825  1.065646  0.433874   \n",
       "11 -0.709038 -3.270481  0.778171  0.400224  0.020348  1.992440  1.182444   \n",
       "12 -0.837098 -1.091814  0.095694 -0.007199 -0.297777  0.851025  0.173454   \n",
       "13  0.411740  3.174164 -0.081461 -1.476627 -0.512489 -1.472918 -1.016073   \n",
       "14  0.377875  3.228982 -0.070239 -1.497753 -0.535361 -1.550593 -0.985662   \n",
       "15 -0.524236  1.690124  0.360152 -1.021231 -0.809366 -0.742110 -0.560704   \n",
       "16  0.022214  2.791778  0.391767 -1.505518 -0.314841 -1.483677 -0.768949   \n",
       "17 -0.332514  2.240826  0.499190 -1.198200 -0.647167 -1.076311 -0.553736   \n",
       "18 -0.150348  1.719013 -0.500162 -0.752633 -0.276743 -0.826216 -0.671714   \n",
       "19 -1.064339 -4.642395  1.728962  0.447515 -0.189246  2.685462  2.132598   \n",
       "20 -1.126585 -1.723158  0.350172  0.296101 -0.025916  0.996593  0.718516   \n",
       "21  0.076902  3.143089  0.232344 -1.177991 -0.591331 -1.297317 -0.641267   \n",
       "22 -1.170744 -5.721959  1.088727  0.241799  0.041649  2.969879  1.494480   \n",
       "23  0.267196  1.715368 -1.195532 -1.015345  0.556674 -0.647347 -1.205126   \n",
       "24  0.233661  3.211500 -0.156633 -1.358121 -0.902594 -1.440895 -0.853459   \n",
       "\n",
       "          7         8         9   ...        90        91        92        93  \\\n",
       "0   0.597933 -2.067078 -0.206579  ... -0.158636  1.425906  1.025101 -0.028354   \n",
       "1   1.246105  1.036349  1.331887  ... -0.887114 -1.587529 -1.264382  0.973971   \n",
       "2   0.557027 -1.793551 -0.557365  ...  0.062814  0.682193  0.660506 -0.592617   \n",
       "3   1.692813  1.136963  1.400865  ... -0.791861 -1.958941 -1.440642  1.214189   \n",
       "4   1.661492  1.121160  1.410003  ... -0.782524 -1.796969 -1.363379  1.163899   \n",
       "5   1.400493  0.766859  0.836819  ... -0.702894 -1.193122 -0.917276  0.615969   \n",
       "6  -0.333925 -2.537221 -1.127924  ... -0.147849  1.183525  0.890112 -1.368696   \n",
       "7   0.610316 -2.420880 -0.271895  ...  0.297817  1.334344  1.391637 -0.642675   \n",
       "8   1.006982 -1.031035  0.293702  ... -0.229055  0.767412  0.814296  0.126623   \n",
       "9   0.974351 -0.841089  0.088883  ... -0.719211  0.411402  0.037168 -0.058658   \n",
       "10  0.451725 -1.429661 -0.734585  ... -0.212799  0.951852  0.293330  0.033528   \n",
       "11  1.078757 -1.577477  0.218275  ... -0.066839  0.956308  0.824618 -0.272232   \n",
       "12  0.892706 -0.683448 -0.051315  ... -0.609853  0.336005 -0.025942 -0.009587   \n",
       "13  1.085818  1.003715  0.850834  ... -0.875578 -1.198156 -1.054554  0.446105   \n",
       "14  1.124721  1.015387  0.874665  ... -0.888018 -1.305566 -1.087319  0.468503   \n",
       "15  0.723358  0.072640  0.204775  ... -0.898124 -0.534893 -0.827300  0.460882   \n",
       "16  1.277467  1.043308  1.300657  ... -0.908197 -1.612261 -1.272619  1.000879   \n",
       "17  0.912188  0.467460  0.556399  ... -0.754629 -0.877227 -1.142292  0.813646   \n",
       "18  1.360025  0.699547  0.722576  ... -0.699954 -1.021949 -0.865052  0.617698   \n",
       "19  0.912721 -2.282389  0.163796  ...  0.050407  1.718530  1.492898 -0.410896   \n",
       "20  0.763507 -1.208054  0.197199  ... -0.055121  0.178722  0.090902  0.094516   \n",
       "21  1.326268  0.965242  1.159404  ... -0.716265 -1.378925 -1.262173  0.997633   \n",
       "22  0.028380 -2.512938 -0.834894  ... -0.162391  1.428777  0.985188 -1.260356   \n",
       "23  0.573093  0.610161 -0.088773  ... -0.565490 -0.325405 -0.162700 -0.039037   \n",
       "24  1.595041  1.413270  1.073089  ... -0.990036 -1.728213 -1.393406  0.556439   \n",
       "\n",
       "          94        95        96        97        98        99  \n",
       "0  -0.048041  1.629410 -0.953644 -0.283592 -0.700354 -1.090133  \n",
       "1  -0.845243 -0.687133  1.482277  0.226025  1.390829  1.959994  \n",
       "2  -0.380943  0.848278 -1.756477 -0.635433 -2.427694 -1.583026  \n",
       "3  -0.765646 -0.853262  1.519032  0.188515  1.302754  2.157268  \n",
       "4  -0.732312 -0.764388  1.450671  0.191647  1.245304  2.095642  \n",
       "5  -0.763657 -0.400789  0.823787  0.114617  0.536050  1.900726  \n",
       "6  -0.020944  1.508202 -2.503601 -0.887476 -3.253759 -2.831345  \n",
       "7  -0.048700  1.547720 -2.346980 -0.859433 -2.758810 -2.664072  \n",
       "8  -0.252945  1.046583 -0.636697  0.065987 -0.687458 -0.310260  \n",
       "9  -0.365759  0.596159 -0.516978 -0.328708 -0.495169  0.442289  \n",
       "10 -0.382728  0.969915 -0.843015 -0.186675 -0.711314 -0.193649  \n",
       "11 -0.050337  1.046628 -1.434512 -0.442222 -1.568484 -0.932112  \n",
       "12 -0.428027  0.578746 -0.477153 -0.132794 -0.557401  0.500838  \n",
       "13 -0.545685 -0.420433  1.150592  0.610831  0.766962  2.025311  \n",
       "14 -0.544103 -0.467280  1.177980  0.612449  0.777439  2.049372  \n",
       "15 -0.457188  0.073724  0.887954  0.485289  0.745222  1.501954  \n",
       "16 -0.858920 -0.716763  1.478209  0.240117  1.382134  1.949388  \n",
       "17 -0.516773 -0.240872  1.053816  0.426998  1.019562  1.634114  \n",
       "18 -0.728692 -0.289909  0.769596  0.075830  0.536395  1.838766  \n",
       "19  0.204248  1.757781 -1.839840 -0.585769 -1.710074 -1.876389  \n",
       "20 -0.227029  0.668818 -0.542119 -0.446755 -0.715341  0.088875  \n",
       "21 -0.603737 -0.562393  1.270813  0.461659  1.141418  2.057800  \n",
       "22  0.100895  1.585065 -2.331019 -0.538911 -2.809221 -2.523257  \n",
       "23 -0.389480  0.093041  0.176263  0.844495 -0.485920  1.027265  \n",
       "24 -0.518007 -0.775438  1.138622  0.140067  0.596830  2.305128  \n",
       "\n",
       "[25 rows x 100 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mags_df = pd.DataFrame (list_of_mag_vectors)\n",
    "mags_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "overhead-poison",
   "metadata": {},
   "source": [
    "---\n",
    "## Clustering\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "integral-qatar",
   "metadata": {},
   "source": [
    "### 1. K-means"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "dramatic-tuning",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>k_range</th>\n",
       "      <th>k_scores</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>-2854.137989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>-669.942820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>-310.319632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>-240.240229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>-179.335104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>-149.289063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>-131.255782</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>-104.342716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>-88.957743</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>-73.295911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>11</td>\n",
       "      <td>-57.250019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>12</td>\n",
       "      <td>-48.431060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>13</td>\n",
       "      <td>-40.247428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>14</td>\n",
       "      <td>-31.251658</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    k_range     k_scores\n",
       "0         1 -2854.137989\n",
       "1         2  -669.942820\n",
       "2         3  -310.319632\n",
       "3         4  -240.240229\n",
       "4         5  -179.335104\n",
       "5         6  -149.289063\n",
       "6         7  -131.255782\n",
       "7         8  -104.342716\n",
       "8         9   -88.957743\n",
       "9        10   -73.295911\n",
       "10       11   -57.250019\n",
       "11       12   -48.431060\n",
       "12       13   -40.247428\n",
       "13       14   -31.251658"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k_range_end = 15 # Usually it is sqrt(# of mags)\n",
    "\n",
    "k_range = range(1, k_range_end)\n",
    "\n",
    "k_means_models = [KMeans (n_clusters = i, random_state = SEED) for i in k_range]\n",
    "k_scores = [k_mean_model.fit(mags_df).score(mags_df) for k_mean_model in k_means_model]\n",
    "k_data = pd.DataFrame ({'k_range':k_range, 'k_scores':k_scores})\n",
    "k_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "closed-video",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-94937833c4b047dabba10a9befe889ea\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-94937833c4b047dabba10a9befe889ea\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-94937833c4b047dabba10a9befe889ea\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.8.1?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function loadScript(lib) {\n",
       "      return new Promise(function(resolve, reject) {\n",
       "        var s = document.createElement('script');\n",
       "        s.src = paths[lib];\n",
       "        s.async = true;\n",
       "        s.onload = () => resolve(paths[lib]);\n",
       "        s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "        document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "      });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else if (typeof vegaEmbed === \"function\") {\n",
       "      displayChart(vegaEmbed);\n",
       "    } else {\n",
       "      loadScript(\"vega\")\n",
       "        .then(() => loadScript(\"vega-lite\"))\n",
       "        .then(() => loadScript(\"vega-embed\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}}, \"data\": {\"name\": \"data-74216e6ede49490d4634d48060309b80\"}, \"mark\": \"line\", \"encoding\": {\"x\": {\"type\": \"quantitative\", \"field\": \"k_range\"}, \"y\": {\"type\": \"quantitative\", \"field\": \"k_scores\"}}, \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.8.1.json\", \"datasets\": {\"data-74216e6ede49490d4634d48060309b80\": [{\"k_range\": 1, \"k_scores\": -2854.137989443729}, {\"k_range\": 2, \"k_scores\": -669.9428200799358}, {\"k_range\": 3, \"k_scores\": -310.3196319867131}, {\"k_range\": 4, \"k_scores\": -240.24022932432987}, {\"k_range\": 5, \"k_scores\": -179.3351044126549}, {\"k_range\": 6, \"k_scores\": -149.2890630987335}, {\"k_range\": 7, \"k_scores\": -131.25578171955}, {\"k_range\": 8, \"k_scores\": -104.3427156787668}, {\"k_range\": 9, \"k_scores\": -88.95774326124139}, {\"k_range\": 10, \"k_scores\": -73.29591071508888}, {\"k_range\": 11, \"k_scores\": -57.25001911404887}, {\"k_range\": 12, \"k_scores\": -48.43105951260313}, {\"k_range\": 13, \"k_scores\": -40.24742754347339}, {\"k_range\": 14, \"k_scores\": -31.25165828575502}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.Chart(...)"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k_num_chart = alt.Chart(data = k_data).mark_line().encode(\n",
    "    alt.X ('k_range:Q'),\n",
    "    alt.Y ('k_scores:Q')\n",
    ")\n",
    "\n",
    "k_num_chart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "clear-donna",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "KMeans(n_clusters=7, random_state=42)"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# We can see from the chart above that 6 or 7 clusters are optimal for this task (where END = 25 MAGs)\n",
    "num_of_clusters = 7\n",
    "\n",
    "k_means_model = KMeans (n_clusters = num_of_clusters, random_state = SEED)\n",
    "k_means_model.fit(mags_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "excited-bunch",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([4, 5, 6, 5, 5, 1, 2, 2, 0, 0, 0, 4, 0, 1, 1, 1, 5, 1, 1, 4, 0, 5,\n",
       "       2, 3, 1], dtype=int32)"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k_means_predicted = k_means_model.predict(mags_df)\n",
    "k_means_predicted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "automatic-kernel",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# Next: try kernel k-means clustering\n",
    "# Next: find evaluation methods\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "million-custody",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "sunset-aquarium",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dirty-briefing",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
