{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "acceptable-mystery",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import altair as alt\n",
    "import os, random\n",
    "from Bio import SeqIO\n",
    "import gensim\n",
    "from gensim.models import Word2Vec\n",
    "from sklearn.cluster import KMeans, DBSCAN\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn import preprocessing, model_selection\n",
    "import altair as alt\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "amateur-sunset",
   "metadata": {},
   "source": [
    "## Defining paths for each and every omic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "moved-prefix",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_root_data = os.path.join ('..', 'Data', 'Extracted', 'Databases')\n",
    "path_all_fasta = os.path.join (path_root_data, 'fasta_files', 'AllBins')\n",
    "path_normalised_metabolomics = os.path.join (path_root_data, 'Metabolomics', 'Normalised_Tables')\n",
    "path_model_save_root = 'Saved_models'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "interracial-consumer",
   "metadata": {},
   "outputs": [],
   "source": [
    "num_of_mags = len([i for i in os.listdir(path_all_fasta) if (i[-2:] == 'fa')])\n",
    "SEED = 42\n",
    "END = 25\n",
    "num_of_workers = 3\n",
    "random.seed(SEED)\n",
    "np.random.seed(SEED)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "forbidden-intervention",
   "metadata": {},
   "source": [
    "---\n",
    "## MAG-related functions\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "rising-opening",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function that splits each genome into k-mers thus creating even longer sentence (MAG)\n",
    "# It returns tokenized genome i.e. [kmer, kmer,...]\n",
    "def split_genome (genome, k = 5):\n",
    "    new_genome = []\n",
    "    n = len(genome)\n",
    "    \n",
    "    if n-k <=0:\n",
    "        return genome\n",
    "    else:\n",
    "        for i in range(n-k):\n",
    "            new_genome.append(genome[i:i+k])\n",
    "        \n",
    "        return new_genome\n",
    "\n",
    "\n",
    "def vectorize_one_mag (one_mag, w2v_model):\n",
    "    \n",
    "    # We have to generate vectors for each word in one MAG and then create vector representation of that MAG\n",
    "    # by averaging vectors of its words\n",
    "    zero_vector = np.zeros(w2v_model.vector_size)\n",
    "    word_vectors = []\n",
    "    one_mag_vector = []\n",
    "    \n",
    "    for sentence in one_mag:\n",
    "        for word in sentence:\n",
    "            if word in w2v_model.wv:\n",
    "                try:\n",
    "                    word_vectors.append(w2v_model.wv[word])\n",
    "                except KeyError:\n",
    "                    print ('Key Error')\n",
    "                    continue\n",
    "    \n",
    "    if word_vectors:\n",
    "        word_vectors = np.asarray(word_vectors)\n",
    "        one_mag_vector = word_vectors.mean (axis=0)\n",
    "    \n",
    "    else:\n",
    "        one_mag_vector = zero_vector\n",
    "    \n",
    "    return one_mag_vector\n",
    "\n",
    "\n",
    "# Function that vectorizes a MAG (document) with a pretrained word2vec model. It returns vector representation of a given MAG\n",
    "# Vectorization is done by averaging word (k-mer) vectors for the whole document (MAG)\n",
    "def vectorize_mags (w2v_model, end = 25):\n",
    "    \n",
    "    print ('Vectorizing MAGs')\n",
    "    \n",
    "    fasta_files = [i for i in os.listdir(path_all_fasta) if (i[-2:] == 'fa')]\n",
    "    list_of_mag_vectors = []\n",
    "    \n",
    "    # This was done so that I could work with first 'end' FASTA files only. Otherwise, I should just remove: i, and enumerate\n",
    "    for i, fasta_file_name in enumerate(fasta_files):\n",
    "        \n",
    "        if i == end:\n",
    "            break\n",
    "        \n",
    "        else:\n",
    "            with open(os.path.join(path_all_fasta, fasta_file_name), 'r') as input_file:\n",
    "                \n",
    "                one_mag = []\n",
    "                for fasta_string in SeqIO.parse(input_file, \"fasta\"):\n",
    "                    \n",
    "                    # Get kmers of a genome and create a sentence (list of words)\n",
    "                    temp_kmers = split_genome (str(fasta_string.seq))\n",
    "                    \n",
    "                    # Create a document (list of sentences)\n",
    "                    one_mag.append(temp_kmers)\n",
    "                    \n",
    "                # Vectorizing MAGs one by one\n",
    "                list_of_mag_vectors.append (vectorize_one_mag (one_mag, w2v_model))\n",
    "    \n",
    "    print ('Finished vectorizing')\n",
    "    \n",
    "    return list_of_mag_vectors\n",
    "    \n",
    "    \n",
    "\n",
    "# If one wants to import MAGs in order to vectorize them, one should use start argument in order to skip first 'start' MAGs\n",
    "# If one wants to import MAGs to train word2vec model, one should use only end argument, so that first 'end' MAGs are used for training\n",
    "# Todo: Implement randomisation in picking MAGs for training, and don't use first 'start' MAGs for training\n",
    "# (create list of indexes from 0 to 1364, use sklearn split train test, then traverse directory and use only MAGs with indexes in train for training w2v)\n",
    "def import_mags_and_build_model (end = 25, path_all_fasta = path_all_fasta):\n",
    "    \n",
    "    print ('Importing MAGs')\n",
    "    \n",
    "    # There are 1364 MAGs enclosed in FASTA files\n",
    "    # I have to traverse every FASTA file, and in each file every sequence\n",
    "    \n",
    "    fasta_files = [i for i in os.listdir(path_all_fasta) if (i[-2:] == 'fa')]\n",
    "    \n",
    "    # This was done so that I could work with first 100 FASTA files only. Otherwise, I should just remove: i, and enumerate\n",
    "    for i, fasta_file_name in enumerate(fasta_files):\n",
    "        \n",
    "        if i == end:\n",
    "            break\n",
    "        \n",
    "        else:\n",
    "            with open(os.path.join(path_all_fasta, fasta_file_name), 'r') as input_file:\n",
    "                \n",
    "                one_mag = []\n",
    "                for fasta_string in SeqIO.parse(input_file, \"fasta\"):\n",
    "                    \n",
    "                    # Get kmers of a genome and create a sentence (list of words)\n",
    "                    temp_kmers = split_genome (str(fasta_string.seq))\n",
    "                    \n",
    "                    # Create a document (list of sentences)\n",
    "                    one_mag.append(temp_kmers)\n",
    "                    \n",
    "                # If we do not have a model, we build one\n",
    "                if i == 0:\n",
    "                    print ('Building w2v model')\n",
    "                    # We build our model on the first MAG\n",
    "                    w2v_model = Word2Vec (sentences = one_mag, size = 100, workers = num_of_workers, seed=SEED)\n",
    "                \n",
    "                # Else we just expand its vocabulary\n",
    "                else:\n",
    "                    # Now we expand our vocabulary\n",
    "                    w2v_model.build_vocab (one_mag, update = True)\n",
    "                    \n",
    "    print ('Finished building')\n",
    "    \n",
    "    return w2v_model\n",
    "\n",
    "\n",
    "def train_model (w2v_model, epochs, end = 25):\n",
    "    \n",
    "    print ('Starting model training')\n",
    "    \n",
    "    # There are 1364 MAGs enclosed in FASTA files\n",
    "    # I have to traverse every FASTA file, and in each file every sequence\n",
    "    \n",
    "    fasta_files = [i for i in os.listdir(path_all_fasta) if (i[-2:] == 'fa')]\n",
    "    \n",
    "    # This was done so that I could work with first 100 FASTA files only. Otherwise, I should just remove: i, and enumerate\n",
    "    for i, fasta_file_name in enumerate(fasta_files):\n",
    "        \n",
    "        if i == end:\n",
    "            break\n",
    "        \n",
    "        else:\n",
    "            with open(os.path.join(path_all_fasta, fasta_file_name), 'r') as input_file:\n",
    "                \n",
    "                one_mag = []\n",
    "                for fasta_string in SeqIO.parse(input_file, \"fasta\"):\n",
    "                    \n",
    "                    # Get kmers of a genome and create a sentence (list of words)\n",
    "                    temp_kmers = split_genome (str(fasta_string.seq))\n",
    "                    \n",
    "                    # Create a document (list of sentences)\n",
    "                    one_mag.append(temp_kmers)\n",
    "                    \n",
    "                \n",
    "                w2v_model.train (one_mag, total_examples = w2v_model.corpus_count, epochs = epochs)\n",
    "                    \n",
    "    print ('Model training finished')\n",
    "    \n",
    "    return w2v_model\n",
    "\n",
    "\n",
    "def visualize_with_pca (data, labels, centers):\n",
    "    \n",
    "    pca_model = PCA (n_components = 2, random_state = SEED)\n",
    "    data_transformed = pca_model.fit_transform (data)\n",
    "    \n",
    "    data_transformed = pd.DataFrame (data_transformed)\n",
    "    data_transformed.columns = ['PC_1', 'PC_2']\n",
    "    data_transformed['Labels'] = labels\n",
    "    \n",
    "    chart_data = alt.Chart(data_transformed).mark_circle().encode(\n",
    "        alt.X ('PC_1:Q'),\n",
    "        alt.Y ('PC_2:Q'),\n",
    "        alt.Color ('Labels:N'),\n",
    "    )\n",
    "    \n",
    "    # This means we are visualising centroids from k_means (there are less centroids that data points)\n",
    "    if data.shape[0] != centers.shape[0]:\n",
    "        \n",
    "        centers_transformed = pca_model.fit_transform (centers)\n",
    "        centers_transformed = pd.DataFrame (centers_transformed)\n",
    "        centers_transformed.columns = ['PC_1', 'PC_2']\n",
    "        \n",
    "        chart_centers = alt.Chart(centers_transformed).mark_point(shape = 'diamond', color = 'black', size = 50, opacity = 0.7).encode(\n",
    "            alt.X ('PC_1:Q'),\n",
    "            alt.Y ('PC_2:Q'),\n",
    "        )\n",
    "        \n",
    "        return chart_data + chart_centers\n",
    "    \n",
    "    # For DBSCAN there are no centroids\n",
    "    else:\n",
    "        return chart_data\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "steady-conservation",
   "metadata": {},
   "source": [
    "---\n",
    "## Importing Metabolomic data\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "informational-hours",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Metabolite</th>\n",
       "      <th>tp</th>\n",
       "      <th>date</th>\n",
       "      <th>type</th>\n",
       "      <th>known_type</th>\n",
       "      <th>means</th>\n",
       "      <th>medians</th>\n",
       "      <th>sds</th>\n",
       "      <th>N</th>\n",
       "      <th>se</th>\n",
       "      <th>ci</th>\n",
       "      <th>type2</th>\n",
       "      <th>measurement</th>\n",
       "      <th>KEGG.Compound.ID</th>\n",
       "      <th>Chebi.Name</th>\n",
       "      <th>Chebi.Name_combined</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Phosphoric_acid_3TMS</td>\n",
       "      <td>D1</td>\n",
       "      <td>2012-04-17</td>\n",
       "      <td>snp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.757113</td>\n",
       "      <td>0.737943</td>\n",
       "      <td>0.098650</td>\n",
       "      <td>3</td>\n",
       "      <td>0.056956</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>C00009</td>\n",
       "      <td>phosphoric acid</td>\n",
       "      <td>phosphoric acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Phosphoric_acid_3TMS</td>\n",
       "      <td>D10</td>\n",
       "      <td>2012-01-25</td>\n",
       "      <td>snp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.720935</td>\n",
       "      <td>0.873509</td>\n",
       "      <td>0.266593</td>\n",
       "      <td>3</td>\n",
       "      <td>0.153918</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>C00009</td>\n",
       "      <td>phosphoric acid</td>\n",
       "      <td>phosphoric acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Phosphoric_acid_3TMS</td>\n",
       "      <td>D11</td>\n",
       "      <td>2012-03-22</td>\n",
       "      <td>snp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.625047</td>\n",
       "      <td>0.652268</td>\n",
       "      <td>0.057820</td>\n",
       "      <td>3</td>\n",
       "      <td>0.033382</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>C00009</td>\n",
       "      <td>phosphoric acid</td>\n",
       "      <td>phosphoric acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Phosphoric_acid_3TMS</td>\n",
       "      <td>D12</td>\n",
       "      <td>2012-01-19</td>\n",
       "      <td>snp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.964956</td>\n",
       "      <td>0.941878</td>\n",
       "      <td>0.053754</td>\n",
       "      <td>3</td>\n",
       "      <td>0.031035</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>C00009</td>\n",
       "      <td>phosphoric acid</td>\n",
       "      <td>phosphoric acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>!Phosphoric_acid_3TMS</td>\n",
       "      <td>D13</td>\n",
       "      <td>2011-05-13</td>\n",
       "      <td>snp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.963984</td>\n",
       "      <td>0.974479</td>\n",
       "      <td>0.071659</td>\n",
       "      <td>3</td>\n",
       "      <td>0.041373</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>C00009</td>\n",
       "      <td>phosphoric acid</td>\n",
       "      <td>phosphoric acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18865</th>\n",
       "      <td>Valine_2TMS</td>\n",
       "      <td>D51</td>\n",
       "      <td>2011-08-11</td>\n",
       "      <td>bp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.704821</td>\n",
       "      <td>0.478251</td>\n",
       "      <td>0.428368</td>\n",
       "      <td>3</td>\n",
       "      <td>0.247319</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C16436,C06417,C00183</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18866</th>\n",
       "      <td>Valine_2TMS</td>\n",
       "      <td>D6</td>\n",
       "      <td>2011-06-03</td>\n",
       "      <td>bp</td>\n",
       "      <td>known</td>\n",
       "      <td>1.085888</td>\n",
       "      <td>1.197471</td>\n",
       "      <td>0.487484</td>\n",
       "      <td>3</td>\n",
       "      <td>0.281449</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C16436,C06417,C00183</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18867</th>\n",
       "      <td>Valine_2TMS</td>\n",
       "      <td>D7</td>\n",
       "      <td>2012-04-04</td>\n",
       "      <td>bp</td>\n",
       "      <td>known</td>\n",
       "      <td>2.958023</td>\n",
       "      <td>2.813145</td>\n",
       "      <td>0.341213</td>\n",
       "      <td>3</td>\n",
       "      <td>0.197000</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C16436,C06417,C00183</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18868</th>\n",
       "      <td>Valine_2TMS</td>\n",
       "      <td>D8</td>\n",
       "      <td>2012-02-08</td>\n",
       "      <td>bp</td>\n",
       "      <td>known</td>\n",
       "      <td>2.515980</td>\n",
       "      <td>2.322248</td>\n",
       "      <td>0.618556</td>\n",
       "      <td>3</td>\n",
       "      <td>0.357123</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C16436,C06417,C00183</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18869</th>\n",
       "      <td>Valine_2TMS</td>\n",
       "      <td>D9</td>\n",
       "      <td>2011-12-28</td>\n",
       "      <td>bp</td>\n",
       "      <td>known</td>\n",
       "      <td>2.356989</td>\n",
       "      <td>2.280757</td>\n",
       "      <td>0.264918</td>\n",
       "      <td>3</td>\n",
       "      <td>0.152950</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C16436,C06417,C00183</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18870 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Metabolite   tp        date type known_type     means  \\\n",
       "0      !Phosphoric_acid_3TMS   D1  2012-04-17  snp      known  0.757113   \n",
       "1      !Phosphoric_acid_3TMS  D10  2012-01-25  snp      known  0.720935   \n",
       "2      !Phosphoric_acid_3TMS  D11  2012-03-22  snp      known  0.625047   \n",
       "3      !Phosphoric_acid_3TMS  D12  2012-01-19  snp      known  0.964956   \n",
       "4      !Phosphoric_acid_3TMS  D13  2011-05-13  snp      known  0.963984   \n",
       "...                      ...  ...         ...  ...        ...       ...   \n",
       "18865            Valine_2TMS  D51  2011-08-11   bp      known  0.704821   \n",
       "18866            Valine_2TMS   D6  2011-06-03   bp      known  1.085888   \n",
       "18867            Valine_2TMS   D7  2012-04-04   bp      known  2.958023   \n",
       "18868            Valine_2TMS   D8  2012-02-08   bp      known  2.515980   \n",
       "18869            Valine_2TMS   D9  2011-12-28   bp      known  2.356989   \n",
       "\n",
       "        medians       sds  N        se        ci          type2 measurement  \\\n",
       "0      0.737943  0.098650  3  0.056956  4.302653  extracellular    nonpolar   \n",
       "1      0.873509  0.266593  3  0.153918  4.302653  extracellular    nonpolar   \n",
       "2      0.652268  0.057820  3  0.033382  4.302653  extracellular    nonpolar   \n",
       "3      0.941878  0.053754  3  0.031035  4.302653  extracellular    nonpolar   \n",
       "4      0.974479  0.071659  3  0.041373  4.302653  extracellular    nonpolar   \n",
       "...         ...       ... ..       ...       ...            ...         ...   \n",
       "18865  0.478251  0.428368  3  0.247319  4.302653  intracellular       polar   \n",
       "18866  1.197471  0.487484  3  0.281449  4.302653  intracellular       polar   \n",
       "18867  2.813145  0.341213  3  0.197000  4.302653  intracellular       polar   \n",
       "18868  2.322248  0.618556  3  0.357123  4.302653  intracellular       polar   \n",
       "18869  2.280757  0.264918  3  0.152950  4.302653  intracellular       polar   \n",
       "\n",
       "           KEGG.Compound.ID                Chebi.Name  \\\n",
       "0                    C00009           phosphoric acid   \n",
       "1                    C00009           phosphoric acid   \n",
       "2                    C00009           phosphoric acid   \n",
       "3                    C00009           phosphoric acid   \n",
       "4                    C00009           phosphoric acid   \n",
       "...                     ...                       ...   \n",
       "18865  C16436,C06417,C00183  valine;D-valine;L-valine   \n",
       "18866  C16436,C06417,C00183  valine;D-valine;L-valine   \n",
       "18867  C16436,C06417,C00183  valine;D-valine;L-valine   \n",
       "18868  C16436,C06417,C00183  valine;D-valine;L-valine   \n",
       "18869  C16436,C06417,C00183  valine;D-valine;L-valine   \n",
       "\n",
       "            Chebi.Name_combined  \n",
       "0               phosphoric acid  \n",
       "1               phosphoric acid  \n",
       "2               phosphoric acid  \n",
       "3               phosphoric acid  \n",
       "4               phosphoric acid  \n",
       "...                         ...  \n",
       "18865  valine;D-valine;L-valine  \n",
       "18866  valine;D-valine;L-valine  \n",
       "18867  valine;D-valine;L-valine  \n",
       "18868  valine;D-valine;L-valine  \n",
       "18869  valine;D-valine;L-valine  \n",
       "\n",
       "[18870 rows x 16 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metabolomics_file_name = os.path.join(path_normalised_metabolomics, os.listdir(path_normalised_metabolomics)[0])\n",
    "metabolomics_df = pd.read_csv (metabolomics_file_name, delimiter = '\\t')\n",
    "metabolomics_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "moved-device",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>date</th>\n",
       "      <th>Metabolite</th>\n",
       "      <th>tp</th>\n",
       "      <th>type</th>\n",
       "      <th>known_type</th>\n",
       "      <th>means</th>\n",
       "      <th>medians</th>\n",
       "      <th>sds</th>\n",
       "      <th>N</th>\n",
       "      <th>se</th>\n",
       "      <th>ci</th>\n",
       "      <th>type2</th>\n",
       "      <th>measurement</th>\n",
       "      <th>KEGG.Compound.ID</th>\n",
       "      <th>Chebi.Name</th>\n",
       "      <th>Chebi.Name_combined</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>Leucine_2TMS</td>\n",
       "      <td>D37</td>\n",
       "      <td>sp</td>\n",
       "      <td>known</td>\n",
       "      <td>1.440122</td>\n",
       "      <td>1.489979</td>\n",
       "      <td>0.146440</td>\n",
       "      <td>3</td>\n",
       "      <td>0.084547</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C16439,C00123,C01570</td>\n",
       "      <td>leucine;L-leucine;D-leucine</td>\n",
       "      <td>leucine;L-leucine;D-leucine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>No match: 1181.78_EM_SP_D1_1_223</td>\n",
       "      <td>D37</td>\n",
       "      <td>sp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>3.778469</td>\n",
       "      <td>3.801417</td>\n",
       "      <td>1.924377</td>\n",
       "      <td>3</td>\n",
       "      <td>1.111040</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>unknown#emu_WW_2144.25</td>\n",
       "      <td>D37</td>\n",
       "      <td>bnp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.949545</td>\n",
       "      <td>0.947209</td>\n",
       "      <td>0.065422</td>\n",
       "      <td>3</td>\n",
       "      <td>0.037771</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>Unknown#bth_pae_001</td>\n",
       "      <td>D37</td>\n",
       "      <td>bp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2.095116</td>\n",
       "      <td>1.809891</td>\n",
       "      <td>1.916896</td>\n",
       "      <td>3</td>\n",
       "      <td>1.106721</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>No match: 2220.34_EM_BP_D1_1_192</td>\n",
       "      <td>D37</td>\n",
       "      <td>bp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.631643</td>\n",
       "      <td>1.571125</td>\n",
       "      <td>0.168501</td>\n",
       "      <td>3</td>\n",
       "      <td>0.097284</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18865</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>Octadecanol_1TMS</td>\n",
       "      <td>D48</td>\n",
       "      <td>bnp</td>\n",
       "      <td>known</td>\n",
       "      <td>1.364105</td>\n",
       "      <td>1.457780</td>\n",
       "      <td>0.170311</td>\n",
       "      <td>3</td>\n",
       "      <td>0.098329</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>D01924</td>\n",
       "      <td>octadecan-1-ol</td>\n",
       "      <td>octadecan-1-ol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18866</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>No match: 1447.93_EM_BP_D2_1_57</td>\n",
       "      <td>D48</td>\n",
       "      <td>bp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>4.815774</td>\n",
       "      <td>1.160165</td>\n",
       "      <td>6.835474</td>\n",
       "      <td>3</td>\n",
       "      <td>3.946463</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18867</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>Lactulose_8TMS</td>\n",
       "      <td>D48</td>\n",
       "      <td>bp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.697489</td>\n",
       "      <td>0.684339</td>\n",
       "      <td>0.049632</td>\n",
       "      <td>3</td>\n",
       "      <td>0.028655</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C07064</td>\n",
       "      <td>lactulose</td>\n",
       "      <td>lactulose</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18868</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>No match: 2004.78_EM_BNP_D10_1_205</td>\n",
       "      <td>D48</td>\n",
       "      <td>bnp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.785996</td>\n",
       "      <td>0.744529</td>\n",
       "      <td>0.139901</td>\n",
       "      <td>3</td>\n",
       "      <td>0.080772</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18869</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>2-Oxobutanoic_acid_1_1MeOX_1TMS</td>\n",
       "      <td>D48</td>\n",
       "      <td>sp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.755565</td>\n",
       "      <td>0.709972</td>\n",
       "      <td>0.149214</td>\n",
       "      <td>3</td>\n",
       "      <td>0.086149</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C00109</td>\n",
       "      <td>2-oxobutanoic acid</td>\n",
       "      <td>2-oxobutanoic acid</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18870 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            date                          Metabolite   tp type known_type  \\\n",
       "0     2011-03-21                        Leucine_2TMS  D37   sp      known   \n",
       "1     2011-03-21    No match: 1181.78_EM_SP_D1_1_223  D37   sp    unknown   \n",
       "2     2011-03-21              unknown#emu_WW_2144.25  D37  bnp    unknown   \n",
       "3     2011-03-21                 Unknown#bth_pae_001  D37   bp    unknown   \n",
       "4     2011-03-21    No match: 2220.34_EM_BP_D1_1_192  D37   bp    unknown   \n",
       "...          ...                                 ...  ...  ...        ...   \n",
       "18865 2012-05-03                    Octadecanol_1TMS  D48  bnp      known   \n",
       "18866 2012-05-03     No match: 1447.93_EM_BP_D2_1_57  D48   bp    unknown   \n",
       "18867 2012-05-03                      Lactulose_8TMS  D48   bp      known   \n",
       "18868 2012-05-03  No match: 2004.78_EM_BNP_D10_1_205  D48  bnp    unknown   \n",
       "18869 2012-05-03     2-Oxobutanoic_acid_1_1MeOX_1TMS  D48   sp      known   \n",
       "\n",
       "          means   medians       sds  N        se        ci          type2  \\\n",
       "0      1.440122  1.489979  0.146440  3  0.084547  4.302653  extracellular   \n",
       "1      3.778469  3.801417  1.924377  3  1.111040  4.302653  extracellular   \n",
       "2      0.949545  0.947209  0.065422  3  0.037771  4.302653  intracellular   \n",
       "3      2.095116  1.809891  1.916896  3  1.106721  4.302653  intracellular   \n",
       "4      1.631643  1.571125  0.168501  3  0.097284  4.302653  intracellular   \n",
       "...         ...       ...       ... ..       ...       ...            ...   \n",
       "18865  1.364105  1.457780  0.170311  3  0.098329  4.302653  intracellular   \n",
       "18866  4.815774  1.160165  6.835474  3  3.946463  4.302653  intracellular   \n",
       "18867  0.697489  0.684339  0.049632  3  0.028655  4.302653  intracellular   \n",
       "18868  0.785996  0.744529  0.139901  3  0.080772  4.302653  intracellular   \n",
       "18869  0.755565  0.709972  0.149214  3  0.086149  4.302653  extracellular   \n",
       "\n",
       "      measurement      KEGG.Compound.ID                   Chebi.Name  \\\n",
       "0           polar  C16439,C00123,C01570  leucine;L-leucine;D-leucine   \n",
       "1           polar                   NaN                          NaN   \n",
       "2        nonpolar                   NaN                          NaN   \n",
       "3           polar                   NaN                          NaN   \n",
       "4           polar                   NaN                          NaN   \n",
       "...           ...                   ...                          ...   \n",
       "18865    nonpolar                D01924               octadecan-1-ol   \n",
       "18866       polar                   NaN                          NaN   \n",
       "18867       polar                C07064                    lactulose   \n",
       "18868    nonpolar                   NaN                          NaN   \n",
       "18869       polar                C00109           2-oxobutanoic acid   \n",
       "\n",
       "               Chebi.Name_combined  \n",
       "0      leucine;L-leucine;D-leucine  \n",
       "1                              NaN  \n",
       "2                              NaN  \n",
       "3                              NaN  \n",
       "4                              NaN  \n",
       "...                            ...  \n",
       "18865               octadecan-1-ol  \n",
       "18866                          NaN  \n",
       "18867                    lactulose  \n",
       "18868                          NaN  \n",
       "18869           2-oxobutanoic acid  \n",
       "\n",
       "[18870 rows x 16 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metabolomics_df['date'] = pd.to_datetime(metabolomics_df['date'])\n",
    "metabolomics_df.insert (0, 'date', metabolomics_df.pop('date'))\n",
    "metabolomics_df.sort_values ('date', inplace = True, ignore_index = True)\n",
    "metabolomics_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "interior-gazette",
   "metadata": {},
   "source": [
    "## MAG examination"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "protective-kansas",
   "metadata": {},
   "source": [
    "To do:\n",
    "1. Make kmers of some length, out of every word in every sentence\n",
    "2. Encode every sentence with Word2Vec and CBOW (because it's faster than skip-gram)\n",
    "3. Reduce dimensionality with PCA or with autoencoder\n",
    "    > Maybe it is good to do the first 3 steps while importing, for each and every sentence. That way, I might use less memory\n",
    "4. Cluster similar sentences (MAGs) to get rMAGs\n",
    "\n",
    "> Important: this is not temporal data, since there are no timestamps!\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "disturbed-onion",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing MAGs\n",
      "Building w2v model\n",
      "Finished building\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# FOR CLUSTERING I SHOULD CREATE A DATAFRAME WITH MAGs INDEXES AND THEIR VECTOR REPRESENTATIONS\n",
    "\n",
    "final_model = import_mags_and_build_model (end = END, path_all_fasta = path_all_fasta)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "disciplinary-constitution",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('CCGGC', 0.5773573517799377),\n",
       " ('TCGGC', 0.5635622143745422),\n",
       " ('ACGGA', 0.5475510358810425),\n",
       " ('GCGGC', 0.5233906507492065),\n",
       " ('ACGGT', 0.4797680675983429),\n",
       " ('ACGGG', 0.47880542278289795),\n",
       " ('ACGAC', 0.29950520396232605),\n",
       " ('ACGTT', 0.2790815234184265),\n",
       " ('ACGAT', 0.26410460472106934),\n",
       " ('AACGA', 0.26403582096099854)]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_model.wv.most_similar('ACGGC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "christian-assistant",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting model training\n",
      "Model training finished\n"
     ]
    }
   ],
   "source": [
    "# Train model. It tooks ~10 minutes for END = 25 amount of MAGs\n",
    "epochs = 10\n",
    "\n",
    "final_model = train_model (final_model, epochs = epochs, end = END)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "sacred-federation",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_model.wv.save_word2vec_format(os.path.join (path_model_save_root, 'model_25.bin'), binary=True) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "medieval-williams",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "loved-interface",
   "metadata": {},
   "source": [
    "Now I should vectorize documents with this model. For further use, I could save this model's weights, and use it to vectorize all mags. That would take a lot, but every MAG will have its vector representation\n",
    "> This could be done by importing one MAG at a time, then tokenizing it (like before), then getting vector representations of that MAG's sentences (genomes) and then finding the vector representation of the whole MAG (document). If I do that for one MAG at a time, There is no need to worry about memory\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "scientific-audience",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Vectorizing MAGs\n",
      "Finished vectorizing\n"
     ]
    }
   ],
   "source": [
    "list_of_mag_vectors = vectorize_mags (final_model, end = END)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "spoken-saturn",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[array([ 1.1176913e+00, -1.2851473e+00,  1.2977009e+00, -3.5607186e-01,\n",
       "         1.1518677e+00, -1.4784318e+00,  6.9590878e-01, -2.0498614e+00,\n",
       "         7.7519298e-01,  9.5224845e-01,  8.7902375e-02, -8.6440951e-01,\n",
       "        -1.4811026e-01,  2.1250219e+00, -6.5102720e-01,  6.4006710e-01,\n",
       "        -7.7712810e-01, -6.8890983e-01, -1.7285050e+00,  3.5550112e-01,\n",
       "         7.3951113e-01, -1.9243872e+00,  1.3869661e+00,  1.4437793e-01,\n",
       "        -1.2514910e+00,  4.9194074e-01,  7.0413029e-01, -2.1186869e-01,\n",
       "         1.1919310e+00,  4.0589210e-01, -1.1578156e+00, -1.0346288e-01,\n",
       "        -4.7114155e-01,  2.5510933e+00,  8.9802682e-01, -1.0469451e+00,\n",
       "        -4.5550957e-01,  5.2043205e-01, -3.8725996e-01,  3.1858125e-01,\n",
       "         8.2555884e-01, -1.8461175e+00,  5.8819723e-01, -6.8201619e-01,\n",
       "         2.8457314e-01, -2.1281188e+00,  1.8199693e+00, -6.0000217e-01,\n",
       "        -8.4681815e-01, -1.4194238e+00, -3.9664999e-01,  1.5638909e-01,\n",
       "         1.8452522e+00, -3.9363700e-01,  9.9437870e-04,  1.1115664e+00,\n",
       "        -8.3923858e-01, -1.3796260e+00, -3.5114899e-01, -8.9791693e-02,\n",
       "         1.2229146e+00, -5.0225228e-01, -5.1954752e-01, -9.7626305e-01,\n",
       "         5.1489139e-01,  9.6944278e-01, -3.5591587e-01, -3.1405132e+00,\n",
       "         5.2186018e-01, -2.4086967e+00,  3.2519007e-01,  1.7473051e-01,\n",
       "         1.2470962e+00,  1.1753468e-01, -4.2037788e-01,  3.6502448e-01,\n",
       "         1.3502605e-01,  4.7087383e-01, -1.2364941e+00, -4.3196341e-01,\n",
       "         5.6995934e-01, -7.3345935e-01, -1.2896657e+00,  4.5480818e-01,\n",
       "         3.5474357e-01, -9.3029511e-01,  3.6287594e-01,  1.5294808e+00,\n",
       "         1.5644047e+00, -2.1622639e+00,  1.5480717e+00, -1.2125455e+00,\n",
       "         1.2973577e+00,  3.0430534e-01, -8.8180378e-02,  7.4170303e-01,\n",
       "        -2.1921003e+00, -1.7600286e+00,  1.2927092e+00, -5.6631410e-01],\n",
       "       dtype=float32),\n",
       " array([-0.39745724, -0.5863701 , -2.1017823 , -0.9578395 , -0.8947163 ,\n",
       "         2.2527587 ,  0.17934464,  1.366534  , -0.14100094, -0.368006  ,\n",
       "         1.2022954 ,  0.23336607, -0.35584146, -1.4575847 , -0.3383674 ,\n",
       "        -1.3487175 ,  0.03054165,  0.6704843 ,  0.56563   , -0.6582459 ,\n",
       "        -0.30309078,  0.31679574,  0.63306063, -0.32028386,  0.5919644 ,\n",
       "         0.5847385 ,  0.2419244 , -0.55095273, -0.05648061,  0.799294  ,\n",
       "         0.6668926 , -1.3909849 , -0.8421477 ,  0.4151582 , -0.327073  ,\n",
       "         1.2033821 ,  0.8354061 ,  0.0832241 , -0.34794497,  0.09240235,\n",
       "        -1.4730088 ,  1.3817904 ,  0.5524177 ,  1.7613115 , -0.62893456,\n",
       "         1.5522633 , -1.7648458 ,  0.3729613 ,  0.7584396 ,  1.511729  ,\n",
       "        -0.6032976 ,  0.24214186, -1.704377  , -0.37426257,  0.6322615 ,\n",
       "        -1.7870628 ,  1.5927831 ,  0.9731983 ,  1.2930869 ,  0.47782254,\n",
       "        -1.4922181 ,  0.5212634 , -0.9407582 ,  2.4608393 ,  0.9967795 ,\n",
       "        -1.5149596 , -0.73958915,  2.5861137 , -0.79474854,  1.1340076 ,\n",
       "         0.7264857 , -0.5695081 , -1.6340992 ,  1.0300808 ,  0.99156946,\n",
       "         0.30827007,  2.2246892 , -0.2302448 ,  1.5955485 ,  0.77574486,\n",
       "         0.01802589, -0.07176805,  0.43433568, -0.8414009 , -0.30650863,\n",
       "         1.6139556 ,  0.41015342, -2.062111  , -0.78547466, -0.12728275,\n",
       "        -0.3241871 ,  0.45694226, -1.7233173 , -1.0898184 ,  1.471538  ,\n",
       "        -1.3320009 ,  2.7974143 ,  0.9621294 , -0.94937855,  0.02721914],\n",
       "       dtype=float32)]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_of_mag_vectors[0:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "comparable-dealer",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>90</th>\n",
       "      <th>91</th>\n",
       "      <th>92</th>\n",
       "      <th>93</th>\n",
       "      <th>94</th>\n",
       "      <th>95</th>\n",
       "      <th>96</th>\n",
       "      <th>97</th>\n",
       "      <th>98</th>\n",
       "      <th>99</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.117691</td>\n",
       "      <td>-1.285147</td>\n",
       "      <td>1.297701</td>\n",
       "      <td>-0.356072</td>\n",
       "      <td>1.151868</td>\n",
       "      <td>-1.478432</td>\n",
       "      <td>0.695909</td>\n",
       "      <td>-2.049861</td>\n",
       "      <td>0.775193</td>\n",
       "      <td>0.952248</td>\n",
       "      <td>...</td>\n",
       "      <td>1.548072</td>\n",
       "      <td>-1.212546</td>\n",
       "      <td>1.297358</td>\n",
       "      <td>0.304305</td>\n",
       "      <td>-0.088180</td>\n",
       "      <td>0.741703</td>\n",
       "      <td>-2.192100</td>\n",
       "      <td>-1.760029</td>\n",
       "      <td>1.292709</td>\n",
       "      <td>-0.566314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-0.397457</td>\n",
       "      <td>-0.586370</td>\n",
       "      <td>-2.101782</td>\n",
       "      <td>-0.957839</td>\n",
       "      <td>-0.894716</td>\n",
       "      <td>2.252759</td>\n",
       "      <td>0.179345</td>\n",
       "      <td>1.366534</td>\n",
       "      <td>-0.141001</td>\n",
       "      <td>-0.368006</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.324187</td>\n",
       "      <td>0.456942</td>\n",
       "      <td>-1.723317</td>\n",
       "      <td>-1.089818</td>\n",
       "      <td>1.471538</td>\n",
       "      <td>-1.332001</td>\n",
       "      <td>2.797414</td>\n",
       "      <td>0.962129</td>\n",
       "      <td>-0.949379</td>\n",
       "      <td>0.027219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.428277</td>\n",
       "      <td>-0.409841</td>\n",
       "      <td>1.821033</td>\n",
       "      <td>-0.632364</td>\n",
       "      <td>0.519591</td>\n",
       "      <td>-1.075682</td>\n",
       "      <td>0.467286</td>\n",
       "      <td>-2.106847</td>\n",
       "      <td>0.623872</td>\n",
       "      <td>-0.444501</td>\n",
       "      <td>...</td>\n",
       "      <td>0.977349</td>\n",
       "      <td>-2.612781</td>\n",
       "      <td>0.853364</td>\n",
       "      <td>1.045666</td>\n",
       "      <td>-0.377918</td>\n",
       "      <td>1.504324</td>\n",
       "      <td>-3.668192</td>\n",
       "      <td>-3.049183</td>\n",
       "      <td>1.911584</td>\n",
       "      <td>-1.719634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-0.807010</td>\n",
       "      <td>-1.132076</td>\n",
       "      <td>-2.040275</td>\n",
       "      <td>-1.018199</td>\n",
       "      <td>-1.241043</td>\n",
       "      <td>2.814131</td>\n",
       "      <td>-0.277895</td>\n",
       "      <td>1.702027</td>\n",
       "      <td>-0.216814</td>\n",
       "      <td>-0.389414</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.576729</td>\n",
       "      <td>0.361689</td>\n",
       "      <td>-1.794586</td>\n",
       "      <td>-0.989517</td>\n",
       "      <td>1.530060</td>\n",
       "      <td>-1.400325</td>\n",
       "      <td>3.109184</td>\n",
       "      <td>1.333641</td>\n",
       "      <td>-1.368727</td>\n",
       "      <td>-0.299508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-0.729850</td>\n",
       "      <td>-1.103235</td>\n",
       "      <td>-1.926998</td>\n",
       "      <td>-1.014075</td>\n",
       "      <td>-1.196003</td>\n",
       "      <td>2.693963</td>\n",
       "      <td>-0.228020</td>\n",
       "      <td>1.613018</td>\n",
       "      <td>-0.233582</td>\n",
       "      <td>-0.356466</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.552327</td>\n",
       "      <td>0.412861</td>\n",
       "      <td>-1.779991</td>\n",
       "      <td>-0.922350</td>\n",
       "      <td>1.528533</td>\n",
       "      <td>-1.356465</td>\n",
       "      <td>2.961123</td>\n",
       "      <td>1.244486</td>\n",
       "      <td>-1.314663</td>\n",
       "      <td>-0.330579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>-0.700547</td>\n",
       "      <td>-0.697610</td>\n",
       "      <td>-0.764258</td>\n",
       "      <td>-0.581024</td>\n",
       "      <td>-1.340819</td>\n",
       "      <td>2.218557</td>\n",
       "      <td>-0.651789</td>\n",
       "      <td>1.153820</td>\n",
       "      <td>-0.019014</td>\n",
       "      <td>-0.771810</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.663449</td>\n",
       "      <td>-0.050688</td>\n",
       "      <td>-1.609203</td>\n",
       "      <td>-0.787775</td>\n",
       "      <td>1.419916</td>\n",
       "      <td>-0.765434</td>\n",
       "      <td>1.982262</td>\n",
       "      <td>0.660961</td>\n",
       "      <td>-1.071012</td>\n",
       "      <td>-0.559409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1.882998</td>\n",
       "      <td>-0.430399</td>\n",
       "      <td>2.193910</td>\n",
       "      <td>-0.951240</td>\n",
       "      <td>1.713361</td>\n",
       "      <td>-2.041455</td>\n",
       "      <td>1.248835</td>\n",
       "      <td>-3.099297</td>\n",
       "      <td>0.470355</td>\n",
       "      <td>0.100278</td>\n",
       "      <td>...</td>\n",
       "      <td>1.631961</td>\n",
       "      <td>-2.643044</td>\n",
       "      <td>1.638079</td>\n",
       "      <td>1.352711</td>\n",
       "      <td>-1.153366</td>\n",
       "      <td>2.204770</td>\n",
       "      <td>-5.934611</td>\n",
       "      <td>-4.415373</td>\n",
       "      <td>3.064456</td>\n",
       "      <td>-2.190602</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1.943645</td>\n",
       "      <td>-1.087292</td>\n",
       "      <td>2.275767</td>\n",
       "      <td>-0.813578</td>\n",
       "      <td>1.121162</td>\n",
       "      <td>-1.855459</td>\n",
       "      <td>1.098353</td>\n",
       "      <td>-3.171567</td>\n",
       "      <td>0.737215</td>\n",
       "      <td>0.383941</td>\n",
       "      <td>...</td>\n",
       "      <td>1.407518</td>\n",
       "      <td>-2.815509</td>\n",
       "      <td>2.073592</td>\n",
       "      <td>1.181931</td>\n",
       "      <td>-0.901339</td>\n",
       "      <td>1.766575</td>\n",
       "      <td>-4.593009</td>\n",
       "      <td>-3.681092</td>\n",
       "      <td>2.426500</td>\n",
       "      <td>-1.965810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.527792</td>\n",
       "      <td>-1.015823</td>\n",
       "      <td>1.037761</td>\n",
       "      <td>-0.317868</td>\n",
       "      <td>-0.009117</td>\n",
       "      <td>-0.300108</td>\n",
       "      <td>0.145670</td>\n",
       "      <td>-1.214258</td>\n",
       "      <td>0.575413</td>\n",
       "      <td>0.319207</td>\n",
       "      <td>...</td>\n",
       "      <td>0.590685</td>\n",
       "      <td>-1.015095</td>\n",
       "      <td>0.404293</td>\n",
       "      <td>-0.036654</td>\n",
       "      <td>0.229922</td>\n",
       "      <td>0.548078</td>\n",
       "      <td>-0.990845</td>\n",
       "      <td>-1.071017</td>\n",
       "      <td>0.159505</td>\n",
       "      <td>-0.640709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0.226097</td>\n",
       "      <td>-0.912251</td>\n",
       "      <td>0.708252</td>\n",
       "      <td>-0.288762</td>\n",
       "      <td>0.087465</td>\n",
       "      <td>0.339399</td>\n",
       "      <td>-0.080053</td>\n",
       "      <td>-0.495883</td>\n",
       "      <td>0.601787</td>\n",
       "      <td>-0.025573</td>\n",
       "      <td>...</td>\n",
       "      <td>0.509829</td>\n",
       "      <td>-0.953439</td>\n",
       "      <td>-0.174323</td>\n",
       "      <td>-0.128326</td>\n",
       "      <td>0.652848</td>\n",
       "      <td>0.279634</td>\n",
       "      <td>-0.737565</td>\n",
       "      <td>-0.845867</td>\n",
       "      <td>0.400943</td>\n",
       "      <td>-0.742461</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>0.791906</td>\n",
       "      <td>-0.050599</td>\n",
       "      <td>1.252299</td>\n",
       "      <td>-0.039181</td>\n",
       "      <td>0.087118</td>\n",
       "      <td>-1.048984</td>\n",
       "      <td>0.134164</td>\n",
       "      <td>-1.191949</td>\n",
       "      <td>0.675383</td>\n",
       "      <td>-0.191731</td>\n",
       "      <td>...</td>\n",
       "      <td>0.611286</td>\n",
       "      <td>-1.361567</td>\n",
       "      <td>0.350822</td>\n",
       "      <td>0.107739</td>\n",
       "      <td>0.065545</td>\n",
       "      <td>0.838603</td>\n",
       "      <td>-1.422344</td>\n",
       "      <td>-1.323586</td>\n",
       "      <td>0.914966</td>\n",
       "      <td>-0.281728</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>0.866563</td>\n",
       "      <td>-1.256081</td>\n",
       "      <td>1.202125</td>\n",
       "      <td>-0.385703</td>\n",
       "      <td>0.761847</td>\n",
       "      <td>-0.611253</td>\n",
       "      <td>0.706594</td>\n",
       "      <td>-1.860197</td>\n",
       "      <td>0.667664</td>\n",
       "      <td>0.385697</td>\n",
       "      <td>...</td>\n",
       "      <td>1.064760</td>\n",
       "      <td>-1.613429</td>\n",
       "      <td>0.695706</td>\n",
       "      <td>0.742108</td>\n",
       "      <td>-0.017866</td>\n",
       "      <td>0.870360</td>\n",
       "      <td>-2.465451</td>\n",
       "      <td>-2.106171</td>\n",
       "      <td>1.109272</td>\n",
       "      <td>-1.342029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>0.165756</td>\n",
       "      <td>-0.837589</td>\n",
       "      <td>0.909629</td>\n",
       "      <td>-0.195248</td>\n",
       "      <td>-0.147802</td>\n",
       "      <td>0.463221</td>\n",
       "      <td>-0.280277</td>\n",
       "      <td>-0.379127</td>\n",
       "      <td>0.591854</td>\n",
       "      <td>-0.234692</td>\n",
       "      <td>...</td>\n",
       "      <td>0.324979</td>\n",
       "      <td>-0.779651</td>\n",
       "      <td>-0.196971</td>\n",
       "      <td>-0.214583</td>\n",
       "      <td>0.578756</td>\n",
       "      <td>0.303365</td>\n",
       "      <td>-0.711127</td>\n",
       "      <td>-0.803022</td>\n",
       "      <td>0.224744</td>\n",
       "      <td>-0.596073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>-0.711264</td>\n",
       "      <td>-0.313377</td>\n",
       "      <td>-1.537928</td>\n",
       "      <td>-0.445861</td>\n",
       "      <td>-1.046072</td>\n",
       "      <td>2.228831</td>\n",
       "      <td>-0.232429</td>\n",
       "      <td>1.365681</td>\n",
       "      <td>-0.359007</td>\n",
       "      <td>-0.560773</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.765562</td>\n",
       "      <td>0.713428</td>\n",
       "      <td>-1.891653</td>\n",
       "      <td>-1.241241</td>\n",
       "      <td>1.121401</td>\n",
       "      <td>-0.779070</td>\n",
       "      <td>2.209726</td>\n",
       "      <td>0.917561</td>\n",
       "      <td>-1.366258</td>\n",
       "      <td>-0.002396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>-0.773445</td>\n",
       "      <td>-0.378233</td>\n",
       "      <td>-1.598493</td>\n",
       "      <td>-0.457237</td>\n",
       "      <td>-1.048272</td>\n",
       "      <td>2.317907</td>\n",
       "      <td>-0.225659</td>\n",
       "      <td>1.418877</td>\n",
       "      <td>-0.345373</td>\n",
       "      <td>-0.563281</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.765340</td>\n",
       "      <td>0.692370</td>\n",
       "      <td>-1.920101</td>\n",
       "      <td>-1.256244</td>\n",
       "      <td>1.126904</td>\n",
       "      <td>-0.811196</td>\n",
       "      <td>2.265481</td>\n",
       "      <td>0.964531</td>\n",
       "      <td>-1.385092</td>\n",
       "      <td>0.000151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>-0.551217</td>\n",
       "      <td>-0.484983</td>\n",
       "      <td>-0.740480</td>\n",
       "      <td>-0.317116</td>\n",
       "      <td>-0.289072</td>\n",
       "      <td>1.203944</td>\n",
       "      <td>-0.258495</td>\n",
       "      <td>0.829469</td>\n",
       "      <td>-0.127503</td>\n",
       "      <td>-0.098855</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.133899</td>\n",
       "      <td>0.557267</td>\n",
       "      <td>-1.310686</td>\n",
       "      <td>-0.872177</td>\n",
       "      <td>0.841495</td>\n",
       "      <td>-0.366921</td>\n",
       "      <td>1.319264</td>\n",
       "      <td>0.608191</td>\n",
       "      <td>-0.727719</td>\n",
       "      <td>0.437993</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>-0.439805</td>\n",
       "      <td>-0.603414</td>\n",
       "      <td>-2.132188</td>\n",
       "      <td>-0.951599</td>\n",
       "      <td>-0.920562</td>\n",
       "      <td>2.266138</td>\n",
       "      <td>0.164758</td>\n",
       "      <td>1.358264</td>\n",
       "      <td>-0.106242</td>\n",
       "      <td>-0.370740</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.310326</td>\n",
       "      <td>0.425036</td>\n",
       "      <td>-1.724499</td>\n",
       "      <td>-1.099207</td>\n",
       "      <td>1.453926</td>\n",
       "      <td>-1.312302</td>\n",
       "      <td>2.784316</td>\n",
       "      <td>0.947054</td>\n",
       "      <td>-0.947618</td>\n",
       "      <td>0.042994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>-0.611443</td>\n",
       "      <td>-0.506999</td>\n",
       "      <td>-1.324136</td>\n",
       "      <td>-0.568980</td>\n",
       "      <td>-0.544524</td>\n",
       "      <td>1.422707</td>\n",
       "      <td>0.092015</td>\n",
       "      <td>1.065604</td>\n",
       "      <td>-0.093371</td>\n",
       "      <td>-0.146054</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.218616</td>\n",
       "      <td>0.667531</td>\n",
       "      <td>-1.635424</td>\n",
       "      <td>-0.655394</td>\n",
       "      <td>0.993311</td>\n",
       "      <td>-0.665229</td>\n",
       "      <td>1.776212</td>\n",
       "      <td>0.746641</td>\n",
       "      <td>-0.892052</td>\n",
       "      <td>0.217256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>-0.637345</td>\n",
       "      <td>-0.636482</td>\n",
       "      <td>-0.669429</td>\n",
       "      <td>-0.576676</td>\n",
       "      <td>-1.266894</td>\n",
       "      <td>2.006546</td>\n",
       "      <td>-0.626739</td>\n",
       "      <td>1.061490</td>\n",
       "      <td>-0.010571</td>\n",
       "      <td>-0.746190</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.587130</td>\n",
       "      <td>-0.089719</td>\n",
       "      <td>-1.631752</td>\n",
       "      <td>-0.722160</td>\n",
       "      <td>1.427183</td>\n",
       "      <td>-0.709005</td>\n",
       "      <td>1.850475</td>\n",
       "      <td>0.589096</td>\n",
       "      <td>-0.998568</td>\n",
       "      <td>-0.583243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1.364293</td>\n",
       "      <td>-1.919275</td>\n",
       "      <td>1.467089</td>\n",
       "      <td>-0.525684</td>\n",
       "      <td>1.599125</td>\n",
       "      <td>-1.581952</td>\n",
       "      <td>1.269167</td>\n",
       "      <td>-2.868625</td>\n",
       "      <td>0.918929</td>\n",
       "      <td>1.291348</td>\n",
       "      <td>...</td>\n",
       "      <td>1.918654</td>\n",
       "      <td>-1.614646</td>\n",
       "      <td>1.931791</td>\n",
       "      <td>0.983550</td>\n",
       "      <td>-0.427123</td>\n",
       "      <td>1.063765</td>\n",
       "      <td>-3.549005</td>\n",
       "      <td>-2.731330</td>\n",
       "      <td>1.863523</td>\n",
       "      <td>-1.438287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>0.437251</td>\n",
       "      <td>-1.131051</td>\n",
       "      <td>0.849429</td>\n",
       "      <td>-0.368388</td>\n",
       "      <td>0.225779</td>\n",
       "      <td>0.117888</td>\n",
       "      <td>0.248102</td>\n",
       "      <td>-0.721550</td>\n",
       "      <td>0.635726</td>\n",
       "      <td>-0.074681</td>\n",
       "      <td>...</td>\n",
       "      <td>0.585914</td>\n",
       "      <td>-1.177844</td>\n",
       "      <td>0.362499</td>\n",
       "      <td>0.283238</td>\n",
       "      <td>0.211147</td>\n",
       "      <td>0.309783</td>\n",
       "      <td>-0.971592</td>\n",
       "      <td>-1.092962</td>\n",
       "      <td>0.562560</td>\n",
       "      <td>-0.794982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>-0.799734</td>\n",
       "      <td>-0.876434</td>\n",
       "      <td>-1.559608</td>\n",
       "      <td>-0.636807</td>\n",
       "      <td>-1.034256</td>\n",
       "      <td>2.258245</td>\n",
       "      <td>-0.151252</td>\n",
       "      <td>1.546439</td>\n",
       "      <td>-0.007002</td>\n",
       "      <td>-0.309009</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.553826</td>\n",
       "      <td>0.810577</td>\n",
       "      <td>-1.769872</td>\n",
       "      <td>-0.858350</td>\n",
       "      <td>1.261316</td>\n",
       "      <td>-1.066611</td>\n",
       "      <td>2.517839</td>\n",
       "      <td>1.145920</td>\n",
       "      <td>-1.396793</td>\n",
       "      <td>0.070701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1.713707</td>\n",
       "      <td>-0.478280</td>\n",
       "      <td>1.761179</td>\n",
       "      <td>-0.597557</td>\n",
       "      <td>1.775618</td>\n",
       "      <td>-2.020615</td>\n",
       "      <td>1.468742</td>\n",
       "      <td>-3.035022</td>\n",
       "      <td>0.640905</td>\n",
       "      <td>0.411676</td>\n",
       "      <td>...</td>\n",
       "      <td>1.707383</td>\n",
       "      <td>-2.140487</td>\n",
       "      <td>1.224570</td>\n",
       "      <td>1.341206</td>\n",
       "      <td>-1.018394</td>\n",
       "      <td>1.966278</td>\n",
       "      <td>-5.483290</td>\n",
       "      <td>-3.913215</td>\n",
       "      <td>2.621877</td>\n",
       "      <td>-1.889695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>-0.293014</td>\n",
       "      <td>0.328315</td>\n",
       "      <td>-0.252548</td>\n",
       "      <td>0.170059</td>\n",
       "      <td>-1.448728</td>\n",
       "      <td>1.113993</td>\n",
       "      <td>-0.747425</td>\n",
       "      <td>0.669318</td>\n",
       "      <td>-0.539600</td>\n",
       "      <td>-1.028947</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.215645</td>\n",
       "      <td>0.015466</td>\n",
       "      <td>-1.475949</td>\n",
       "      <td>-1.005116</td>\n",
       "      <td>0.598915</td>\n",
       "      <td>0.322653</td>\n",
       "      <td>0.771331</td>\n",
       "      <td>0.043018</td>\n",
       "      <td>-1.204419</td>\n",
       "      <td>-0.299253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>-1.167658</td>\n",
       "      <td>-0.870320</td>\n",
       "      <td>-1.657678</td>\n",
       "      <td>-0.971837</td>\n",
       "      <td>-1.470639</td>\n",
       "      <td>2.984151</td>\n",
       "      <td>-0.369626</td>\n",
       "      <td>1.703925</td>\n",
       "      <td>-0.233460</td>\n",
       "      <td>-0.828888</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.689074</td>\n",
       "      <td>0.522325</td>\n",
       "      <td>-2.510032</td>\n",
       "      <td>-0.788953</td>\n",
       "      <td>1.633304</td>\n",
       "      <td>-1.064721</td>\n",
       "      <td>2.147650</td>\n",
       "      <td>0.821055</td>\n",
       "      <td>-1.486556</td>\n",
       "      <td>-1.035391</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25 rows × 100 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0         1         2         3         4         5         6   \\\n",
       "0   1.117691 -1.285147  1.297701 -0.356072  1.151868 -1.478432  0.695909   \n",
       "1  -0.397457 -0.586370 -2.101782 -0.957839 -0.894716  2.252759  0.179345   \n",
       "2   1.428277 -0.409841  1.821033 -0.632364  0.519591 -1.075682  0.467286   \n",
       "3  -0.807010 -1.132076 -2.040275 -1.018199 -1.241043  2.814131 -0.277895   \n",
       "4  -0.729850 -1.103235 -1.926998 -1.014075 -1.196003  2.693963 -0.228020   \n",
       "5  -0.700547 -0.697610 -0.764258 -0.581024 -1.340819  2.218557 -0.651789   \n",
       "6   1.882998 -0.430399  2.193910 -0.951240  1.713361 -2.041455  1.248835   \n",
       "7   1.943645 -1.087292  2.275767 -0.813578  1.121162 -1.855459  1.098353   \n",
       "8   0.527792 -1.015823  1.037761 -0.317868 -0.009117 -0.300108  0.145670   \n",
       "9   0.226097 -0.912251  0.708252 -0.288762  0.087465  0.339399 -0.080053   \n",
       "10  0.791906 -0.050599  1.252299 -0.039181  0.087118 -1.048984  0.134164   \n",
       "11  0.866563 -1.256081  1.202125 -0.385703  0.761847 -0.611253  0.706594   \n",
       "12  0.165756 -0.837589  0.909629 -0.195248 -0.147802  0.463221 -0.280277   \n",
       "13 -0.711264 -0.313377 -1.537928 -0.445861 -1.046072  2.228831 -0.232429   \n",
       "14 -0.773445 -0.378233 -1.598493 -0.457237 -1.048272  2.317907 -0.225659   \n",
       "15 -0.551217 -0.484983 -0.740480 -0.317116 -0.289072  1.203944 -0.258495   \n",
       "16 -0.439805 -0.603414 -2.132188 -0.951599 -0.920562  2.266138  0.164758   \n",
       "17 -0.611443 -0.506999 -1.324136 -0.568980 -0.544524  1.422707  0.092015   \n",
       "18 -0.637345 -0.636482 -0.669429 -0.576676 -1.266894  2.006546 -0.626739   \n",
       "19  1.364293 -1.919275  1.467089 -0.525684  1.599125 -1.581952  1.269167   \n",
       "20  0.437251 -1.131051  0.849429 -0.368388  0.225779  0.117888  0.248102   \n",
       "21 -0.799734 -0.876434 -1.559608 -0.636807 -1.034256  2.258245 -0.151252   \n",
       "22  1.713707 -0.478280  1.761179 -0.597557  1.775618 -2.020615  1.468742   \n",
       "23 -0.293014  0.328315 -0.252548  0.170059 -1.448728  1.113993 -0.747425   \n",
       "24 -1.167658 -0.870320 -1.657678 -0.971837 -1.470639  2.984151 -0.369626   \n",
       "\n",
       "          7         8         9   ...        90        91        92        93  \\\n",
       "0  -2.049861  0.775193  0.952248  ...  1.548072 -1.212546  1.297358  0.304305   \n",
       "1   1.366534 -0.141001 -0.368006  ... -0.324187  0.456942 -1.723317 -1.089818   \n",
       "2  -2.106847  0.623872 -0.444501  ...  0.977349 -2.612781  0.853364  1.045666   \n",
       "3   1.702027 -0.216814 -0.389414  ... -0.576729  0.361689 -1.794586 -0.989517   \n",
       "4   1.613018 -0.233582 -0.356466  ... -0.552327  0.412861 -1.779991 -0.922350   \n",
       "5   1.153820 -0.019014 -0.771810  ... -0.663449 -0.050688 -1.609203 -0.787775   \n",
       "6  -3.099297  0.470355  0.100278  ...  1.631961 -2.643044  1.638079  1.352711   \n",
       "7  -3.171567  0.737215  0.383941  ...  1.407518 -2.815509  2.073592  1.181931   \n",
       "8  -1.214258  0.575413  0.319207  ...  0.590685 -1.015095  0.404293 -0.036654   \n",
       "9  -0.495883  0.601787 -0.025573  ...  0.509829 -0.953439 -0.174323 -0.128326   \n",
       "10 -1.191949  0.675383 -0.191731  ...  0.611286 -1.361567  0.350822  0.107739   \n",
       "11 -1.860197  0.667664  0.385697  ...  1.064760 -1.613429  0.695706  0.742108   \n",
       "12 -0.379127  0.591854 -0.234692  ...  0.324979 -0.779651 -0.196971 -0.214583   \n",
       "13  1.365681 -0.359007 -0.560773  ... -0.765562  0.713428 -1.891653 -1.241241   \n",
       "14  1.418877 -0.345373 -0.563281  ... -0.765340  0.692370 -1.920101 -1.256244   \n",
       "15  0.829469 -0.127503 -0.098855  ... -0.133899  0.557267 -1.310686 -0.872177   \n",
       "16  1.358264 -0.106242 -0.370740  ... -0.310326  0.425036 -1.724499 -1.099207   \n",
       "17  1.065604 -0.093371 -0.146054  ... -0.218616  0.667531 -1.635424 -0.655394   \n",
       "18  1.061490 -0.010571 -0.746190  ... -0.587130 -0.089719 -1.631752 -0.722160   \n",
       "19 -2.868625  0.918929  1.291348  ...  1.918654 -1.614646  1.931791  0.983550   \n",
       "20 -0.721550  0.635726 -0.074681  ...  0.585914 -1.177844  0.362499  0.283238   \n",
       "21  1.546439 -0.007002 -0.309009  ... -0.553826  0.810577 -1.769872 -0.858350   \n",
       "22 -3.035022  0.640905  0.411676  ...  1.707383 -2.140487  1.224570  1.341206   \n",
       "23  0.669318 -0.539600 -1.028947  ... -1.215645  0.015466 -1.475949 -1.005116   \n",
       "24  1.703925 -0.233460 -0.828888  ... -0.689074  0.522325 -2.510032 -0.788953   \n",
       "\n",
       "          94        95        96        97        98        99  \n",
       "0  -0.088180  0.741703 -2.192100 -1.760029  1.292709 -0.566314  \n",
       "1   1.471538 -1.332001  2.797414  0.962129 -0.949379  0.027219  \n",
       "2  -0.377918  1.504324 -3.668192 -3.049183  1.911584 -1.719634  \n",
       "3   1.530060 -1.400325  3.109184  1.333641 -1.368727 -0.299508  \n",
       "4   1.528533 -1.356465  2.961123  1.244486 -1.314663 -0.330579  \n",
       "5   1.419916 -0.765434  1.982262  0.660961 -1.071012 -0.559409  \n",
       "6  -1.153366  2.204770 -5.934611 -4.415373  3.064456 -2.190602  \n",
       "7  -0.901339  1.766575 -4.593009 -3.681092  2.426500 -1.965810  \n",
       "8   0.229922  0.548078 -0.990845 -1.071017  0.159505 -0.640709  \n",
       "9   0.652848  0.279634 -0.737565 -0.845867  0.400943 -0.742461  \n",
       "10  0.065545  0.838603 -1.422344 -1.323586  0.914966 -0.281728  \n",
       "11 -0.017866  0.870360 -2.465451 -2.106171  1.109272 -1.342029  \n",
       "12  0.578756  0.303365 -0.711127 -0.803022  0.224744 -0.596073  \n",
       "13  1.121401 -0.779070  2.209726  0.917561 -1.366258 -0.002396  \n",
       "14  1.126904 -0.811196  2.265481  0.964531 -1.385092  0.000151  \n",
       "15  0.841495 -0.366921  1.319264  0.608191 -0.727719  0.437993  \n",
       "16  1.453926 -1.312302  2.784316  0.947054 -0.947618  0.042994  \n",
       "17  0.993311 -0.665229  1.776212  0.746641 -0.892052  0.217256  \n",
       "18  1.427183 -0.709005  1.850475  0.589096 -0.998568 -0.583243  \n",
       "19 -0.427123  1.063765 -3.549005 -2.731330  1.863523 -1.438287  \n",
       "20  0.211147  0.309783 -0.971592 -1.092962  0.562560 -0.794982  \n",
       "21  1.261316 -1.066611  2.517839  1.145920 -1.396793  0.070701  \n",
       "22 -1.018394  1.966278 -5.483290 -3.913215  2.621877 -1.889695  \n",
       "23  0.598915  0.322653  0.771331  0.043018 -1.204419 -0.299253  \n",
       "24  1.633304 -1.064721  2.147650  0.821055 -1.486556 -1.035391  \n",
       "\n",
       "[25 rows x 100 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mags_df = pd.DataFrame (list_of_mag_vectors)\n",
    "mags_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "colonial-ethernet",
   "metadata": {},
   "source": [
    "---\n",
    "## Data preprocessing\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "catholic-joshua",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[ 1.00382113, -1.19856783,  0.90286325, ..., -0.63545758,\n",
       "         0.85830795,  0.12988954],\n",
       "       [-0.57417116,  0.35773879, -1.35580206, ...,  0.89849409,\n",
       "        -0.69996512,  0.95923406],\n",
       "       [ 1.3272889 ,  0.7509027 ,  1.2505726 , ..., -1.36190356,\n",
       "         1.28843227, -1.48164563],\n",
       "       ...,\n",
       "       [ 1.6245575 ,  0.59847658,  1.21080464, ..., -1.8487907 ,\n",
       "         1.78209276, -1.71927117],\n",
       "       [-0.46539595,  2.39491147, -0.12714411, ...,  0.38056967,\n",
       "        -0.87722101,  0.50305431],\n",
       "       [-1.37631748, -0.27466974, -1.06073248, ...,  0.81899784,\n",
       "        -1.07330886, -0.52555183]])"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mag_scaler = preprocessing.StandardScaler()\n",
    "scaled_mags_df = mag_scaler.fit_transform(mags_df)\n",
    "\n",
    "scaled_mags_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "incoming-integer",
   "metadata": {},
   "source": [
    "---\n",
    "## Clustering\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "parental-batch",
   "metadata": {},
   "source": [
    "### 1. K-means"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "twelve-occasion",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>k_range</th>\n",
       "      <th>k_scores</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>-2500.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>-945.723227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>-595.438301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>-466.881248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>-369.161081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>-294.748963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>-253.504334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>-214.095257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>-186.160850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>-162.447143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>11</td>\n",
       "      <td>-137.881258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>12</td>\n",
       "      <td>-117.227010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>13</td>\n",
       "      <td>-97.025712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>14</td>\n",
       "      <td>-79.204936</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    k_range     k_scores\n",
       "0         1 -2500.000000\n",
       "1         2  -945.723227\n",
       "2         3  -595.438301\n",
       "3         4  -466.881248\n",
       "4         5  -369.161081\n",
       "5         6  -294.748963\n",
       "6         7  -253.504334\n",
       "7         8  -214.095257\n",
       "8         9  -186.160850\n",
       "9        10  -162.447143\n",
       "10       11  -137.881258\n",
       "11       12  -117.227010\n",
       "12       13   -97.025712\n",
       "13       14   -79.204936"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k_range_end = 15 # Usually it is sqrt(# of mags)\n",
    "\n",
    "k_range = range(1, k_range_end)\n",
    "\n",
    "k_mean_models = [KMeans (n_clusters = i, random_state = SEED) for i in k_range]\n",
    "k_scores = [k_mean_model.fit(scaled_mags_df).score(scaled_mags_df) for k_mean_model in k_mean_models]\n",
    "k_data = pd.DataFrame ({'k_range':k_range, 'k_scores':k_scores})\n",
    "k_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "grand-knitting",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-93e7955b0bc7415f9d413c32deabeb8f\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-93e7955b0bc7415f9d413c32deabeb8f\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-93e7955b0bc7415f9d413c32deabeb8f\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.8.1?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function loadScript(lib) {\n",
       "      return new Promise(function(resolve, reject) {\n",
       "        var s = document.createElement('script');\n",
       "        s.src = paths[lib];\n",
       "        s.async = true;\n",
       "        s.onload = () => resolve(paths[lib]);\n",
       "        s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "        document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "      });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else if (typeof vegaEmbed === \"function\") {\n",
       "      displayChart(vegaEmbed);\n",
       "    } else {\n",
       "      loadScript(\"vega\")\n",
       "        .then(() => loadScript(\"vega-lite\"))\n",
       "        .then(() => loadScript(\"vega-embed\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}}, \"data\": {\"name\": \"data-fa978f46b25af447234f1cb3bfeee413\"}, \"mark\": \"line\", \"encoding\": {\"x\": {\"type\": \"quantitative\", \"field\": \"k_range\"}, \"y\": {\"type\": \"quantitative\", \"field\": \"k_scores\"}}, \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.8.1.json\", \"datasets\": {\"data-fa978f46b25af447234f1cb3bfeee413\": [{\"k_range\": 1, \"k_scores\": -2500.0000000000005}, {\"k_range\": 2, \"k_scores\": -945.7232272874026}, {\"k_range\": 3, \"k_scores\": -595.4383014391532}, {\"k_range\": 4, \"k_scores\": -466.88124815552794}, {\"k_range\": 5, \"k_scores\": -369.161080768216}, {\"k_range\": 6, \"k_scores\": -294.7489626465943}, {\"k_range\": 7, \"k_scores\": -253.50433416800453}, {\"k_range\": 8, \"k_scores\": -214.09525745891628}, {\"k_range\": 9, \"k_scores\": -186.16085046576845}, {\"k_range\": 10, \"k_scores\": -162.44714257115055}, {\"k_range\": 11, \"k_scores\": -137.8812584709489}, {\"k_range\": 12, \"k_scores\": -117.22701035784957}, {\"k_range\": 13, \"k_scores\": -97.02571245367629}, {\"k_range\": 14, \"k_scores\": -79.20493590685432}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.Chart(...)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k_num_chart = alt.Chart(data = k_data).mark_line().encode(\n",
    "    alt.X ('k_range:Q'),\n",
    "    alt.Y ('k_scores:Q')\n",
    ")\n",
    "\n",
    "k_num_chart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "banner-framework",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([6, 1, 2, 1, 1, 4, 2, 2, 0, 0, 0, 0, 0, 4, 4, 5, 1, 5, 4, 6, 0, 1,\n",
       "       2, 3, 4], dtype=int32)"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# We can see from the chart above that 6 or 7 clusters are optimal for this task (where END = 25 MAGs)\n",
    "num_of_clusters = 7\n",
    "\n",
    "k_means_model = KMeans (n_clusters = num_of_clusters, random_state = SEED)\n",
    "k_means_predicted = k_means_model.fit_predict(scaled_mags_df)\n",
    "k_means_predicted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "controversial-plymouth",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-18c759729401480aa5c3ff192f52abad\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-18c759729401480aa5c3ff192f52abad\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-18c759729401480aa5c3ff192f52abad\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.8.1?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function loadScript(lib) {\n",
       "      return new Promise(function(resolve, reject) {\n",
       "        var s = document.createElement('script');\n",
       "        s.src = paths[lib];\n",
       "        s.async = true;\n",
       "        s.onload = () => resolve(paths[lib]);\n",
       "        s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "        document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "      });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else if (typeof vegaEmbed === \"function\") {\n",
       "      displayChart(vegaEmbed);\n",
       "    } else {\n",
       "      loadScript(\"vega\")\n",
       "        .then(() => loadScript(\"vega-lite\"))\n",
       "        .then(() => loadScript(\"vega-embed\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}}, \"layer\": [{\"data\": {\"name\": \"data-f0e53477dba4de0d89449f7204f8d793\"}, \"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"nominal\", \"field\": \"Labels\"}, \"x\": {\"type\": \"quantitative\", \"field\": \"PC_1\"}, \"y\": {\"type\": \"quantitative\", \"field\": \"PC_2\"}}}, {\"data\": {\"name\": \"data-c0f09e6cb70c8c818325c5023c5b6d08\"}, \"mark\": {\"type\": \"point\", \"color\": \"black\", \"opacity\": 0.7, \"shape\": \"diamond\", \"size\": 50}, \"encoding\": {\"x\": {\"type\": \"quantitative\", \"field\": \"PC_1\"}, \"y\": {\"type\": \"quantitative\", \"field\": \"PC_2\"}}}], \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.8.1.json\", \"datasets\": {\"data-f0e53477dba4de0d89449f7204f8d793\": [{\"PC_1\": 7.608232831242397, \"PC_2\": -4.703450391765291, \"Labels\": 6}, {\"PC_1\": -8.398305059411609, \"PC_2\": -3.12367321402007, \"Labels\": 1}, {\"PC_1\": 11.424670592093205, \"PC_2\": 3.1493756255249004, \"Labels\": 2}, {\"PC_1\": -9.403078535843957, \"PC_2\": -2.059398380670939, \"Labels\": 1}, {\"PC_1\": -9.10839009880538, \"PC_2\": -1.9434162944406181, \"Labels\": 1}, {\"PC_1\": -6.572188552138569, \"PC_2\": 1.8770470984578347, \"Labels\": 4}, {\"PC_1\": 16.369160427318477, \"PC_2\": 2.399478197070116, \"Labels\": 2}, {\"PC_1\": 14.68500453643797, \"PC_2\": -0.11515699602987278, \"Labels\": 2}, {\"PC_1\": 3.1442587405363587, \"PC_2\": -0.7608408411129592, \"Labels\": 0}, {\"PC_1\": 1.900519113934637, \"PC_2\": -0.09561781591184619, \"Labels\": 0}, {\"PC_1\": 5.708898475304954, \"PC_2\": 0.6730624247038894, \"Labels\": 0}, {\"PC_1\": 7.174977522394448, \"PC_2\": -1.0203525568655218, \"Labels\": 0}, {\"PC_1\": 1.5508142875671727, \"PC_2\": 1.2069373423292145, \"Labels\": 0}, {\"PC_1\": -8.431243158529288, \"PC_2\": 1.7951146264395323, \"Labels\": 4}, {\"PC_1\": -8.633390127330554, \"PC_2\": 1.6970697966965236, \"Labels\": 4}, {\"PC_1\": -4.891743448604073, \"PC_2\": -0.5917820010007707, \"Labels\": 5}, {\"PC_1\": -8.357935819261321, \"PC_2\": -3.1922889382295194, \"Labels\": 1}, {\"PC_1\": -6.289077812333235, \"PC_2\": -1.8637426076236705, \"Labels\": 5}, {\"PC_1\": -6.087136538802871, \"PC_2\": 1.8362791532967526, \"Labels\": 4}, {\"PC_1\": 10.643814103573433, \"PC_2\": -5.310494535431499, \"Labels\": 6}, {\"PC_1\": 3.4894773721390733, \"PC_2\": -0.9804083216789435, \"Labels\": 0}, {\"PC_1\": -8.732147476904709, \"PC_2\": -1.551019253379136, \"Labels\": 1}, {\"PC_1\": 14.252328297122418, \"PC_2\": 1.1804211349844376, \"Labels\": 2}, {\"PC_1\": -3.5858870443661037, \"PC_2\": 9.098295501178026, \"Labels\": 3}, {\"PC_1\": -9.461632627332877, \"PC_2\": 2.398561247479437, \"Labels\": 4}], \"data-c0f09e6cb70c8c818325c5023c5b6d08\": [{\"PC_1\": 3.7144029967426975, \"PC_2\": -0.3180541157553076}, {\"PC_1\": -8.524916233181528, \"PC_2\": -4.226950180062868}, {\"PC_1\": 13.816815953025314, \"PC_2\": 2.1216039022424438}, {\"PC_1\": -4.856577301455141, \"PC_2\": 8.789566867085018}, {\"PC_1\": -8.068344729378932, \"PC_2\": 0.04780842109577909}, {\"PC_1\": -5.585107146947802, \"PC_2\": -2.6418730587485117}, {\"PC_1\": 9.503726461195395, \"PC_2\": -3.7721018358565526}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.LayerChart(...)"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k_means_chart = visualize_with_pca (scaled_mags_df, k_means_predicted, k_means_model.cluster_centers_)\n",
    "k_means_chart\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "sophisticated-penguin",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "miniature-mention",
   "metadata": {},
   "source": [
    "### 2. DBSCAN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "variable-bread",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 0,  1,  2,  1,  1,  1,  2,  2,  0,  0,  0,  0,  0,  1,  1,  1,  1,\n",
       "        1,  1,  0,  0,  1,  2, -1,  1])"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MIN_SAMPLES = 3\n",
    "EPS = 7\n",
    "\n",
    "dbscan_model = DBSCAN(eps = EPS, min_samples = MIN_SAMPLES, n_jobs = num_of_workers)\n",
    "dbscan_predicted = dbscan_model.fit_predict (scaled_mags_df)\n",
    "dbscan_predicted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "congressional-savage",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[ True  True  True  True  True  True False  True  True  True  True  True\n",
      "  True  True  True  True  True  True  True  True  True  True  True False\n",
      "  True]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([ 0,  1,  2,  3,  4,  5,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,\n",
       "       18, 19, 20, 21, 22, 24])"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dbscan_core_samples = np.zeros_like(dbscan_predicted, dtype=bool)\n",
    "dbscan_core_samples[dbscan_model.core_sample_indices_] = 1\n",
    "print(dbscan_core_samples)\n",
    "dbscan_model.core_sample_indices_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "purple-nepal",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-52617d41ab7a4eb4ab0b554253c94a56\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-52617d41ab7a4eb4ab0b554253c94a56\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-52617d41ab7a4eb4ab0b554253c94a56\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.8.1?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function loadScript(lib) {\n",
       "      return new Promise(function(resolve, reject) {\n",
       "        var s = document.createElement('script');\n",
       "        s.src = paths[lib];\n",
       "        s.async = true;\n",
       "        s.onload = () => resolve(paths[lib]);\n",
       "        s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "        document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "      });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else if (typeof vegaEmbed === \"function\") {\n",
       "      displayChart(vegaEmbed);\n",
       "    } else {\n",
       "      loadScript(\"vega\")\n",
       "        .then(() => loadScript(\"vega-lite\"))\n",
       "        .then(() => loadScript(\"vega-embed\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}}, \"data\": {\"name\": \"data-6b20a8da93926eada30128335f693068\"}, \"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"nominal\", \"field\": \"Labels\"}, \"x\": {\"type\": \"quantitative\", \"field\": \"PC_1\"}, \"y\": {\"type\": \"quantitative\", \"field\": \"PC_2\"}}, \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.8.1.json\", \"datasets\": {\"data-6b20a8da93926eada30128335f693068\": [{\"PC_1\": 7.608232831242397, \"PC_2\": -4.703450391765291, \"Labels\": 0}, {\"PC_1\": -8.398305059411609, \"PC_2\": -3.12367321402007, \"Labels\": 1}, {\"PC_1\": 11.424670592093205, \"PC_2\": 3.1493756255249004, \"Labels\": 2}, {\"PC_1\": -9.403078535843957, \"PC_2\": -2.059398380670939, \"Labels\": 1}, {\"PC_1\": -9.10839009880538, \"PC_2\": -1.9434162944406181, \"Labels\": 1}, {\"PC_1\": -6.572188552138569, \"PC_2\": 1.8770470984578347, \"Labels\": 1}, {\"PC_1\": 16.369160427318477, \"PC_2\": 2.399478197070116, \"Labels\": 2}, {\"PC_1\": 14.68500453643797, \"PC_2\": -0.11515699602987278, \"Labels\": 2}, {\"PC_1\": 3.1442587405363587, \"PC_2\": -0.7608408411129592, \"Labels\": 0}, {\"PC_1\": 1.900519113934637, \"PC_2\": -0.09561781591184619, \"Labels\": 0}, {\"PC_1\": 5.708898475304954, \"PC_2\": 0.6730624247038894, \"Labels\": 0}, {\"PC_1\": 7.174977522394448, \"PC_2\": -1.0203525568655218, \"Labels\": 0}, {\"PC_1\": 1.5508142875671727, \"PC_2\": 1.2069373423292145, \"Labels\": 0}, {\"PC_1\": -8.431243158529288, \"PC_2\": 1.7951146264395323, \"Labels\": 1}, {\"PC_1\": -8.633390127330554, \"PC_2\": 1.6970697966965236, \"Labels\": 1}, {\"PC_1\": -4.891743448604073, \"PC_2\": -0.5917820010007707, \"Labels\": 1}, {\"PC_1\": -8.357935819261321, \"PC_2\": -3.1922889382295194, \"Labels\": 1}, {\"PC_1\": -6.289077812333235, \"PC_2\": -1.8637426076236705, \"Labels\": 1}, {\"PC_1\": -6.087136538802871, \"PC_2\": 1.8362791532967526, \"Labels\": 1}, {\"PC_1\": 10.643814103573433, \"PC_2\": -5.310494535431499, \"Labels\": 0}, {\"PC_1\": 3.4894773721390733, \"PC_2\": -0.9804083216789435, \"Labels\": 0}, {\"PC_1\": -8.732147476904709, \"PC_2\": -1.551019253379136, \"Labels\": 1}, {\"PC_1\": 14.252328297122418, \"PC_2\": 1.1804211349844376, \"Labels\": 2}, {\"PC_1\": -3.5858870443661037, \"PC_2\": 9.098295501178026, \"Labels\": -1}, {\"PC_1\": -9.461632627332877, \"PC_2\": 2.398561247479437, \"Labels\": 1}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.Chart(...)"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Visualize clusters\n",
    "dbscan_chart = visualize_with_pca (scaled_mags_df, dbscan_predicted, dbscan_core_samples)\n",
    "dbscan_chart\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "played-consumption",
   "metadata": {},
   "source": [
    "---\n",
    "## Evaluation\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "sapphire-dinner",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
