{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "handy-invalid",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import altair as alt\n",
    "import os, random\n",
    "from Bio import SeqIO\n",
    "import gensim\n",
    "from gensim.models import Word2Vec\n",
    "from sklearn.cluster import KMeans, DBSCAN\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn import preprocessing, model_selection\n",
    "import altair as alt\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "charitable-watershed",
   "metadata": {},
   "source": [
    "## Defining paths for each and every omic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "literary-ukraine",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_root_data = os.path.join ('..', 'Data', 'Extracted', 'Databases')\n",
    "path_all_fasta = os.path.join (path_root_data, 'fasta_files', 'AllBins')\n",
    "path_normalised_metabolomics = os.path.join (path_root_data, 'Metabolomics', 'Normalised_Tables')\n",
    "path_model_save_root = 'Saved_models'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "attempted-truth",
   "metadata": {},
   "outputs": [],
   "source": [
    "num_of_mags = len([i for i in os.listdir(path_all_fasta) if (i[-2:] == 'fa')])\n",
    "SEED = 42\n",
    "END = 25\n",
    "num_of_workers = 3\n",
    "random.seed(SEED)\n",
    "np.random.seed(SEED)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "political-civilization",
   "metadata": {},
   "source": [
    "---\n",
    "## MAG-related functions\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "skilled-ordering",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function that splits each genome into k-mers thus creating even longer sentence (MAG)\n",
    "# It returns tokenized genome i.e. [kmer, kmer,...]\n",
    "def split_genome (genome, k = 5):\n",
    "    new_genome = []\n",
    "    n = len(genome)\n",
    "    \n",
    "    if n-k <=0:\n",
    "        return genome\n",
    "    else:\n",
    "        for i in range(n-k):\n",
    "            new_genome.append(genome[i:i+k])\n",
    "        \n",
    "        return new_genome\n",
    "\n",
    "\n",
    "def vectorize_one_mag (one_mag, w2v_model):\n",
    "    \n",
    "    # We have to generate vectors for each word in one MAG and then create vector representation of that MAG\n",
    "    # by averaging vectors of its words\n",
    "    zero_vector = np.zeros(w2v_model.vector_size)\n",
    "    word_vectors = []\n",
    "    one_mag_vector = []\n",
    "    \n",
    "    for sentence in one_mag:\n",
    "        for word in sentence:\n",
    "            if word in w2v_model.wv:\n",
    "                try:\n",
    "                    word_vectors.append(w2v_model.wv[word])\n",
    "                except KeyError:\n",
    "                    print ('Key Error')\n",
    "                    continue\n",
    "    \n",
    "    if word_vectors:\n",
    "        word_vectors = np.asarray(word_vectors)\n",
    "        one_mag_vector = word_vectors.mean (axis=0)\n",
    "    \n",
    "    else:\n",
    "        one_mag_vector = zero_vector\n",
    "    \n",
    "    return one_mag_vector\n",
    "\n",
    "\n",
    "# Function that vectorizes a MAG (document) with a pretrained word2vec model. It returns vector representation of a given MAG\n",
    "# Vectorization is done by averaging word (k-mer) vectors for the whole document (MAG)\n",
    "def vectorize_mags (w2v_model, end = 25):\n",
    "    \n",
    "    print ('Vectorizing MAGs')\n",
    "    \n",
    "    fasta_files = [i for i in os.listdir(path_all_fasta) if (i[-2:] == 'fa')]\n",
    "    list_of_mag_vectors = []\n",
    "    \n",
    "    # This was done so that I could work with first 'end' FASTA files only. Otherwise, I should just remove: i, and enumerate\n",
    "    for i, fasta_file_name in enumerate(fasta_files):\n",
    "        \n",
    "        if i == end:\n",
    "            break\n",
    "        \n",
    "        else:\n",
    "            with open(os.path.join(path_all_fasta, fasta_file_name), 'r') as input_file:\n",
    "                \n",
    "                one_mag = []\n",
    "                for fasta_string in SeqIO.parse(input_file, \"fasta\"):\n",
    "                    \n",
    "                    # Get kmers of a genome and create a sentence (list of words)\n",
    "                    temp_kmers = split_genome (str(fasta_string.seq))\n",
    "                    \n",
    "                    # Create a document (list of sentences)\n",
    "                    one_mag.append(temp_kmers)\n",
    "                    \n",
    "                # Vectorizing MAGs one by one\n",
    "                list_of_mag_vectors.append (vectorize_one_mag (one_mag, w2v_model))\n",
    "    \n",
    "    print ('Finished vectorizing')\n",
    "    \n",
    "    return list_of_mag_vectors\n",
    "    \n",
    "    \n",
    "\n",
    "# If one wants to import MAGs in order to vectorize them, one should use start argument in order to skip first 'start' MAGs\n",
    "# If one wants to import MAGs to train word2vec model, one should use only end argument, so that first 'end' MAGs are used for training\n",
    "# Todo: Implement randomisation in picking MAGs for training, and don't use first 'start' MAGs for training\n",
    "# (create list of indexes from 0 to 1364, use sklearn split train test, then traverse directory and use only MAGs with indexes in train for training w2v)\n",
    "def import_mags_and_build_model (end = 25, path_all_fasta = path_all_fasta):\n",
    "    \n",
    "    print ('Importing MAGs')\n",
    "    \n",
    "    # There are 1364 MAGs enclosed in FASTA files\n",
    "    # I have to traverse every FASTA file, and in each file every sequence\n",
    "    \n",
    "    fasta_files = [i for i in os.listdir(path_all_fasta) if (i[-2:] == 'fa')]\n",
    "    \n",
    "    # This was done so that I could work with first 100 FASTA files only. Otherwise, I should just remove: i, and enumerate\n",
    "    for i, fasta_file_name in enumerate(fasta_files):\n",
    "        \n",
    "        if i == end:\n",
    "            break\n",
    "        \n",
    "        else:\n",
    "            with open(os.path.join(path_all_fasta, fasta_file_name), 'r') as input_file:\n",
    "                \n",
    "                one_mag = []\n",
    "                for fasta_string in SeqIO.parse(input_file, \"fasta\"):\n",
    "                    \n",
    "                    # Get kmers of a genome and create a sentence (list of words)\n",
    "                    temp_kmers = split_genome (str(fasta_string.seq))\n",
    "                    \n",
    "                    # Create a document (list of sentences)\n",
    "                    one_mag.append(temp_kmers)\n",
    "                    \n",
    "                # If we do not have a model, we build one\n",
    "                if i == 0:\n",
    "                    print ('Building w2v model')\n",
    "                    # We build our model on the first MAG\n",
    "                    w2v_model = Word2Vec (sentences = one_mag, size = 100, workers = num_of_workers, seed=SEED)\n",
    "                \n",
    "                # Else we just expand its vocabulary\n",
    "                else:\n",
    "                    # Now we expand our vocabulary\n",
    "                    w2v_model.build_vocab (one_mag, update = True)\n",
    "                    \n",
    "    print ('Finished building')\n",
    "    \n",
    "    return w2v_model\n",
    "\n",
    "\n",
    "def train_model (w2v_model, epochs, end = 25):\n",
    "    \n",
    "    print ('Starting model training')\n",
    "    \n",
    "    # There are 1364 MAGs enclosed in FASTA files\n",
    "    # I have to traverse every FASTA file, and in each file every sequence\n",
    "    \n",
    "    fasta_files = [i for i in os.listdir(path_all_fasta) if (i[-2:] == 'fa')]\n",
    "    \n",
    "    # This was done so that I could work with first 100 FASTA files only. Otherwise, I should just remove: i, and enumerate\n",
    "    for i, fasta_file_name in enumerate(fasta_files):\n",
    "        \n",
    "        if i == end:\n",
    "            break\n",
    "        \n",
    "        else:\n",
    "            with open(os.path.join(path_all_fasta, fasta_file_name), 'r') as input_file:\n",
    "                \n",
    "                one_mag = []\n",
    "                for fasta_string in SeqIO.parse(input_file, \"fasta\"):\n",
    "                    \n",
    "                    # Get kmers of a genome and create a sentence (list of words)\n",
    "                    temp_kmers = split_genome (str(fasta_string.seq))\n",
    "                    \n",
    "                    # Create a document (list of sentences)\n",
    "                    one_mag.append(temp_kmers)\n",
    "                    \n",
    "                \n",
    "                w2v_model.train (one_mag, total_examples = w2v_model.corpus_count, epochs = epochs)\n",
    "                    \n",
    "    print ('Model training finished')\n",
    "    \n",
    "    return w2v_model\n",
    "\n",
    "\n",
    "def visualize_with_pca (data, labels):\n",
    "    \n",
    "    pca_model = PCA (n_components = 2, random_state = SEED)\n",
    "    data_transformed = pca_model.fit_transform (data)\n",
    "    data_transformed = pd.DataFrame (data_transformed)\n",
    "    data_transformed.columns = ['PC_1', 'PC_2']\n",
    "    data_transformed['Labels'] = labels\n",
    "    \n",
    "    chart = alt.Chart(data_transformed).mark_circle().encode(\n",
    "        alt.X ('PC_1:Q'),\n",
    "        alt.Y ('PC_2:Q'),\n",
    "        alt.Color ('Labels:N')\n",
    "    )\n",
    "    \n",
    "    return chart\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "welcome-temperature",
   "metadata": {},
   "source": [
    "---\n",
    "## Importing Metabolomic data\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "incident-doctor",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Metabolite</th>\n",
       "      <th>tp</th>\n",
       "      <th>date</th>\n",
       "      <th>type</th>\n",
       "      <th>known_type</th>\n",
       "      <th>means</th>\n",
       "      <th>medians</th>\n",
       "      <th>sds</th>\n",
       "      <th>N</th>\n",
       "      <th>se</th>\n",
       "      <th>ci</th>\n",
       "      <th>type2</th>\n",
       "      <th>measurement</th>\n",
       "      <th>KEGG.Compound.ID</th>\n",
       "      <th>Chebi.Name</th>\n",
       "      <th>Chebi.Name_combined</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>!Phosphoric_acid_3TMS</td>\n",
       "      <td>D1</td>\n",
       "      <td>2012-04-17</td>\n",
       "      <td>snp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.757113</td>\n",
       "      <td>0.737943</td>\n",
       "      <td>0.098650</td>\n",
       "      <td>3</td>\n",
       "      <td>0.056956</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>C00009</td>\n",
       "      <td>phosphoric acid</td>\n",
       "      <td>phosphoric acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>!Phosphoric_acid_3TMS</td>\n",
       "      <td>D10</td>\n",
       "      <td>2012-01-25</td>\n",
       "      <td>snp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.720935</td>\n",
       "      <td>0.873509</td>\n",
       "      <td>0.266593</td>\n",
       "      <td>3</td>\n",
       "      <td>0.153918</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>C00009</td>\n",
       "      <td>phosphoric acid</td>\n",
       "      <td>phosphoric acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>!Phosphoric_acid_3TMS</td>\n",
       "      <td>D11</td>\n",
       "      <td>2012-03-22</td>\n",
       "      <td>snp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.625047</td>\n",
       "      <td>0.652268</td>\n",
       "      <td>0.057820</td>\n",
       "      <td>3</td>\n",
       "      <td>0.033382</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>C00009</td>\n",
       "      <td>phosphoric acid</td>\n",
       "      <td>phosphoric acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>!Phosphoric_acid_3TMS</td>\n",
       "      <td>D12</td>\n",
       "      <td>2012-01-19</td>\n",
       "      <td>snp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.964956</td>\n",
       "      <td>0.941878</td>\n",
       "      <td>0.053754</td>\n",
       "      <td>3</td>\n",
       "      <td>0.031035</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>C00009</td>\n",
       "      <td>phosphoric acid</td>\n",
       "      <td>phosphoric acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>!Phosphoric_acid_3TMS</td>\n",
       "      <td>D13</td>\n",
       "      <td>2011-05-13</td>\n",
       "      <td>snp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.963984</td>\n",
       "      <td>0.974479</td>\n",
       "      <td>0.071659</td>\n",
       "      <td>3</td>\n",
       "      <td>0.041373</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>C00009</td>\n",
       "      <td>phosphoric acid</td>\n",
       "      <td>phosphoric acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18865</th>\n",
       "      <td>Valine_2TMS</td>\n",
       "      <td>D51</td>\n",
       "      <td>2011-08-11</td>\n",
       "      <td>bp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.704821</td>\n",
       "      <td>0.478251</td>\n",
       "      <td>0.428368</td>\n",
       "      <td>3</td>\n",
       "      <td>0.247319</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C16436,C06417,C00183</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18866</th>\n",
       "      <td>Valine_2TMS</td>\n",
       "      <td>D6</td>\n",
       "      <td>2011-06-03</td>\n",
       "      <td>bp</td>\n",
       "      <td>known</td>\n",
       "      <td>1.085888</td>\n",
       "      <td>1.197471</td>\n",
       "      <td>0.487484</td>\n",
       "      <td>3</td>\n",
       "      <td>0.281449</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C16436,C06417,C00183</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18867</th>\n",
       "      <td>Valine_2TMS</td>\n",
       "      <td>D7</td>\n",
       "      <td>2012-04-04</td>\n",
       "      <td>bp</td>\n",
       "      <td>known</td>\n",
       "      <td>2.958023</td>\n",
       "      <td>2.813145</td>\n",
       "      <td>0.341213</td>\n",
       "      <td>3</td>\n",
       "      <td>0.197000</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C16436,C06417,C00183</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18868</th>\n",
       "      <td>Valine_2TMS</td>\n",
       "      <td>D8</td>\n",
       "      <td>2012-02-08</td>\n",
       "      <td>bp</td>\n",
       "      <td>known</td>\n",
       "      <td>2.515980</td>\n",
       "      <td>2.322248</td>\n",
       "      <td>0.618556</td>\n",
       "      <td>3</td>\n",
       "      <td>0.357123</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C16436,C06417,C00183</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18869</th>\n",
       "      <td>Valine_2TMS</td>\n",
       "      <td>D9</td>\n",
       "      <td>2011-12-28</td>\n",
       "      <td>bp</td>\n",
       "      <td>known</td>\n",
       "      <td>2.356989</td>\n",
       "      <td>2.280757</td>\n",
       "      <td>0.264918</td>\n",
       "      <td>3</td>\n",
       "      <td>0.152950</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C16436,C06417,C00183</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "      <td>valine;D-valine;L-valine</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18870 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Metabolite   tp        date type known_type     means  \\\n",
       "0      !Phosphoric_acid_3TMS   D1  2012-04-17  snp      known  0.757113   \n",
       "1      !Phosphoric_acid_3TMS  D10  2012-01-25  snp      known  0.720935   \n",
       "2      !Phosphoric_acid_3TMS  D11  2012-03-22  snp      known  0.625047   \n",
       "3      !Phosphoric_acid_3TMS  D12  2012-01-19  snp      known  0.964956   \n",
       "4      !Phosphoric_acid_3TMS  D13  2011-05-13  snp      known  0.963984   \n",
       "...                      ...  ...         ...  ...        ...       ...   \n",
       "18865            Valine_2TMS  D51  2011-08-11   bp      known  0.704821   \n",
       "18866            Valine_2TMS   D6  2011-06-03   bp      known  1.085888   \n",
       "18867            Valine_2TMS   D7  2012-04-04   bp      known  2.958023   \n",
       "18868            Valine_2TMS   D8  2012-02-08   bp      known  2.515980   \n",
       "18869            Valine_2TMS   D9  2011-12-28   bp      known  2.356989   \n",
       "\n",
       "        medians       sds  N        se        ci          type2 measurement  \\\n",
       "0      0.737943  0.098650  3  0.056956  4.302653  extracellular    nonpolar   \n",
       "1      0.873509  0.266593  3  0.153918  4.302653  extracellular    nonpolar   \n",
       "2      0.652268  0.057820  3  0.033382  4.302653  extracellular    nonpolar   \n",
       "3      0.941878  0.053754  3  0.031035  4.302653  extracellular    nonpolar   \n",
       "4      0.974479  0.071659  3  0.041373  4.302653  extracellular    nonpolar   \n",
       "...         ...       ... ..       ...       ...            ...         ...   \n",
       "18865  0.478251  0.428368  3  0.247319  4.302653  intracellular       polar   \n",
       "18866  1.197471  0.487484  3  0.281449  4.302653  intracellular       polar   \n",
       "18867  2.813145  0.341213  3  0.197000  4.302653  intracellular       polar   \n",
       "18868  2.322248  0.618556  3  0.357123  4.302653  intracellular       polar   \n",
       "18869  2.280757  0.264918  3  0.152950  4.302653  intracellular       polar   \n",
       "\n",
       "           KEGG.Compound.ID                Chebi.Name  \\\n",
       "0                    C00009           phosphoric acid   \n",
       "1                    C00009           phosphoric acid   \n",
       "2                    C00009           phosphoric acid   \n",
       "3                    C00009           phosphoric acid   \n",
       "4                    C00009           phosphoric acid   \n",
       "...                     ...                       ...   \n",
       "18865  C16436,C06417,C00183  valine;D-valine;L-valine   \n",
       "18866  C16436,C06417,C00183  valine;D-valine;L-valine   \n",
       "18867  C16436,C06417,C00183  valine;D-valine;L-valine   \n",
       "18868  C16436,C06417,C00183  valine;D-valine;L-valine   \n",
       "18869  C16436,C06417,C00183  valine;D-valine;L-valine   \n",
       "\n",
       "            Chebi.Name_combined  \n",
       "0               phosphoric acid  \n",
       "1               phosphoric acid  \n",
       "2               phosphoric acid  \n",
       "3               phosphoric acid  \n",
       "4               phosphoric acid  \n",
       "...                         ...  \n",
       "18865  valine;D-valine;L-valine  \n",
       "18866  valine;D-valine;L-valine  \n",
       "18867  valine;D-valine;L-valine  \n",
       "18868  valine;D-valine;L-valine  \n",
       "18869  valine;D-valine;L-valine  \n",
       "\n",
       "[18870 rows x 16 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metabolomics_file_name = os.path.join(path_normalised_metabolomics, os.listdir(path_normalised_metabolomics)[0])\n",
    "metabolomics_df = pd.read_csv (metabolomics_file_name, delimiter = '\\t')\n",
    "metabolomics_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "elect-bahamas",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>date</th>\n",
       "      <th>Metabolite</th>\n",
       "      <th>tp</th>\n",
       "      <th>type</th>\n",
       "      <th>known_type</th>\n",
       "      <th>means</th>\n",
       "      <th>medians</th>\n",
       "      <th>sds</th>\n",
       "      <th>N</th>\n",
       "      <th>se</th>\n",
       "      <th>ci</th>\n",
       "      <th>type2</th>\n",
       "      <th>measurement</th>\n",
       "      <th>KEGG.Compound.ID</th>\n",
       "      <th>Chebi.Name</th>\n",
       "      <th>Chebi.Name_combined</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>Leucine_2TMS</td>\n",
       "      <td>D37</td>\n",
       "      <td>sp</td>\n",
       "      <td>known</td>\n",
       "      <td>1.440122</td>\n",
       "      <td>1.489979</td>\n",
       "      <td>0.146440</td>\n",
       "      <td>3</td>\n",
       "      <td>0.084547</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C16439,C00123,C01570</td>\n",
       "      <td>leucine;L-leucine;D-leucine</td>\n",
       "      <td>leucine;L-leucine;D-leucine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>No match: 1181.78_EM_SP_D1_1_223</td>\n",
       "      <td>D37</td>\n",
       "      <td>sp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>3.778469</td>\n",
       "      <td>3.801417</td>\n",
       "      <td>1.924377</td>\n",
       "      <td>3</td>\n",
       "      <td>1.111040</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>unknown#emu_WW_2144.25</td>\n",
       "      <td>D37</td>\n",
       "      <td>bnp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.949545</td>\n",
       "      <td>0.947209</td>\n",
       "      <td>0.065422</td>\n",
       "      <td>3</td>\n",
       "      <td>0.037771</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>Unknown#bth_pae_001</td>\n",
       "      <td>D37</td>\n",
       "      <td>bp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2.095116</td>\n",
       "      <td>1.809891</td>\n",
       "      <td>1.916896</td>\n",
       "      <td>3</td>\n",
       "      <td>1.106721</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2011-03-21</td>\n",
       "      <td>No match: 2220.34_EM_BP_D1_1_192</td>\n",
       "      <td>D37</td>\n",
       "      <td>bp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.631643</td>\n",
       "      <td>1.571125</td>\n",
       "      <td>0.168501</td>\n",
       "      <td>3</td>\n",
       "      <td>0.097284</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18865</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>Octadecanol_1TMS</td>\n",
       "      <td>D48</td>\n",
       "      <td>bnp</td>\n",
       "      <td>known</td>\n",
       "      <td>1.364105</td>\n",
       "      <td>1.457780</td>\n",
       "      <td>0.170311</td>\n",
       "      <td>3</td>\n",
       "      <td>0.098329</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>D01924</td>\n",
       "      <td>octadecan-1-ol</td>\n",
       "      <td>octadecan-1-ol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18866</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>No match: 1447.93_EM_BP_D2_1_57</td>\n",
       "      <td>D48</td>\n",
       "      <td>bp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>4.815774</td>\n",
       "      <td>1.160165</td>\n",
       "      <td>6.835474</td>\n",
       "      <td>3</td>\n",
       "      <td>3.946463</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18867</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>Lactulose_8TMS</td>\n",
       "      <td>D48</td>\n",
       "      <td>bp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.697489</td>\n",
       "      <td>0.684339</td>\n",
       "      <td>0.049632</td>\n",
       "      <td>3</td>\n",
       "      <td>0.028655</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C07064</td>\n",
       "      <td>lactulose</td>\n",
       "      <td>lactulose</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18868</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>No match: 2004.78_EM_BNP_D10_1_205</td>\n",
       "      <td>D48</td>\n",
       "      <td>bnp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.785996</td>\n",
       "      <td>0.744529</td>\n",
       "      <td>0.139901</td>\n",
       "      <td>3</td>\n",
       "      <td>0.080772</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>intracellular</td>\n",
       "      <td>nonpolar</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18869</th>\n",
       "      <td>2012-05-03</td>\n",
       "      <td>2-Oxobutanoic_acid_1_1MeOX_1TMS</td>\n",
       "      <td>D48</td>\n",
       "      <td>sp</td>\n",
       "      <td>known</td>\n",
       "      <td>0.755565</td>\n",
       "      <td>0.709972</td>\n",
       "      <td>0.149214</td>\n",
       "      <td>3</td>\n",
       "      <td>0.086149</td>\n",
       "      <td>4.302653</td>\n",
       "      <td>extracellular</td>\n",
       "      <td>polar</td>\n",
       "      <td>C00109</td>\n",
       "      <td>2-oxobutanoic acid</td>\n",
       "      <td>2-oxobutanoic acid</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18870 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            date                          Metabolite   tp type known_type  \\\n",
       "0     2011-03-21                        Leucine_2TMS  D37   sp      known   \n",
       "1     2011-03-21    No match: 1181.78_EM_SP_D1_1_223  D37   sp    unknown   \n",
       "2     2011-03-21              unknown#emu_WW_2144.25  D37  bnp    unknown   \n",
       "3     2011-03-21                 Unknown#bth_pae_001  D37   bp    unknown   \n",
       "4     2011-03-21    No match: 2220.34_EM_BP_D1_1_192  D37   bp    unknown   \n",
       "...          ...                                 ...  ...  ...        ...   \n",
       "18865 2012-05-03                    Octadecanol_1TMS  D48  bnp      known   \n",
       "18866 2012-05-03     No match: 1447.93_EM_BP_D2_1_57  D48   bp    unknown   \n",
       "18867 2012-05-03                      Lactulose_8TMS  D48   bp      known   \n",
       "18868 2012-05-03  No match: 2004.78_EM_BNP_D10_1_205  D48  bnp    unknown   \n",
       "18869 2012-05-03     2-Oxobutanoic_acid_1_1MeOX_1TMS  D48   sp      known   \n",
       "\n",
       "          means   medians       sds  N        se        ci          type2  \\\n",
       "0      1.440122  1.489979  0.146440  3  0.084547  4.302653  extracellular   \n",
       "1      3.778469  3.801417  1.924377  3  1.111040  4.302653  extracellular   \n",
       "2      0.949545  0.947209  0.065422  3  0.037771  4.302653  intracellular   \n",
       "3      2.095116  1.809891  1.916896  3  1.106721  4.302653  intracellular   \n",
       "4      1.631643  1.571125  0.168501  3  0.097284  4.302653  intracellular   \n",
       "...         ...       ...       ... ..       ...       ...            ...   \n",
       "18865  1.364105  1.457780  0.170311  3  0.098329  4.302653  intracellular   \n",
       "18866  4.815774  1.160165  6.835474  3  3.946463  4.302653  intracellular   \n",
       "18867  0.697489  0.684339  0.049632  3  0.028655  4.302653  intracellular   \n",
       "18868  0.785996  0.744529  0.139901  3  0.080772  4.302653  intracellular   \n",
       "18869  0.755565  0.709972  0.149214  3  0.086149  4.302653  extracellular   \n",
       "\n",
       "      measurement      KEGG.Compound.ID                   Chebi.Name  \\\n",
       "0           polar  C16439,C00123,C01570  leucine;L-leucine;D-leucine   \n",
       "1           polar                   NaN                          NaN   \n",
       "2        nonpolar                   NaN                          NaN   \n",
       "3           polar                   NaN                          NaN   \n",
       "4           polar                   NaN                          NaN   \n",
       "...           ...                   ...                          ...   \n",
       "18865    nonpolar                D01924               octadecan-1-ol   \n",
       "18866       polar                   NaN                          NaN   \n",
       "18867       polar                C07064                    lactulose   \n",
       "18868    nonpolar                   NaN                          NaN   \n",
       "18869       polar                C00109           2-oxobutanoic acid   \n",
       "\n",
       "               Chebi.Name_combined  \n",
       "0      leucine;L-leucine;D-leucine  \n",
       "1                              NaN  \n",
       "2                              NaN  \n",
       "3                              NaN  \n",
       "4                              NaN  \n",
       "...                            ...  \n",
       "18865               octadecan-1-ol  \n",
       "18866                          NaN  \n",
       "18867                    lactulose  \n",
       "18868                          NaN  \n",
       "18869           2-oxobutanoic acid  \n",
       "\n",
       "[18870 rows x 16 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metabolomics_df['date'] = pd.to_datetime(metabolomics_df['date'])\n",
    "metabolomics_df.insert (0, 'date', metabolomics_df.pop('date'))\n",
    "metabolomics_df.sort_values ('date', inplace = True, ignore_index = True)\n",
    "metabolomics_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "special-montreal",
   "metadata": {},
   "source": [
    "## MAG examination"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "silver-depression",
   "metadata": {},
   "source": [
    "To do:\n",
    "1. Make kmers of some length, out of every word in every sentence\n",
    "2. Encode every sentence with Word2Vec and CBOW (because it's faster than skip-gram)\n",
    "3. Reduce dimensionality with PCA or with autoencoder\n",
    "    > Maybe it is good to do the first 3 steps while importing, for each and every sentence. That way, I might use less memory\n",
    "4. Cluster similar sentences (MAGs) to get rMAGs\n",
    "\n",
    "> Important: this is not temporal data, since there are no timestamps!\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "noticed-association",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Importing MAGs\n",
      "Building w2v model\n",
      "Finished building\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# FOR CLUSTERING I SHOULD CREATE A DATAFRAME WITH MAGs INDEXES AND THEIR VECTOR REPRESENTATIONS\n",
    "\n",
    "final_model = import_mags_and_build_model (end = END, path_all_fasta = path_all_fasta)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "coral-hypothesis",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('CCGGC', 0.6029188632965088),\n",
       " ('ACGGA', 0.5602467060089111),\n",
       " ('TCGGC', 0.5392060279846191),\n",
       " ('GCGGC', 0.5378471612930298),\n",
       " ('ACGGT', 0.48065903782844543),\n",
       " ('ACGGG', 0.4784856140613556),\n",
       " ('CTACG', 0.28459882736206055),\n",
       " ('TACGA', 0.2798745036125183),\n",
       " ('AACGA', 0.2734222412109375),\n",
       " ('ACGTT', 0.273221492767334)]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_model.wv.most_similar('ACGGC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "demonstrated-thinking",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting model training\n",
      "Model training finished\n"
     ]
    }
   ],
   "source": [
    "# Train model. It tooks ~10 minutes for END = 25 amount of MAGs\n",
    "epochs = 10\n",
    "\n",
    "final_model = train_model (final_model, epochs = epochs, end = END)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "toxic-christopher",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_model.wv.save_word2vec_format(os.path.join (path_model_save_root, 'model_25.bin'), binary=True) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "undefined-wagner",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "south-marijuana",
   "metadata": {},
   "source": [
    "Now I should vectorize documents with this model. For further use, I could save this model's weights, and use it to vectorize all mags. That would take a lot, but every MAG will have its vector representation\n",
    "> This could be done by importing one MAG at a time, then tokenizing it (like before), then getting vector representations of that MAG's sentences (genomes) and then finding the vector representation of the whole MAG (document). If I do that for one MAG at a time, There is no need to worry about memory\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "adjacent-state",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Vectorizing MAGs\n",
      "Finished vectorizing\n"
     ]
    }
   ],
   "source": [
    "list_of_mag_vectors = vectorize_mags (final_model, end = END)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "chicken-guidance",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[array([ 0.3447247 , -0.98493254,  0.31138113,  0.9457629 , -0.6776731 ,\n",
       "        -0.16597524,  0.9591458 ,  2.5192375 ,  0.8832711 , -0.2430107 ,\n",
       "         1.6672454 , -2.3872983 ,  0.65017176, -0.28312284, -1.4965318 ,\n",
       "        -1.2259525 ,  0.23684104, -0.35258517,  0.08118253,  0.62865   ,\n",
       "        -0.17559732, -0.10104723,  1.0475872 ,  0.9804583 ,  1.5938567 ,\n",
       "        -1.3744875 , -0.5204637 ,  0.01642581,  1.2500559 ,  2.1099281 ,\n",
       "        -0.49030125,  0.6041928 , -0.87914413,  0.08175804,  1.5866882 ,\n",
       "        -0.3342121 , -0.83550924, -1.4288213 ,  0.07086587, -1.9811511 ,\n",
       "        -1.0529662 , -0.990749  ,  2.3318615 , -0.08571481,  1.5839869 ,\n",
       "         0.05786707,  2.2484546 , -1.2275187 , -0.3562304 ,  1.1188687 ,\n",
       "        -0.13782707, -0.03078756, -1.6636964 ,  1.1800835 ,  1.4556597 ,\n",
       "        -1.3416642 , -0.05779186,  1.6490489 , -0.62548006,  1.5076993 ,\n",
       "         1.0270246 ,  0.828531  , -0.2174203 , -0.28671232, -1.3789287 ,\n",
       "        -0.3863798 ,  0.354306  , -0.7671363 , -0.50262713,  1.1501355 ,\n",
       "         0.326566  ,  0.06137726, -0.7858441 , -1.7276732 ,  0.74253213,\n",
       "        -2.350372  , -0.36903518, -0.68736947, -2.2025032 ,  1.1155628 ,\n",
       "        -1.5600027 , -0.9776511 ,  1.992551  , -1.7193599 , -0.43857273,\n",
       "        -0.80924124,  1.0510292 ,  1.0522406 ,  0.6792343 , -0.04714993,\n",
       "         1.0273055 , -0.68425584,  0.8423442 ,  0.98279345, -1.1151959 ,\n",
       "        -1.3894882 ,  0.3606208 , -2.1903095 ,  1.0444397 ,  0.3445911 ],\n",
       "       dtype=float32),\n",
       " array([ 0.0862533 ,  0.48152915,  0.28543922, -1.5974874 ,  0.08400733,\n",
       "         0.5027265 , -1.1244688 , -2.7118134 , -0.08269083,  0.80805624,\n",
       "        -2.5342026 ,  1.0569525 , -1.4793999 ,  0.26065853,  0.97189367,\n",
       "         0.3608742 , -0.5281864 ,  0.57776827, -0.18825257,  0.60462105,\n",
       "        -0.02812269,  1.6119161 , -2.4362185 ,  0.11042367, -1.3866956 ,\n",
       "         3.0519419 , -0.13126701, -0.15497635, -1.0448885 , -1.9309072 ,\n",
       "         0.6863494 , -0.23142286, -0.63116235, -1.638833  , -0.3422584 ,\n",
       "         0.01597553,  1.1347748 ,  0.7428918 ,  1.0702393 ,  1.074417  ,\n",
       "         0.6996188 ,  0.25205404, -1.0696585 , -0.6027482 , -1.8014576 ,\n",
       "        -1.4292419 , -3.2431657 , -0.42430863,  1.0132906 , -0.758023  ,\n",
       "         0.563429  , -1.0823264 ,  0.63374925, -0.19125819, -0.42829555,\n",
       "         0.12681372, -0.06488036, -0.6363876 ,  0.5740274 ,  0.68306565,\n",
       "         0.19315882, -1.4799703 ,  0.8212972 ,  0.18644463,  0.58301115,\n",
       "         0.94935393, -2.0438788 , -0.6254538 ,  0.36048505, -0.5966196 ,\n",
       "         0.2869914 , -0.87278944,  2.216706  ,  1.3083268 , -1.5392052 ,\n",
       "        -0.5063748 ,  0.20029314, -0.17192197,  1.2717096 , -0.24208985,\n",
       "         0.46685755,  0.61918426, -1.5395591 ,  0.8831208 , -1.391985  ,\n",
       "        -0.33395845, -1.0784066 , -0.79115766,  0.67973137, -0.84917355,\n",
       "         0.7519467 , -0.5672727 , -0.5476874 , -0.01814327,  0.4825466 ,\n",
       "         1.2929289 , -0.4804222 ,  2.6266718 , -1.3837788 , -0.45134282],\n",
       "       dtype=float32)]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_of_mag_vectors[0:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "korean-package",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>90</th>\n",
       "      <th>91</th>\n",
       "      <th>92</th>\n",
       "      <th>93</th>\n",
       "      <th>94</th>\n",
       "      <th>95</th>\n",
       "      <th>96</th>\n",
       "      <th>97</th>\n",
       "      <th>98</th>\n",
       "      <th>99</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.344725</td>\n",
       "      <td>-0.984933</td>\n",
       "      <td>0.311381</td>\n",
       "      <td>0.945763</td>\n",
       "      <td>-0.677673</td>\n",
       "      <td>-0.165975</td>\n",
       "      <td>0.959146</td>\n",
       "      <td>2.519238</td>\n",
       "      <td>0.883271</td>\n",
       "      <td>-0.243011</td>\n",
       "      <td>...</td>\n",
       "      <td>1.027305</td>\n",
       "      <td>-0.684256</td>\n",
       "      <td>0.842344</td>\n",
       "      <td>0.982793</td>\n",
       "      <td>-1.115196</td>\n",
       "      <td>-1.389488</td>\n",
       "      <td>0.360621</td>\n",
       "      <td>-2.190310</td>\n",
       "      <td>1.044440</td>\n",
       "      <td>0.344591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.086253</td>\n",
       "      <td>0.481529</td>\n",
       "      <td>0.285439</td>\n",
       "      <td>-1.597487</td>\n",
       "      <td>0.084007</td>\n",
       "      <td>0.502726</td>\n",
       "      <td>-1.124469</td>\n",
       "      <td>-2.711813</td>\n",
       "      <td>-0.082691</td>\n",
       "      <td>0.808056</td>\n",
       "      <td>...</td>\n",
       "      <td>0.751947</td>\n",
       "      <td>-0.567273</td>\n",
       "      <td>-0.547687</td>\n",
       "      <td>-0.018143</td>\n",
       "      <td>0.482547</td>\n",
       "      <td>1.292929</td>\n",
       "      <td>-0.480422</td>\n",
       "      <td>2.626672</td>\n",
       "      <td>-1.383779</td>\n",
       "      <td>-0.451343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-0.818952</td>\n",
       "      <td>-1.239758</td>\n",
       "      <td>0.806250</td>\n",
       "      <td>1.519856</td>\n",
       "      <td>-0.290418</td>\n",
       "      <td>-0.264609</td>\n",
       "      <td>0.906635</td>\n",
       "      <td>2.390739</td>\n",
       "      <td>-0.411772</td>\n",
       "      <td>-0.718464</td>\n",
       "      <td>...</td>\n",
       "      <td>0.274231</td>\n",
       "      <td>-0.460688</td>\n",
       "      <td>1.217699</td>\n",
       "      <td>0.447533</td>\n",
       "      <td>-0.244538</td>\n",
       "      <td>-2.007320</td>\n",
       "      <td>1.051488</td>\n",
       "      <td>-2.756425</td>\n",
       "      <td>2.020143</td>\n",
       "      <td>0.185595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.087183</td>\n",
       "      <td>0.527681</td>\n",
       "      <td>0.683097</td>\n",
       "      <td>-1.658966</td>\n",
       "      <td>-0.213686</td>\n",
       "      <td>0.772661</td>\n",
       "      <td>-1.532721</td>\n",
       "      <td>-3.115765</td>\n",
       "      <td>-0.120055</td>\n",
       "      <td>1.104269</td>\n",
       "      <td>...</td>\n",
       "      <td>0.539762</td>\n",
       "      <td>-0.430546</td>\n",
       "      <td>-0.241169</td>\n",
       "      <td>-0.086811</td>\n",
       "      <td>0.808656</td>\n",
       "      <td>1.170911</td>\n",
       "      <td>-0.667988</td>\n",
       "      <td>2.736131</td>\n",
       "      <td>-1.349261</td>\n",
       "      <td>-0.597348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.104078</td>\n",
       "      <td>0.523030</td>\n",
       "      <td>0.664976</td>\n",
       "      <td>-1.584116</td>\n",
       "      <td>-0.150568</td>\n",
       "      <td>0.790583</td>\n",
       "      <td>-1.426932</td>\n",
       "      <td>-3.024777</td>\n",
       "      <td>-0.114511</td>\n",
       "      <td>1.100467</td>\n",
       "      <td>...</td>\n",
       "      <td>0.484436</td>\n",
       "      <td>-0.492954</td>\n",
       "      <td>-0.212373</td>\n",
       "      <td>-0.144619</td>\n",
       "      <td>0.706650</td>\n",
       "      <td>1.106278</td>\n",
       "      <td>-0.596556</td>\n",
       "      <td>2.609393</td>\n",
       "      <td>-1.271026</td>\n",
       "      <td>-0.525063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.014663</td>\n",
       "      <td>0.432116</td>\n",
       "      <td>1.100762</td>\n",
       "      <td>-0.587587</td>\n",
       "      <td>-0.268986</td>\n",
       "      <td>0.792427</td>\n",
       "      <td>-1.194816</td>\n",
       "      <td>-2.399201</td>\n",
       "      <td>-0.469793</td>\n",
       "      <td>1.280533</td>\n",
       "      <td>...</td>\n",
       "      <td>0.088165</td>\n",
       "      <td>-0.541312</td>\n",
       "      <td>0.376861</td>\n",
       "      <td>-0.096240</td>\n",
       "      <td>0.815520</td>\n",
       "      <td>0.834893</td>\n",
       "      <td>-0.308112</td>\n",
       "      <td>1.879369</td>\n",
       "      <td>0.026564</td>\n",
       "      <td>-0.450374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>-0.604413</td>\n",
       "      <td>-1.880543</td>\n",
       "      <td>-0.051779</td>\n",
       "      <td>2.134320</td>\n",
       "      <td>-0.342273</td>\n",
       "      <td>-0.850293</td>\n",
       "      <td>1.257860</td>\n",
       "      <td>4.233354</td>\n",
       "      <td>-0.423073</td>\n",
       "      <td>-1.518789</td>\n",
       "      <td>...</td>\n",
       "      <td>0.403746</td>\n",
       "      <td>-0.517206</td>\n",
       "      <td>0.977598</td>\n",
       "      <td>0.030805</td>\n",
       "      <td>-0.722312</td>\n",
       "      <td>-2.436873</td>\n",
       "      <td>1.935860</td>\n",
       "      <td>-4.078766</td>\n",
       "      <td>1.318376</td>\n",
       "      <td>0.259708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>-0.466918</td>\n",
       "      <td>-1.944027</td>\n",
       "      <td>0.441384</td>\n",
       "      <td>1.657203</td>\n",
       "      <td>-0.654853</td>\n",
       "      <td>-0.361336</td>\n",
       "      <td>1.080996</td>\n",
       "      <td>3.812093</td>\n",
       "      <td>0.016572</td>\n",
       "      <td>-1.234429</td>\n",
       "      <td>...</td>\n",
       "      <td>0.813006</td>\n",
       "      <td>-0.365175</td>\n",
       "      <td>1.063718</td>\n",
       "      <td>0.750750</td>\n",
       "      <td>-0.798019</td>\n",
       "      <td>-2.832770</td>\n",
       "      <td>1.128418</td>\n",
       "      <td>-3.906541</td>\n",
       "      <td>1.899582</td>\n",
       "      <td>0.186630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.351202</td>\n",
       "      <td>-0.651379</td>\n",
       "      <td>0.771672</td>\n",
       "      <td>0.734120</td>\n",
       "      <td>-0.723834</td>\n",
       "      <td>0.184430</td>\n",
       "      <td>0.033661</td>\n",
       "      <td>0.868674</td>\n",
       "      <td>0.213349</td>\n",
       "      <td>0.448700</td>\n",
       "      <td>...</td>\n",
       "      <td>0.851811</td>\n",
       "      <td>-0.615773</td>\n",
       "      <td>0.845860</td>\n",
       "      <td>0.614016</td>\n",
       "      <td>-0.545951</td>\n",
       "      <td>-0.934811</td>\n",
       "      <td>0.101101</td>\n",
       "      <td>-0.997692</td>\n",
       "      <td>1.270653</td>\n",
       "      <td>0.073550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>-0.055697</td>\n",
       "      <td>-0.071840</td>\n",
       "      <td>0.714666</td>\n",
       "      <td>0.696301</td>\n",
       "      <td>-0.560185</td>\n",
       "      <td>0.275882</td>\n",
       "      <td>0.003019</td>\n",
       "      <td>0.233549</td>\n",
       "      <td>-0.053218</td>\n",
       "      <td>0.568498</td>\n",
       "      <td>...</td>\n",
       "      <td>0.304297</td>\n",
       "      <td>-0.734093</td>\n",
       "      <td>0.968042</td>\n",
       "      <td>0.301389</td>\n",
       "      <td>0.070306</td>\n",
       "      <td>-0.267304</td>\n",
       "      <td>0.279300</td>\n",
       "      <td>-0.504349</td>\n",
       "      <td>0.932093</td>\n",
       "      <td>0.187632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>-0.237913</td>\n",
       "      <td>-0.598244</td>\n",
       "      <td>0.644930</td>\n",
       "      <td>1.069067</td>\n",
       "      <td>-0.601827</td>\n",
       "      <td>-0.038800</td>\n",
       "      <td>0.791883</td>\n",
       "      <td>1.418516</td>\n",
       "      <td>0.366272</td>\n",
       "      <td>0.062157</td>\n",
       "      <td>...</td>\n",
       "      <td>0.547454</td>\n",
       "      <td>-0.530424</td>\n",
       "      <td>0.942949</td>\n",
       "      <td>0.606467</td>\n",
       "      <td>-0.452768</td>\n",
       "      <td>-0.862444</td>\n",
       "      <td>0.295688</td>\n",
       "      <td>-1.228855</td>\n",
       "      <td>1.678287</td>\n",
       "      <td>0.308781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>-0.148395</td>\n",
       "      <td>-1.105558</td>\n",
       "      <td>0.382042</td>\n",
       "      <td>1.171390</td>\n",
       "      <td>-0.461197</td>\n",
       "      <td>0.149104</td>\n",
       "      <td>0.524762</td>\n",
       "      <td>1.564994</td>\n",
       "      <td>-0.028786</td>\n",
       "      <td>-0.284429</td>\n",
       "      <td>...</td>\n",
       "      <td>0.569270</td>\n",
       "      <td>-0.733589</td>\n",
       "      <td>1.058700</td>\n",
       "      <td>0.463804</td>\n",
       "      <td>-0.480641</td>\n",
       "      <td>-1.502399</td>\n",
       "      <td>0.578655</td>\n",
       "      <td>-1.946262</td>\n",
       "      <td>1.206935</td>\n",
       "      <td>0.145237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>-0.064093</td>\n",
       "      <td>-0.039293</td>\n",
       "      <td>0.743804</td>\n",
       "      <td>0.758627</td>\n",
       "      <td>-0.749009</td>\n",
       "      <td>0.392671</td>\n",
       "      <td>-0.145965</td>\n",
       "      <td>0.127638</td>\n",
       "      <td>-0.143598</td>\n",
       "      <td>0.563006</td>\n",
       "      <td>...</td>\n",
       "      <td>0.309192</td>\n",
       "      <td>-0.657410</td>\n",
       "      <td>1.066946</td>\n",
       "      <td>0.159204</td>\n",
       "      <td>-0.013811</td>\n",
       "      <td>-0.430098</td>\n",
       "      <td>0.264688</td>\n",
       "      <td>-0.373346</td>\n",
       "      <td>0.987297</td>\n",
       "      <td>0.206829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>0.267164</td>\n",
       "      <td>0.386644</td>\n",
       "      <td>0.495075</td>\n",
       "      <td>-0.610948</td>\n",
       "      <td>-0.200776</td>\n",
       "      <td>0.639278</td>\n",
       "      <td>-1.268994</td>\n",
       "      <td>-2.851595</td>\n",
       "      <td>-0.366958</td>\n",
       "      <td>1.045451</td>\n",
       "      <td>...</td>\n",
       "      <td>0.647557</td>\n",
       "      <td>-0.609587</td>\n",
       "      <td>-0.108954</td>\n",
       "      <td>-0.491477</td>\n",
       "      <td>0.401165</td>\n",
       "      <td>1.284045</td>\n",
       "      <td>-0.217907</td>\n",
       "      <td>2.398955</td>\n",
       "      <td>-0.757857</td>\n",
       "      <td>-0.243582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>0.258273</td>\n",
       "      <td>0.386517</td>\n",
       "      <td>0.508212</td>\n",
       "      <td>-0.641460</td>\n",
       "      <td>-0.237646</td>\n",
       "      <td>0.642224</td>\n",
       "      <td>-1.333043</td>\n",
       "      <td>-2.938588</td>\n",
       "      <td>-0.376913</td>\n",
       "      <td>1.043629</td>\n",
       "      <td>...</td>\n",
       "      <td>0.674588</td>\n",
       "      <td>-0.591110</td>\n",
       "      <td>-0.115226</td>\n",
       "      <td>-0.474830</td>\n",
       "      <td>0.510926</td>\n",
       "      <td>1.310520</td>\n",
       "      <td>-0.239643</td>\n",
       "      <td>2.472327</td>\n",
       "      <td>-0.845485</td>\n",
       "      <td>-0.286182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>0.388046</td>\n",
       "      <td>0.303580</td>\n",
       "      <td>0.374766</td>\n",
       "      <td>-0.308154</td>\n",
       "      <td>-0.444269</td>\n",
       "      <td>0.383437</td>\n",
       "      <td>-0.507701</td>\n",
       "      <td>-1.409627</td>\n",
       "      <td>0.147173</td>\n",
       "      <td>0.789374</td>\n",
       "      <td>...</td>\n",
       "      <td>0.671324</td>\n",
       "      <td>-0.648832</td>\n",
       "      <td>0.191926</td>\n",
       "      <td>-0.036511</td>\n",
       "      <td>0.031080</td>\n",
       "      <td>0.851078</td>\n",
       "      <td>-0.247260</td>\n",
       "      <td>1.517264</td>\n",
       "      <td>-0.662310</td>\n",
       "      <td>0.013303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>0.066099</td>\n",
       "      <td>0.449663</td>\n",
       "      <td>0.273700</td>\n",
       "      <td>-1.604737</td>\n",
       "      <td>0.045393</td>\n",
       "      <td>0.500019</td>\n",
       "      <td>-1.166838</td>\n",
       "      <td>-2.682436</td>\n",
       "      <td>-0.082244</td>\n",
       "      <td>0.782839</td>\n",
       "      <td>...</td>\n",
       "      <td>0.767075</td>\n",
       "      <td>-0.581743</td>\n",
       "      <td>-0.533779</td>\n",
       "      <td>0.012559</td>\n",
       "      <td>0.496415</td>\n",
       "      <td>1.274615</td>\n",
       "      <td>-0.468524</td>\n",
       "      <td>2.644280</td>\n",
       "      <td>-1.412417</td>\n",
       "      <td>-0.436565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>0.319671</td>\n",
       "      <td>0.196100</td>\n",
       "      <td>0.048452</td>\n",
       "      <td>-0.877526</td>\n",
       "      <td>-0.286205</td>\n",
       "      <td>0.535245</td>\n",
       "      <td>-0.745279</td>\n",
       "      <td>-1.955159</td>\n",
       "      <td>0.059869</td>\n",
       "      <td>0.729140</td>\n",
       "      <td>...</td>\n",
       "      <td>0.535518</td>\n",
       "      <td>-0.717063</td>\n",
       "      <td>-0.123622</td>\n",
       "      <td>-0.132882</td>\n",
       "      <td>0.254660</td>\n",
       "      <td>1.086064</td>\n",
       "      <td>-0.370138</td>\n",
       "      <td>2.070380</td>\n",
       "      <td>-1.194862</td>\n",
       "      <td>-0.358723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>-0.014087</td>\n",
       "      <td>0.415910</td>\n",
       "      <td>1.118081</td>\n",
       "      <td>-0.517990</td>\n",
       "      <td>-0.231222</td>\n",
       "      <td>0.742162</td>\n",
       "      <td>-1.050030</td>\n",
       "      <td>-2.231887</td>\n",
       "      <td>-0.438067</td>\n",
       "      <td>1.294657</td>\n",
       "      <td>...</td>\n",
       "      <td>0.016728</td>\n",
       "      <td>-0.569043</td>\n",
       "      <td>0.420864</td>\n",
       "      <td>-0.071543</td>\n",
       "      <td>0.768963</td>\n",
       "      <td>0.802863</td>\n",
       "      <td>-0.274859</td>\n",
       "      <td>1.764337</td>\n",
       "      <td>0.126429</td>\n",
       "      <td>-0.408897</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>0.201650</td>\n",
       "      <td>-1.610545</td>\n",
       "      <td>-0.073393</td>\n",
       "      <td>1.509113</td>\n",
       "      <td>-0.820674</td>\n",
       "      <td>-0.118941</td>\n",
       "      <td>0.857989</td>\n",
       "      <td>3.183227</td>\n",
       "      <td>0.732018</td>\n",
       "      <td>-0.657939</td>\n",
       "      <td>...</td>\n",
       "      <td>1.013418</td>\n",
       "      <td>-0.818447</td>\n",
       "      <td>1.032210</td>\n",
       "      <td>0.724327</td>\n",
       "      <td>-1.042743</td>\n",
       "      <td>-2.114437</td>\n",
       "      <td>0.730771</td>\n",
       "      <td>-3.172199</td>\n",
       "      <td>0.979456</td>\n",
       "      <td>0.355066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>0.055401</td>\n",
       "      <td>-0.582526</td>\n",
       "      <td>0.487721</td>\n",
       "      <td>0.440858</td>\n",
       "      <td>-0.619517</td>\n",
       "      <td>0.209370</td>\n",
       "      <td>0.238413</td>\n",
       "      <td>0.541963</td>\n",
       "      <td>0.012372</td>\n",
       "      <td>0.044936</td>\n",
       "      <td>...</td>\n",
       "      <td>0.606983</td>\n",
       "      <td>-0.308914</td>\n",
       "      <td>0.875686</td>\n",
       "      <td>0.392352</td>\n",
       "      <td>-0.076132</td>\n",
       "      <td>-0.785183</td>\n",
       "      <td>-0.033932</td>\n",
       "      <td>-0.792483</td>\n",
       "      <td>0.700628</td>\n",
       "      <td>-0.330127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>0.359764</td>\n",
       "      <td>0.418667</td>\n",
       "      <td>0.185245</td>\n",
       "      <td>-1.167228</td>\n",
       "      <td>-0.418677</td>\n",
       "      <td>0.849160</td>\n",
       "      <td>-1.278315</td>\n",
       "      <td>-2.841949</td>\n",
       "      <td>-0.106587</td>\n",
       "      <td>1.009048</td>\n",
       "      <td>...</td>\n",
       "      <td>0.553045</td>\n",
       "      <td>-0.640072</td>\n",
       "      <td>-0.131178</td>\n",
       "      <td>-0.305709</td>\n",
       "      <td>0.480756</td>\n",
       "      <td>1.187944</td>\n",
       "      <td>-0.610885</td>\n",
       "      <td>2.603128</td>\n",
       "      <td>-1.275098</td>\n",
       "      <td>-0.453794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>-0.422700</td>\n",
       "      <td>-1.774670</td>\n",
       "      <td>-0.224373</td>\n",
       "      <td>2.243236</td>\n",
       "      <td>-0.268128</td>\n",
       "      <td>-0.623241</td>\n",
       "      <td>1.238087</td>\n",
       "      <td>3.553332</td>\n",
       "      <td>-0.163262</td>\n",
       "      <td>-1.207438</td>\n",
       "      <td>...</td>\n",
       "      <td>0.544014</td>\n",
       "      <td>-0.799662</td>\n",
       "      <td>0.991320</td>\n",
       "      <td>0.083721</td>\n",
       "      <td>-0.774948</td>\n",
       "      <td>-2.030537</td>\n",
       "      <td>1.720451</td>\n",
       "      <td>-3.654131</td>\n",
       "      <td>1.314532</td>\n",
       "      <td>0.437816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>0.072642</td>\n",
       "      <td>-0.427801</td>\n",
       "      <td>1.343444</td>\n",
       "      <td>0.664575</td>\n",
       "      <td>-0.550097</td>\n",
       "      <td>0.397489</td>\n",
       "      <td>-0.678432</td>\n",
       "      <td>-1.745667</td>\n",
       "      <td>-0.766703</td>\n",
       "      <td>0.677340</td>\n",
       "      <td>...</td>\n",
       "      <td>0.457688</td>\n",
       "      <td>-0.365134</td>\n",
       "      <td>0.512235</td>\n",
       "      <td>-0.391230</td>\n",
       "      <td>0.423829</td>\n",
       "      <td>0.277259</td>\n",
       "      <td>0.156696</td>\n",
       "      <td>1.145475</td>\n",
       "      <td>0.995143</td>\n",
       "      <td>0.056791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>0.144246</td>\n",
       "      <td>0.530277</td>\n",
       "      <td>0.773706</td>\n",
       "      <td>-0.947087</td>\n",
       "      <td>-0.102724</td>\n",
       "      <td>0.910229</td>\n",
       "      <td>-1.788558</td>\n",
       "      <td>-3.410334</td>\n",
       "      <td>-0.891370</td>\n",
       "      <td>1.438627</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.242513</td>\n",
       "      <td>-0.784440</td>\n",
       "      <td>0.184291</td>\n",
       "      <td>-0.792968</td>\n",
       "      <td>1.372638</td>\n",
       "      <td>1.477131</td>\n",
       "      <td>-0.071839</td>\n",
       "      <td>2.781913</td>\n",
       "      <td>-1.105542</td>\n",
       "      <td>-0.799165</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25 rows × 100 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0         1         2         3         4         5         6   \\\n",
       "0   0.344725 -0.984933  0.311381  0.945763 -0.677673 -0.165975  0.959146   \n",
       "1   0.086253  0.481529  0.285439 -1.597487  0.084007  0.502726 -1.124469   \n",
       "2  -0.818952 -1.239758  0.806250  1.519856 -0.290418 -0.264609  0.906635   \n",
       "3   0.087183  0.527681  0.683097 -1.658966 -0.213686  0.772661 -1.532721   \n",
       "4   0.104078  0.523030  0.664976 -1.584116 -0.150568  0.790583 -1.426932   \n",
       "5   0.014663  0.432116  1.100762 -0.587587 -0.268986  0.792427 -1.194816   \n",
       "6  -0.604413 -1.880543 -0.051779  2.134320 -0.342273 -0.850293  1.257860   \n",
       "7  -0.466918 -1.944027  0.441384  1.657203 -0.654853 -0.361336  1.080996   \n",
       "8   0.351202 -0.651379  0.771672  0.734120 -0.723834  0.184430  0.033661   \n",
       "9  -0.055697 -0.071840  0.714666  0.696301 -0.560185  0.275882  0.003019   \n",
       "10 -0.237913 -0.598244  0.644930  1.069067 -0.601827 -0.038800  0.791883   \n",
       "11 -0.148395 -1.105558  0.382042  1.171390 -0.461197  0.149104  0.524762   \n",
       "12 -0.064093 -0.039293  0.743804  0.758627 -0.749009  0.392671 -0.145965   \n",
       "13  0.267164  0.386644  0.495075 -0.610948 -0.200776  0.639278 -1.268994   \n",
       "14  0.258273  0.386517  0.508212 -0.641460 -0.237646  0.642224 -1.333043   \n",
       "15  0.388046  0.303580  0.374766 -0.308154 -0.444269  0.383437 -0.507701   \n",
       "16  0.066099  0.449663  0.273700 -1.604737  0.045393  0.500019 -1.166838   \n",
       "17  0.319671  0.196100  0.048452 -0.877526 -0.286205  0.535245 -0.745279   \n",
       "18 -0.014087  0.415910  1.118081 -0.517990 -0.231222  0.742162 -1.050030   \n",
       "19  0.201650 -1.610545 -0.073393  1.509113 -0.820674 -0.118941  0.857989   \n",
       "20  0.055401 -0.582526  0.487721  0.440858 -0.619517  0.209370  0.238413   \n",
       "21  0.359764  0.418667  0.185245 -1.167228 -0.418677  0.849160 -1.278315   \n",
       "22 -0.422700 -1.774670 -0.224373  2.243236 -0.268128 -0.623241  1.238087   \n",
       "23  0.072642 -0.427801  1.343444  0.664575 -0.550097  0.397489 -0.678432   \n",
       "24  0.144246  0.530277  0.773706 -0.947087 -0.102724  0.910229 -1.788558   \n",
       "\n",
       "          7         8         9   ...        90        91        92        93  \\\n",
       "0   2.519238  0.883271 -0.243011  ...  1.027305 -0.684256  0.842344  0.982793   \n",
       "1  -2.711813 -0.082691  0.808056  ...  0.751947 -0.567273 -0.547687 -0.018143   \n",
       "2   2.390739 -0.411772 -0.718464  ...  0.274231 -0.460688  1.217699  0.447533   \n",
       "3  -3.115765 -0.120055  1.104269  ...  0.539762 -0.430546 -0.241169 -0.086811   \n",
       "4  -3.024777 -0.114511  1.100467  ...  0.484436 -0.492954 -0.212373 -0.144619   \n",
       "5  -2.399201 -0.469793  1.280533  ...  0.088165 -0.541312  0.376861 -0.096240   \n",
       "6   4.233354 -0.423073 -1.518789  ...  0.403746 -0.517206  0.977598  0.030805   \n",
       "7   3.812093  0.016572 -1.234429  ...  0.813006 -0.365175  1.063718  0.750750   \n",
       "8   0.868674  0.213349  0.448700  ...  0.851811 -0.615773  0.845860  0.614016   \n",
       "9   0.233549 -0.053218  0.568498  ...  0.304297 -0.734093  0.968042  0.301389   \n",
       "10  1.418516  0.366272  0.062157  ...  0.547454 -0.530424  0.942949  0.606467   \n",
       "11  1.564994 -0.028786 -0.284429  ...  0.569270 -0.733589  1.058700  0.463804   \n",
       "12  0.127638 -0.143598  0.563006  ...  0.309192 -0.657410  1.066946  0.159204   \n",
       "13 -2.851595 -0.366958  1.045451  ...  0.647557 -0.609587 -0.108954 -0.491477   \n",
       "14 -2.938588 -0.376913  1.043629  ...  0.674588 -0.591110 -0.115226 -0.474830   \n",
       "15 -1.409627  0.147173  0.789374  ...  0.671324 -0.648832  0.191926 -0.036511   \n",
       "16 -2.682436 -0.082244  0.782839  ...  0.767075 -0.581743 -0.533779  0.012559   \n",
       "17 -1.955159  0.059869  0.729140  ...  0.535518 -0.717063 -0.123622 -0.132882   \n",
       "18 -2.231887 -0.438067  1.294657  ...  0.016728 -0.569043  0.420864 -0.071543   \n",
       "19  3.183227  0.732018 -0.657939  ...  1.013418 -0.818447  1.032210  0.724327   \n",
       "20  0.541963  0.012372  0.044936  ...  0.606983 -0.308914  0.875686  0.392352   \n",
       "21 -2.841949 -0.106587  1.009048  ...  0.553045 -0.640072 -0.131178 -0.305709   \n",
       "22  3.553332 -0.163262 -1.207438  ...  0.544014 -0.799662  0.991320  0.083721   \n",
       "23 -1.745667 -0.766703  0.677340  ...  0.457688 -0.365134  0.512235 -0.391230   \n",
       "24 -3.410334 -0.891370  1.438627  ... -0.242513 -0.784440  0.184291 -0.792968   \n",
       "\n",
       "          94        95        96        97        98        99  \n",
       "0  -1.115196 -1.389488  0.360621 -2.190310  1.044440  0.344591  \n",
       "1   0.482547  1.292929 -0.480422  2.626672 -1.383779 -0.451343  \n",
       "2  -0.244538 -2.007320  1.051488 -2.756425  2.020143  0.185595  \n",
       "3   0.808656  1.170911 -0.667988  2.736131 -1.349261 -0.597348  \n",
       "4   0.706650  1.106278 -0.596556  2.609393 -1.271026 -0.525063  \n",
       "5   0.815520  0.834893 -0.308112  1.879369  0.026564 -0.450374  \n",
       "6  -0.722312 -2.436873  1.935860 -4.078766  1.318376  0.259708  \n",
       "7  -0.798019 -2.832770  1.128418 -3.906541  1.899582  0.186630  \n",
       "8  -0.545951 -0.934811  0.101101 -0.997692  1.270653  0.073550  \n",
       "9   0.070306 -0.267304  0.279300 -0.504349  0.932093  0.187632  \n",
       "10 -0.452768 -0.862444  0.295688 -1.228855  1.678287  0.308781  \n",
       "11 -0.480641 -1.502399  0.578655 -1.946262  1.206935  0.145237  \n",
       "12 -0.013811 -0.430098  0.264688 -0.373346  0.987297  0.206829  \n",
       "13  0.401165  1.284045 -0.217907  2.398955 -0.757857 -0.243582  \n",
       "14  0.510926  1.310520 -0.239643  2.472327 -0.845485 -0.286182  \n",
       "15  0.031080  0.851078 -0.247260  1.517264 -0.662310  0.013303  \n",
       "16  0.496415  1.274615 -0.468524  2.644280 -1.412417 -0.436565  \n",
       "17  0.254660  1.086064 -0.370138  2.070380 -1.194862 -0.358723  \n",
       "18  0.768963  0.802863 -0.274859  1.764337  0.126429 -0.408897  \n",
       "19 -1.042743 -2.114437  0.730771 -3.172199  0.979456  0.355066  \n",
       "20 -0.076132 -0.785183 -0.033932 -0.792483  0.700628 -0.330127  \n",
       "21  0.480756  1.187944 -0.610885  2.603128 -1.275098 -0.453794  \n",
       "22 -0.774948 -2.030537  1.720451 -3.654131  1.314532  0.437816  \n",
       "23  0.423829  0.277259  0.156696  1.145475  0.995143  0.056791  \n",
       "24  1.372638  1.477131 -0.071839  2.781913 -1.105542 -0.799165  \n",
       "\n",
       "[25 rows x 100 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mags_df = pd.DataFrame (list_of_mag_vectors)\n",
    "mags_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "oriental-concentration",
   "metadata": {},
   "source": [
    "---\n",
    "## Data preprocessing\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "traditional-catalyst",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[ 1.08036059, -0.79333764, -0.53660622, ..., -0.97476679,\n",
       "         0.7086667 ,  1.26536613],\n",
       "       [ 0.24232101,  0.94163843, -0.60586657, ...,  1.03505808,\n",
       "        -1.35285303, -0.98385411],\n",
       "       [-2.69261742, -1.09482215,  0.78460615, ..., -1.21097122,\n",
       "         1.53702388,  0.81606068],\n",
       "       ...,\n",
       "       [-1.40785227, -1.72767858, -1.96697477, ..., -1.58552797,\n",
       "         0.93797092,  1.52880821],\n",
       "       [ 0.19818926, -0.1341928 ,  2.21882058, ...,  0.41704726,\n",
       "         0.66681469,  0.45207539],\n",
       "       [ 0.43035118,  0.99931165,  0.69771964, ...,  1.09983054,\n",
       "        -1.11663466, -1.96675936]])"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mag_scaler = preprocessing.StandardScaler()\n",
    "scaled_mags_df = mag_scaler.fit_transform(mags_df)\n",
    "\n",
    "scaled_mags_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fresh-essex",
   "metadata": {},
   "source": [
    "---\n",
    "## Clustering\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "welcome-thanksgiving",
   "metadata": {},
   "source": [
    "### 1. K-means"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "painted-tiger",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>k_range</th>\n",
       "      <th>k_scores</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>-2500.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>-913.125017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>-596.802598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>-476.311370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>-384.997930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>-318.072367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>-259.527094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>-213.584082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>-183.692437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>-161.485470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>11</td>\n",
       "      <td>-136.850105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>12</td>\n",
       "      <td>-111.842814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>13</td>\n",
       "      <td>-88.444251</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>14</td>\n",
       "      <td>-71.113668</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    k_range     k_scores\n",
       "0         1 -2500.000000\n",
       "1         2  -913.125017\n",
       "2         3  -596.802598\n",
       "3         4  -476.311370\n",
       "4         5  -384.997930\n",
       "5         6  -318.072367\n",
       "6         7  -259.527094\n",
       "7         8  -213.584082\n",
       "8         9  -183.692437\n",
       "9        10  -161.485470\n",
       "10       11  -136.850105\n",
       "11       12  -111.842814\n",
       "12       13   -88.444251\n",
       "13       14   -71.113668"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k_range_end = 15 # Usually it is sqrt(# of mags)\n",
    "\n",
    "k_range = range(1, k_range_end)\n",
    "\n",
    "k_mean_models = [KMeans (n_clusters = i, random_state = SEED) for i in k_range]\n",
    "k_scores = [k_mean_model.fit(scaled_mags_df).score(scaled_mags_df) for k_mean_model in k_mean_models]\n",
    "k_data = pd.DataFrame ({'k_range':k_range, 'k_scores':k_scores})\n",
    "k_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "finite-surname",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-04ea61df9cb94895bca45ed90947080d\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-04ea61df9cb94895bca45ed90947080d\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-04ea61df9cb94895bca45ed90947080d\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.8.1?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function loadScript(lib) {\n",
       "      return new Promise(function(resolve, reject) {\n",
       "        var s = document.createElement('script');\n",
       "        s.src = paths[lib];\n",
       "        s.async = true;\n",
       "        s.onload = () => resolve(paths[lib]);\n",
       "        s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "        document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "      });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else if (typeof vegaEmbed === \"function\") {\n",
       "      displayChart(vegaEmbed);\n",
       "    } else {\n",
       "      loadScript(\"vega\")\n",
       "        .then(() => loadScript(\"vega-lite\"))\n",
       "        .then(() => loadScript(\"vega-embed\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}}, \"data\": {\"name\": \"data-b54816b952ae85852ae14d00d67a2ca8\"}, \"mark\": \"line\", \"encoding\": {\"x\": {\"type\": \"quantitative\", \"field\": \"k_range\"}, \"y\": {\"type\": \"quantitative\", \"field\": \"k_scores\"}}, \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.8.1.json\", \"datasets\": {\"data-b54816b952ae85852ae14d00d67a2ca8\": [{\"k_range\": 1, \"k_scores\": -2500.000000000001}, {\"k_range\": 2, \"k_scores\": -913.1250165126056}, {\"k_range\": 3, \"k_scores\": -596.8025977568058}, {\"k_range\": 4, \"k_scores\": -476.311370074231}, {\"k_range\": 5, \"k_scores\": -384.9979302386223}, {\"k_range\": 6, \"k_scores\": -318.0723673427134}, {\"k_range\": 7, \"k_scores\": -259.52709399581136}, {\"k_range\": 8, \"k_scores\": -213.58408214373264}, {\"k_range\": 9, \"k_scores\": -183.69243664237374}, {\"k_range\": 10, \"k_scores\": -161.48547028728026}, {\"k_range\": 11, \"k_scores\": -136.85010499212098}, {\"k_range\": 12, \"k_scores\": -111.84281414118608}, {\"k_range\": 13, \"k_scores\": -88.44425130869627}, {\"k_range\": 14, \"k_scores\": -71.11366772662606}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.Chart(...)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k_num_chart = alt.Chart(data = k_data).mark_line().encode(\n",
    "    alt.X ('k_range:Q'),\n",
    "    alt.Y ('k_scores:Q')\n",
    ")\n",
    "\n",
    "k_num_chart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "local-socket",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([5, 3, 2, 3, 3, 6, 2, 2, 0, 0, 0, 0, 0, 1, 1, 1, 3, 1, 6, 5, 0, 1,\n",
       "       2, 4, 6], dtype=int32)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# We can see from the chart above that 6 or 7 clusters are optimal for this task (where END = 25 MAGs)\n",
    "num_of_clusters = 7\n",
    "\n",
    "k_means_model = KMeans (n_clusters = num_of_clusters, random_state = SEED)\n",
    "k_means_predicted = k_means_model.fit_predict(scaled_mags_df)\n",
    "k_means_predicted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "cooperative-junior",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-e6890cb8a0304c53adeb9369305309ce\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-e6890cb8a0304c53adeb9369305309ce\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-e6890cb8a0304c53adeb9369305309ce\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.8.1?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function loadScript(lib) {\n",
       "      return new Promise(function(resolve, reject) {\n",
       "        var s = document.createElement('script');\n",
       "        s.src = paths[lib];\n",
       "        s.async = true;\n",
       "        s.onload = () => resolve(paths[lib]);\n",
       "        s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "        document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "      });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else if (typeof vegaEmbed === \"function\") {\n",
       "      displayChart(vegaEmbed);\n",
       "    } else {\n",
       "      loadScript(\"vega\")\n",
       "        .then(() => loadScript(\"vega-lite\"))\n",
       "        .then(() => loadScript(\"vega-embed\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}}, \"data\": {\"name\": \"data-efce506d634eae1db097519b10b460ac\"}, \"mark\": \"circle\", \"encoding\": {\"color\": {\"type\": \"nominal\", \"field\": \"Labels\"}, \"x\": {\"type\": \"quantitative\", \"field\": \"PC_1\"}, \"y\": {\"type\": \"quantitative\", \"field\": \"PC_2\"}}, \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.8.1.json\", \"datasets\": {\"data-efce506d634eae1db097519b10b460ac\": [{\"PC_1\": 8.67595851154927, \"PC_2\": 6.498028439080464, \"Labels\": 5}, {\"PC_1\": -8.04556962889203, \"PC_2\": 1.6242825749924958, \"Labels\": 3}, {\"PC_1\": 10.324004279655782, \"PC_2\": -5.272728892157344, \"Labels\": 2}, {\"PC_1\": -9.251824010959352, \"PC_2\": 0.8909085869087093, \"Labels\": 3}, {\"PC_1\": -8.950284325217487, \"PC_2\": 0.9045663297758316, \"Labels\": 3}, {\"PC_1\": -7.04292947429486, \"PC_2\": -2.0947489484467643, \"Labels\": 6}, {\"PC_1\": 16.060097871625082, \"PC_2\": -4.693188112610714, \"Labels\": 2}, {\"PC_1\": 14.579843863094167, \"PC_2\": -0.953560789836873, \"Labels\": 2}, {\"PC_1\": 3.487721138986819, \"PC_2\": 2.842742115775665, \"Labels\": 0}, {\"PC_1\": 2.0232470735660266, \"PC_2\": 0.6435857795247202, \"Labels\": 0}, {\"PC_1\": 5.026431526116229, \"PC_2\": 0.1509306695482429, \"Labels\": 0}, {\"PC_1\": 7.473867804571189, \"PC_2\": 0.7725614727851433, \"Labels\": 0}, {\"PC_1\": 1.3498435600785317, \"PC_2\": -0.0894473124964643, \"Labels\": 0}, {\"PC_1\": -8.613025416525266, \"PC_2\": -0.7216190288624579, \"Labels\": 1}, {\"PC_1\": -8.765000390490124, \"PC_2\": -0.7497041053138674, \"Labels\": 1}, {\"PC_1\": -4.512213960248289, \"PC_2\": 2.411532594920826, \"Labels\": 1}, {\"PC_1\": -7.979912729850816, \"PC_2\": 1.708001623067547, \"Labels\": 3}, {\"PC_1\": -5.861127397787956, \"PC_2\": 2.0419612532351477, \"Labels\": 1}, {\"PC_1\": -6.573989704490055, \"PC_2\": -2.127193002326341, \"Labels\": 6}, {\"PC_1\": 12.200037961264503, \"PC_2\": 6.0062301674380585, \"Labels\": 5}, {\"PC_1\": 3.3921273838715154, \"PC_2\": 0.40168209257160736, \"Labels\": 0}, {\"PC_1\": -8.412703600322383, \"PC_2\": 1.840859778213692, \"Labels\": 1}, {\"PC_1\": 14.579095361453522, \"PC_2\": -2.35098947239124, \"Labels\": 2}, {\"PC_1\": -5.525061368581925, \"PC_2\": -5.872167462622031, \"Labels\": 4}, {\"PC_1\": -9.638634328172104, \"PC_2\": -3.8125263507740463, \"Labels\": 6}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.Chart(...)"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k_means_chart = visualize_with_pca (scaled_mags_df, k_means_predicted)\n",
    "k_means_chart\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "suspected-revision",
   "metadata": {},
   "source": [
    "### 2. DBSCAN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "ambient-dream",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 0,  1, -1,  1,  1,  1, -1, -1,  0,  0,  0,  0,  0,  1,  1,  1,  1,\n",
       "        1,  1,  0,  0,  1, -1, -1,  1])"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MIN_SAMPLES = 3\n",
    "EPS = 7\n",
    "\n",
    "dbscan_model = DBSCAN(eps = EPS, min_samples = MIN_SAMPLES, n_jobs = num_of_workers)\n",
    "dbscan_predicted = dbscan_model.fit_predict (scaled_mags_df)\n",
    "dbscan_predicted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "realistic-mailman",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[ True  True False  True  True  True False False  True  True  True  True\n",
      "  True  True  True  True  True  True  True False  True  True False False\n",
      "  True]\n"
     ]
    }
   ],
   "source": [
    "dbscan_core_samples = np.zeros_like(dbscan_predicted, dtype=bool)\n",
    "dbscan_core_samples[dbscan_model.core_sample_indices_] = True\n",
    "print(dbscan_core_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "changing-fancy",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Visualize clusters"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "instrumental-walnut",
   "metadata": {},
   "source": [
    "---\n",
    "## Evaluation\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "suburban-application",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
